{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Organize freezed data\n",
    "Aug 19th, 2019\n",
    "\n",
    "Sep 13th, 2019 (data freeze)\n",
    "\n",
    "Fangming Xie\n",
    "\n",
    "- reformat raw data (from the dropbox) into standardized python readable format\n",
    "- use ensembl id to refer genes\n",
    "- preserve content as much as possible, including naming conventions of table header and index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, '/cndd/fangming/CEMBA/snmcseq_dev')\n",
    "import importlib\n",
    "\n",
    "from __init__ import *\n",
    "from __init__jupyterlab import *\n",
    "import snmcseq_utils\n",
    "importlib.reload(snmcseq_utils)\n",
    "import CEMBA_clst_utils\n",
    "from scipy.io import mmread\n",
    "from scipy import sparse\n",
    "import time\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "SRC_DIR = '/cndd/fangming/CEMBA/data/MOp_all/data_freeze'\n",
    "DST_DIR = '/cndd/fangming/CEMBA/data/MOp_all/data_freeze_supptable'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(53379, 6)\n",
      "(53278, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>ENSMUSG00000109644</td>\n",
       "      <td>chr7</td>\n",
       "      <td>45567794</td>\n",
       "      <td>45575327</td>\n",
       "      <td>-</td>\n",
       "      <td>antisense_RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <td>ENSMUSG00000108652</td>\n",
       "      <td>chr7</td>\n",
       "      <td>74818818</td>\n",
       "      <td>74853813</td>\n",
       "      <td>-</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>ENSMUSG00000007777</td>\n",
       "      <td>chr11</td>\n",
       "      <td>51685386</td>\n",
       "      <td>51688874</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>ENSMUSG00000086714</td>\n",
       "      <td>chr2</td>\n",
       "      <td>26445696</td>\n",
       "      <td>26459390</td>\n",
       "      <td>+</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>ENSMUSG00000043644</td>\n",
       "      <td>chr11</td>\n",
       "      <td>120348678</td>\n",
       "      <td>120351190</td>\n",
       "      <td>+</td>\n",
       "      <td>bidirectional_promoter_lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          gene_id    chr      start        end strand  \\\n",
       "gene_name                                                               \n",
       "0610005C13Rik  ENSMUSG00000109644   chr7   45567794   45575327      -   \n",
       "0610006L08Rik  ENSMUSG00000108652   chr7   74818818   74853813      -   \n",
       "0610009B22Rik  ENSMUSG00000007777  chr11   51685386   51688874      -   \n",
       "0610009E02Rik  ENSMUSG00000086714   chr2   26445696   26459390      +   \n",
       "0610009L18Rik  ENSMUSG00000043644  chr11  120348678  120351190      +   \n",
       "\n",
       "                                   gene_type  \n",
       "gene_name                                     \n",
       "0610005C13Rik                  antisense_RNA  \n",
       "0610006L08Rik                        lincRNA  \n",
       "0610009B22Rik                 protein_coding  \n",
       "0610009E02Rik           processed_transcript  \n",
       "0610009L18Rik  bidirectional_promoter_lncRNA  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = PATH_GENEBODY_ANNOTATION \n",
    "df_genes = pd.read_csv(f, sep=\"\\t\")\n",
    "df_genes['gene_id'] = df_genes['gene_id'].apply(lambda x: x.split('.')[0])\n",
    "df_genes = df_genes.set_index('gene_id')\n",
    "df_genes_v2 = df_genes.reset_index().groupby('gene_name').first()\n",
    "print(df_genes.shape)\n",
    "df_genes.head()\n",
    "print(df_genes_v2.shape)\n",
    "df_genes_v2.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 10x_nuclei_v3_macosko \n",
    "- 2 metadata .csv\n",
    "- 1 data matrix in .rds format\n",
    "- concatenate male and female, get autosomo genes only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_supptable/10x_nuclei_v3_macosko_metadata.tsv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze/10x_nuclei_v3_macosko/./reformatted/raw_female.mtx\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_supptable/10x_nuclei_v3_macosko_counts.npz\n"
     ]
    }
   ],
   "source": [
    "dataset_name = '10x_nuclei_v3_macosko'\n",
    "\n",
    "\n",
    "# outdated\n",
    "f_meta = os.path.join(SRC_DIR, dataset_name, 'motor_cortex_v3_mop_data_metadata.csv') \n",
    "# f_meta4 = os.path.join(SRC_DIR, dataset_name, 'zizhen_processing', 'sample_metadata.csv') \n",
    "\n",
    "# zizhen's list \n",
    "f_meta1 = os.path.join(SRC_DIR, dataset_name, 'zizhen_processing', 'cluster.membership.csv') \n",
    "f_meta2 = os.path.join(SRC_DIR, dataset_name, 'zizhen_processing', 'cluster.annotation.csv') \n",
    "f_meta3 = os.path.join(SRC_DIR, dataset_name, 'zizhen_processing', 'QC.csv') \n",
    "\n",
    "f_data = os.path.join(SRC_DIR, dataset_name, './reformatted/raw_female.mtx') \n",
    "f_data_gene = os.path.join(SRC_DIR, dataset_name, './reformatted/raw_female_gene.csv') \n",
    "f_data_cell = os.path.join(SRC_DIR, dataset_name, './reformatted/raw_female_cell.csv') \n",
    "\n",
    "f_data2 = os.path.join(SRC_DIR, dataset_name, './reformatted/raw_male.mtx') \n",
    "f_data2_gene = os.path.join(SRC_DIR, dataset_name, './reformatted/raw_male_gene.csv') \n",
    "f_data2_cell = os.path.join(SRC_DIR, dataset_name, './reformatted/raw_male_cell.csv') \n",
    "\n",
    "fout_meta = os.path.join(DST_DIR, dataset_name+'_metadata.tsv') \n",
    "fout_data = os.path.join(DST_DIR, dataset_name+'_counts.npz') \n",
    "fout_data_gene = os.path.join(DST_DIR, dataset_name+'_counts.gene') \n",
    "fout_data_cell = os.path.join(DST_DIR, dataset_name+'_counts.cell') \n",
    "\n",
    "# print(f_meta)\n",
    "print(fout_meta)\n",
    "print(f_data)\n",
    "print(fout_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "672.5768630504608\n",
      "1225.8807742595673\n"
     ]
    }
   ],
   "source": [
    "ti = time.time()\n",
    "# data = mmread(f_data)\n",
    "gene = pd.read_csv(f_data_gene)['x'].values\n",
    "cell = pd.read_csv(f_data_cell)['x'].values\n",
    "print(time.time()-ti)\n",
    "\n",
    "data2 = mmread(f_data2)\n",
    "gene2 = pd.read_csv(f_data2_gene)['x'].values\n",
    "cell2 = pd.read_csv(f_data2_cell)['x'].values\n",
    "print(time.time()-ti)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# assert np.all(gene==gene2)\n",
    "# gene.shape\n",
    "# gene2.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(27574,) (27495,) (24809,) (24809,)\n",
      "(24809,) (24809,)\n",
      "(24809, 215823)\n"
     ]
    }
   ],
   "source": [
    "# autosome genes\n",
    "chromosomes = ['chr{}'.format(chrom) for chrom in snmcseq_utils.get_mouse_chromosomes(include_x=False)]\n",
    "gene_table1 = df_genes_v2[df_genes_v2['chr'].isin(chromosomes)].reindex(gene).dropna()\n",
    "gene_table2 = df_genes_v2[df_genes_v2['chr'].isin(chromosomes)].reindex(gene2).dropna()\n",
    "\n",
    "genes_common = np.intersect1d(gene_table1.index.values, gene_table2.index.values)\n",
    "genes_common_ids = gene_table1.loc[genes_common, 'gene_id'].values \n",
    "print(gene.shape, \n",
    "      gene2.shape, \n",
    "      genes_common.shape, \n",
    "      genes_common_ids.shape, \n",
    "     )\n",
    "\n",
    "selected_genes_idx1 = snmcseq_utils.get_index_from_array(gene, genes_common)\n",
    "selected_genes_idx2 = snmcseq_utils.get_index_from_array(gene2, genes_common)\n",
    "print(selected_genes_idx1.shape, \n",
    "      selected_genes_idx2.shape, \n",
    "     )\n",
    "\n",
    "# \n",
    "data_comb = sparse.hstack([\n",
    "                data.tocsr()[selected_genes_idx1,:], data2.tocsr()[selected_genes_idx2,:]\n",
    "            ])\n",
    "print(data_comb.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "990.3496780395508\n"
     ]
    }
   ],
   "source": [
    "# output\n",
    "# save \n",
    "ti = time.time()\n",
    "gc_mat = GC_matrix(genes_common_ids, \n",
    "                   np.hstack([cell, cell2]), \n",
    "                   data_comb, \n",
    "                  )\n",
    "snmcseq_utils.save_gc_matrix(gc_mat, fout_data_gene, fout_data_cell, fout_data)\n",
    "print(time.time()-ti)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(215823, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nUMI</th>\n",
       "      <th>nGene</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pBICCNsMMrMOpRMiM003d190318_AAACCCACACAAGCTT</th>\n",
       "      <td>4859</td>\n",
       "      <td>2336</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pBICCNsMMrMOpRMiM003d190318_AAACCCACAGTCCCGA</th>\n",
       "      <td>24082</td>\n",
       "      <td>6164</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pBICCNsMMrMOpRMiM003d190318_AAACCCAGTGAATTAG</th>\n",
       "      <td>5194</td>\n",
       "      <td>2519</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pBICCNsMMrMOpRMiM003d190318_AAACCCAGTGGTAACG</th>\n",
       "      <td>1343</td>\n",
       "      <td>998</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pBICCNsMMrMOpRMiM003d190318_AAACCCAGTTCCAAAC</th>\n",
       "      <td>9022</td>\n",
       "      <td>3558</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               nUMI  nGene   sex\n",
       "pBICCNsMMrMOpRMiM003d190318_AAACCCACACAAGCTT   4859   2336  MALE\n",
       "pBICCNsMMrMOpRMiM003d190318_AAACCCACAGTCCCGA  24082   6164  MALE\n",
       "pBICCNsMMrMOpRMiM003d190318_AAACCCAGTGAATTAG   5194   2519  MALE\n",
       "pBICCNsMMrMOpRMiM003d190318_AAACCCAGTGGTAACG   1343    998  MALE\n",
       "pBICCNsMMrMOpRMiM003d190318_AAACCCAGTTCCAAAC   9022   3558  MALE"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # assemble metadata (keep order, add male and female)\n",
    "\n",
    "meta = pd.read_csv(f_meta, index_col=0).rename(columns={'dataset': 'sex'})[['nUMI', 'nGene', 'sex']]\n",
    "# meta.index.name = 'cell'\n",
    "# meta = meta.loc[np.hstack([cell, cell2])]\n",
    "print(meta.shape)\n",
    "meta.head()\n",
    "\n",
    "# # meta.to_csv(fout_meta, sep=\"\\t\", header=True, index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(159738, 19)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>QC</th>\n",
       "      <th>cluster_label</th>\n",
       "      <th>cluster_color</th>\n",
       "      <th>class_label</th>\n",
       "      <th>subclass_label</th>\n",
       "      <th>size</th>\n",
       "      <th>gene.counts</th>\n",
       "      <th>umi.counts</th>\n",
       "      <th>Broad.QC.doublet</th>\n",
       "      <th>Broad.QC.Mito</th>\n",
       "      <th>Broad.passQC</th>\n",
       "      <th>MALE</th>\n",
       "      <th>Comb.QC</th>\n",
       "      <th>cl</th>\n",
       "      <th>nUMI</th>\n",
       "      <th>nGene</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pBICCNsMMrMOpRMiM003d190318_AAACCCACACAAGCTT</td>\n",
       "      <td>90</td>\n",
       "      <td>1</td>\n",
       "      <td>Oligo Opalin_4</td>\n",
       "      <td>#474662</td>\n",
       "      <td>Non-Neuronal</td>\n",
       "      <td>Oligo</td>\n",
       "      <td>16566</td>\n",
       "      <td>1740.348364</td>\n",
       "      <td>3419.937704</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.487323</td>\n",
       "      <td>NaN</td>\n",
       "      <td>150</td>\n",
       "      <td>4859</td>\n",
       "      <td>2336</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pBICCNsMMrMOpRMiM003d190318_AAAGGATCAGACCGCT</td>\n",
       "      <td>90</td>\n",
       "      <td>19</td>\n",
       "      <td>Oligo Opalin_4</td>\n",
       "      <td>#474662</td>\n",
       "      <td>Non-Neuronal</td>\n",
       "      <td>Oligo</td>\n",
       "      <td>16566</td>\n",
       "      <td>1740.348364</td>\n",
       "      <td>3419.937704</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.487323</td>\n",
       "      <td>NaN</td>\n",
       "      <td>150</td>\n",
       "      <td>3172</td>\n",
       "      <td>1739</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pBICCNsMMrMOpRMiM003d190318_AAAGGATCATGGAACG</td>\n",
       "      <td>90</td>\n",
       "      <td>20</td>\n",
       "      <td>Oligo Opalin_4</td>\n",
       "      <td>#474662</td>\n",
       "      <td>Non-Neuronal</td>\n",
       "      <td>Oligo</td>\n",
       "      <td>16566</td>\n",
       "      <td>1740.348364</td>\n",
       "      <td>3419.937704</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.487323</td>\n",
       "      <td>NaN</td>\n",
       "      <td>150</td>\n",
       "      <td>6306</td>\n",
       "      <td>2760</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pBICCNsMMrMOpRMiM003d190318_AAAGGTAGTAATACCC</td>\n",
       "      <td>90</td>\n",
       "      <td>37</td>\n",
       "      <td>Oligo Opalin_4</td>\n",
       "      <td>#474662</td>\n",
       "      <td>Non-Neuronal</td>\n",
       "      <td>Oligo</td>\n",
       "      <td>16566</td>\n",
       "      <td>1740.348364</td>\n",
       "      <td>3419.937704</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.487323</td>\n",
       "      <td>NaN</td>\n",
       "      <td>150</td>\n",
       "      <td>3310</td>\n",
       "      <td>1806</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pBICCNsMMrMOpRMiM003d190318_AAAGTCCCAAGGTCAG</td>\n",
       "      <td>90</td>\n",
       "      <td>39</td>\n",
       "      <td>Oligo Opalin_4</td>\n",
       "      <td>#474662</td>\n",
       "      <td>Non-Neuronal</td>\n",
       "      <td>Oligo</td>\n",
       "      <td>16566</td>\n",
       "      <td>1740.348364</td>\n",
       "      <td>3419.937704</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.487323</td>\n",
       "      <td>NaN</td>\n",
       "      <td>150</td>\n",
       "      <td>2843</td>\n",
       "      <td>1543</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           cell  cluster_id  QC  \\\n",
       "0  pBICCNsMMrMOpRMiM003d190318_AAACCCACACAAGCTT          90   1   \n",
       "1  pBICCNsMMrMOpRMiM003d190318_AAAGGATCAGACCGCT          90  19   \n",
       "2  pBICCNsMMrMOpRMiM003d190318_AAAGGATCATGGAACG          90  20   \n",
       "3  pBICCNsMMrMOpRMiM003d190318_AAAGGTAGTAATACCC          90  37   \n",
       "4  pBICCNsMMrMOpRMiM003d190318_AAAGTCCCAAGGTCAG          90  39   \n",
       "\n",
       "    cluster_label cluster_color   class_label subclass_label   size  \\\n",
       "0  Oligo Opalin_4       #474662  Non-Neuronal          Oligo  16566   \n",
       "1  Oligo Opalin_4       #474662  Non-Neuronal          Oligo  16566   \n",
       "2  Oligo Opalin_4       #474662  Non-Neuronal          Oligo  16566   \n",
       "3  Oligo Opalin_4       #474662  Non-Neuronal          Oligo  16566   \n",
       "4  Oligo Opalin_4       #474662  Non-Neuronal          Oligo  16566   \n",
       "\n",
       "   gene.counts   umi.counts  Broad.QC.doublet  Broad.QC.Mito  Broad.passQC  \\\n",
       "0  1740.348364  3419.937704              0.06            0.0          0.94   \n",
       "1  1740.348364  3419.937704              0.06            0.0          0.94   \n",
       "2  1740.348364  3419.937704              0.06            0.0          0.94   \n",
       "3  1740.348364  3419.937704              0.06            0.0          0.94   \n",
       "4  1740.348364  3419.937704              0.06            0.0          0.94   \n",
       "\n",
       "       MALE Comb.QC   cl  nUMI  nGene   sex  \n",
       "0  0.487323     NaN  150  4859   2336  MALE  \n",
       "1  0.487323     NaN  150  3172   1739  MALE  \n",
       "2  0.487323     NaN  150  6306   2760  MALE  \n",
       "3  0.487323     NaN  150  3310   1806  MALE  \n",
       "4  0.487323     NaN  150  2843   1543  MALE  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "meta1 = pd.read_csv(f_meta1, names=['cell', 'cluster_id'], header=0) # cluster assignment\n",
    "meta2 = pd.read_csv(f_meta2) # cluster annot [all metrics]\n",
    "meta3 = pd.read_csv(f_meta3, names=['QC', 'cell'], header=0) # QC'ed cells\n",
    "meta1 = pd.merge(meta1, meta3, on='cell') # cluster id on QC'ed cells\n",
    "meta1 = pd.merge(meta1, meta2, on='cluster_id')\n",
    "meta1 = pd.merge(meta1, meta, left_on='cell', right_index=True)\n",
    "\n",
    "print(meta1.shape)\n",
    "meta1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta1.to_csv(fout_meta, sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 10x_nuclei_v3\n",
    "\n",
    "- gene_ids\n",
    "- all cells passed QC (QC.csv) metadata\n",
    "\n",
    "- all cells data\n",
    "- output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze/10x_nuclei_v3/cluster.membership.csv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_supptable/10x_nuclei_v3_metadata.tsv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze/10x_nuclei_v3/matrix.mtx.gz\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_supptable/10x_nuclei_v3_counts.npz\n"
     ]
    }
   ],
   "source": [
    "dataset_name = '10x_nuclei_v3'\n",
    "\n",
    "f_meta = os.path.join(SRC_DIR, dataset_name, 'cluster.membership.csv') \n",
    "f_meta2 = os.path.join(SRC_DIR, dataset_name, 'cluster.annotation.csv') \n",
    "f_meta3 = os.path.join(SRC_DIR, dataset_name, 'QC.csv') \n",
    "\n",
    "f_data = os.path.join(SRC_DIR, dataset_name, 'matrix.mtx.gz') \n",
    "f_data_gene = os.path.join(SRC_DIR, dataset_name, 'features.tsv.gz') \n",
    "f_data_cell = os.path.join(SRC_DIR, dataset_name, 'barcodes.tsv.gz') \n",
    "\n",
    "fout_meta = os.path.join(DST_DIR, dataset_name+'_metadata.tsv') \n",
    "fout_data = os.path.join(DST_DIR, dataset_name+'_counts.npz') \n",
    "fout_data_gene = os.path.join(DST_DIR, dataset_name+'_counts.gene') \n",
    "fout_data_cell = os.path.join(DST_DIR, dataset_name+'_counts.cell') \n",
    "\n",
    "print(f_meta)\n",
    "print(fout_meta)\n",
    "print(f_data)\n",
    "print(fout_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(40166, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>QC</th>\n",
       "      <th>cluster_label</th>\n",
       "      <th>subclass_label</th>\n",
       "      <th>class_label</th>\n",
       "      <th>cluster_color</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAAGCTCTTCC-1L8TX_181211_01_A02</td>\n",
       "      <td>59</td>\n",
       "      <td>1</td>\n",
       "      <td>L6b Shisa6</td>\n",
       "      <td>L6b</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#2B9880</td>\n",
       "      <td>247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAGGGCTCGCGATCG-1L8TX_181211_01_A02</td>\n",
       "      <td>59</td>\n",
       "      <td>47</td>\n",
       "      <td>L6b Shisa6</td>\n",
       "      <td>L6b</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#2B9880</td>\n",
       "      <td>247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAGTGACATCGCTGG-1L8TX_181211_01_A02</td>\n",
       "      <td>59</td>\n",
       "      <td>80</td>\n",
       "      <td>L6b Shisa6</td>\n",
       "      <td>L6b</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#2B9880</td>\n",
       "      <td>247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AACAACCGTGCTCTCT-1L8TX_181211_01_A02</td>\n",
       "      <td>59</td>\n",
       "      <td>108</td>\n",
       "      <td>L6b Shisa6</td>\n",
       "      <td>L6b</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#2B9880</td>\n",
       "      <td>247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AACAAGAAGCCTGCCA-1L8TX_181211_01_A02</td>\n",
       "      <td>59</td>\n",
       "      <td>112</td>\n",
       "      <td>L6b Shisa6</td>\n",
       "      <td>L6b</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#2B9880</td>\n",
       "      <td>247</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   cell  cluster_id   QC cluster_label  \\\n",
       "0  AAACCCAAGCTCTTCC-1L8TX_181211_01_A02          59    1    L6b Shisa6   \n",
       "1  AAAGGGCTCGCGATCG-1L8TX_181211_01_A02          59   47    L6b Shisa6   \n",
       "2  AAAGTGACATCGCTGG-1L8TX_181211_01_A02          59   80    L6b Shisa6   \n",
       "3  AACAACCGTGCTCTCT-1L8TX_181211_01_A02          59  108    L6b Shisa6   \n",
       "4  AACAAGAAGCCTGCCA-1L8TX_181211_01_A02          59  112    L6b Shisa6   \n",
       "\n",
       "  subclass_label    class_label cluster_color  size  \n",
       "0            L6b  Glutamatergic       #2B9880   247  \n",
       "1            L6b  Glutamatergic       #2B9880   247  \n",
       "2            L6b  Glutamatergic       #2B9880   247  \n",
       "3            L6b  Glutamatergic       #2B9880   247  \n",
       "4            L6b  Glutamatergic       #2B9880   247  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv(f_meta, names=['cell', 'cluster_id'], header=0) \n",
    "meta2 = pd.read_csv(f_meta2)\n",
    "meta3 = pd.read_csv(f_meta3, names=['QC', 'cell'], header=0)\n",
    "meta = pd.merge(meta, meta3, on='cell')\n",
    "meta = pd.merge(meta, meta2, on='cluster_id')\n",
    "\n",
    "print(meta.shape)\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(90266,)\n",
      "(31053,)\n",
      "(31053, 90266)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['ENSMUSG00000051951', 'ENSMUSG00000089699', 'ENSMUSG00000102343',\n",
       "       ..., 'ENSMUSG00000063897', 'ENSMUSG00000096730',\n",
       "       'ENSMUSG00000095742'], dtype=object)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_cell = snmcseq_utils.import_single_textcol(f_data_cell)\n",
    "data_gene = snmcseq_utils.import_single_textcol(f_data_gene, col=0)\n",
    "data = mmread(f_data)\n",
    "print(data_cell.shape)\n",
    "print(data_gene.shape)\n",
    "print(data.shape)\n",
    "data_cell\n",
    "data_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "gc_mat = GC_matrix(data_gene, data_cell, data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta.to_csv(fout_meta, sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "snmcseq_utils.save_gc_matrix(gc_mat, \n",
    "                             fout_data_gene, fout_data_cell, fout_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 10x_cells_v3 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze/10x_cells_v3/cluster.membership.csv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/10x_cells_v3_metadata.tsv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze/10x_cells_v3/matrix.mtx.gz\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/10x_cells_v3_counts.npz\n"
     ]
    }
   ],
   "source": [
    "dataset_name = '10x_cells_v3'\n",
    "\n",
    "f_meta = os.path.join(SRC_DIR, dataset_name, 'cluster.membership.csv') \n",
    "f_meta2 = os.path.join(SRC_DIR, dataset_name, 'cluster.annotation.csv') \n",
    "f_meta3 = os.path.join(SRC_DIR, dataset_name, 'QC.csv') \n",
    "\n",
    "f_data = os.path.join(SRC_DIR, dataset_name, 'matrix.mtx.gz') \n",
    "f_data_gene = os.path.join(SRC_DIR, dataset_name, 'features.tsv.gz') \n",
    "f_data_cell = os.path.join(SRC_DIR, dataset_name, 'barcodes.tsv.gz') \n",
    "\n",
    "fout_meta = os.path.join(DST_DIR, dataset_name+'_metadata.tsv') \n",
    "fout_data = os.path.join(DST_DIR, dataset_name+'_counts.npz') \n",
    "fout_data_gene = os.path.join(DST_DIR, dataset_name+'_counts.gene') \n",
    "fout_data_cell = os.path.join(DST_DIR, dataset_name+'_counts.cell') \n",
    "\n",
    "print(f_meta)\n",
    "print(fout_meta)\n",
    "print(f_data)\n",
    "print(fout_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(71183, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>QC</th>\n",
       "      <th>cluster_label</th>\n",
       "      <th>subclass_label</th>\n",
       "      <th>class_label</th>\n",
       "      <th>cluster_color</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAAGCTTCATG-1L8TX_181211_01_G12</td>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>L5 IT Tcap_2</td>\n",
       "      <td>L5 IT</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#52CA74</td>\n",
       "      <td>17334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCACACCAGCCA-1L8TX_181211_01_G12</td>\n",
       "      <td>42</td>\n",
       "      <td>3</td>\n",
       "      <td>L5 IT Tcap_2</td>\n",
       "      <td>L5 IT</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#52CA74</td>\n",
       "      <td>17334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACGAACAACGATTC-1L8TX_181211_01_G12</td>\n",
       "      <td>42</td>\n",
       "      <td>8</td>\n",
       "      <td>L5 IT Tcap_2</td>\n",
       "      <td>L5 IT</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#52CA74</td>\n",
       "      <td>17334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACGAATCTCGTGAA-1L8TX_181211_01_G12</td>\n",
       "      <td>42</td>\n",
       "      <td>11</td>\n",
       "      <td>L5 IT Tcap_2</td>\n",
       "      <td>L5 IT</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#52CA74</td>\n",
       "      <td>17334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGCTGTAGTCACT-1L8TX_181211_01_G12</td>\n",
       "      <td>42</td>\n",
       "      <td>14</td>\n",
       "      <td>L5 IT Tcap_2</td>\n",
       "      <td>L5 IT</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#52CA74</td>\n",
       "      <td>17334</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   cell  cluster_id  QC cluster_label  \\\n",
       "0  AAACCCAAGCTTCATG-1L8TX_181211_01_G12          42   1  L5 IT Tcap_2   \n",
       "1  AAACCCACACCAGCCA-1L8TX_181211_01_G12          42   3  L5 IT Tcap_2   \n",
       "2  AAACGAACAACGATTC-1L8TX_181211_01_G12          42   8  L5 IT Tcap_2   \n",
       "3  AAACGAATCTCGTGAA-1L8TX_181211_01_G12          42  11  L5 IT Tcap_2   \n",
       "4  AAACGCTGTAGTCACT-1L8TX_181211_01_G12          42  14  L5 IT Tcap_2   \n",
       "\n",
       "  subclass_label    class_label cluster_color   size  \n",
       "0          L5 IT  Glutamatergic       #52CA74  17334  \n",
       "1          L5 IT  Glutamatergic       #52CA74  17334  \n",
       "2          L5 IT  Glutamatergic       #52CA74  17334  \n",
       "3          L5 IT  Glutamatergic       #52CA74  17334  \n",
       "4          L5 IT  Glutamatergic       #52CA74  17334  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv(f_meta, names=['cell', 'cluster_id'], header=0) \n",
    "meta2 = pd.read_csv(f_meta2)\n",
    "meta3 = pd.read_csv(f_meta3, names=['QC', 'cell'], header=0)\n",
    "meta = pd.merge(meta, meta3, on='cell')\n",
    "meta = pd.merge(meta, meta2, on='cluster_id')\n",
    "\n",
    "print(meta.shape)\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(176584,)\n",
      "(31053,)\n",
      "(31053, 176584)\n"
     ]
    }
   ],
   "source": [
    "data_cell = snmcseq_utils.import_single_textcol(f_data_cell)\n",
    "data_gene = snmcseq_utils.import_single_textcol(f_data_gene, col=0)\n",
    "data = mmread(f_data)\n",
    "print(data_cell.shape)\n",
    "print(data_gene.shape)\n",
    "print(data.shape)\n",
    "data_cell\n",
    "data_gene\n",
    "\n",
    "gc_mat = GC_matrix(data_gene, data_cell, data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta.to_csv(fout_meta, sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "snmcseq_utils.save_gc_matrix(gc_mat, \n",
    "                             fout_data_gene, fout_data_cell, fout_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 10x_cells_v2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze/10x_cells_v2/cluster.membership.csv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/10x_cells_v2_metadata.tsv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze/10x_cells_v2/matrix.mtx.gz\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/10x_cells_v2_counts.npz\n"
     ]
    }
   ],
   "source": [
    "dataset_name = '10x_cells_v2'\n",
    "\n",
    "f_meta = os.path.join(SRC_DIR, dataset_name, 'cluster.membership.csv') \n",
    "f_meta2 = os.path.join(SRC_DIR, dataset_name, 'cluster.annotation.csv') \n",
    "f_meta3 = os.path.join(SRC_DIR, dataset_name, 'QC.csv') \n",
    "\n",
    "f_data = os.path.join(SRC_DIR, dataset_name, 'matrix.mtx.gz') \n",
    "f_data_gene = os.path.join(SRC_DIR, dataset_name, 'features.tsv.gz') \n",
    "f_data_cell = os.path.join(SRC_DIR, dataset_name, 'barcodes.tsv.gz') \n",
    "\n",
    "fout_meta = os.path.join(DST_DIR, dataset_name+'_metadata.tsv') \n",
    "fout_data = os.path.join(DST_DIR, dataset_name+'_counts.npz') \n",
    "fout_data_gene = os.path.join(DST_DIR, dataset_name+'_counts.gene') \n",
    "fout_data_cell = os.path.join(DST_DIR, dataset_name+'_counts.cell') \n",
    "\n",
    "print(f_meta)\n",
    "print(fout_meta)\n",
    "print(f_data)\n",
    "print(fout_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(122641, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>QC</th>\n",
       "      <th>cluster_label</th>\n",
       "      <th>subclass_label</th>\n",
       "      <th>class_label</th>\n",
       "      <th>cluster_color</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGAGGAGTCTG-1L8TX_171026_01_F03</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Sncg Slc17a8</td>\n",
       "      <td>Sncg</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#9440F3</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACCTTTAGTACAGCAG-1L8TX_171026_01_F03</td>\n",
       "      <td>9</td>\n",
       "      <td>673</td>\n",
       "      <td>Sncg Slc17a8</td>\n",
       "      <td>Sncg</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#9440F3</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACGATACCACCCAGTG-1L8TX_171026_01_F03</td>\n",
       "      <td>9</td>\n",
       "      <td>724</td>\n",
       "      <td>Sncg Slc17a8</td>\n",
       "      <td>Sncg</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#9440F3</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACGGGTCAGTGGGCTA-1L8TX_171026_01_F03</td>\n",
       "      <td>9</td>\n",
       "      <td>885</td>\n",
       "      <td>Sncg Slc17a8</td>\n",
       "      <td>Sncg</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#9440F3</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGCAGCCCAGTTCATG-1L8TX_171026_01_F03</td>\n",
       "      <td>9</td>\n",
       "      <td>1205</td>\n",
       "      <td>Sncg Slc17a8</td>\n",
       "      <td>Sncg</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#9440F3</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   cell  cluster_id    QC cluster_label  \\\n",
       "0  AAACCTGAGGAGTCTG-1L8TX_171026_01_F03           9     1  Sncg Slc17a8   \n",
       "1  ACCTTTAGTACAGCAG-1L8TX_171026_01_F03           9   673  Sncg Slc17a8   \n",
       "2  ACGATACCACCCAGTG-1L8TX_171026_01_F03           9   724  Sncg Slc17a8   \n",
       "3  ACGGGTCAGTGGGCTA-1L8TX_171026_01_F03           9   885  Sncg Slc17a8   \n",
       "4  AGCAGCCCAGTTCATG-1L8TX_171026_01_F03           9  1205  Sncg Slc17a8   \n",
       "\n",
       "  subclass_label class_label cluster_color  size  \n",
       "0           Sncg   GABAergic       #9440F3   293  \n",
       "1           Sncg   GABAergic       #9440F3   293  \n",
       "2           Sncg   GABAergic       #9440F3   293  \n",
       "3           Sncg   GABAergic       #9440F3   293  \n",
       "4           Sncg   GABAergic       #9440F3   293  "
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv(f_meta, names=['cell', 'cluster_id'], header=0) \n",
    "meta2 = pd.read_csv(f_meta2)\n",
    "meta3 = pd.read_csv(f_meta3, names=['QC', 'cell'], header=0)\n",
    "meta = pd.merge(meta, meta3, on='cell')\n",
    "meta = pd.merge(meta, meta2, on='cluster_id')\n",
    "\n",
    "print(meta.shape)\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(145748,)\n",
      "(31053,)\n",
      "(31053, 145748)\n"
     ]
    }
   ],
   "source": [
    "data_cell = snmcseq_utils.import_single_textcol(f_data_cell)\n",
    "data_gene = snmcseq_utils.import_single_textcol(f_data_gene, col=0)\n",
    "data = mmread(f_data)\n",
    "print(data_cell.shape)\n",
    "print(data_gene.shape)\n",
    "print(data.shape)\n",
    "data_cell\n",
    "data_gene\n",
    "\n",
    "gc_mat = GC_matrix(data_gene, data_cell, data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta.to_csv(fout_meta, sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "snmcseq_utils.save_gc_matrix(gc_mat, \n",
    "                             fout_data_gene, fout_data_cell, fout_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Smart-cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze/smarter_cells/cluster.membership.csv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/smarter_cells_metadata.tsv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/smarter_cells_raw.npz\n"
     ]
    }
   ],
   "source": [
    "dataset_name = 'smarter_cells'\n",
    "\n",
    "f_meta = os.path.join(SRC_DIR, dataset_name, 'cluster.membership.csv') \n",
    "f_meta2 = os.path.join(SRC_DIR, dataset_name, 'cluster.annotation.csv') \n",
    "f_meta3 = os.path.join(SRC_DIR, dataset_name, 'QC.csv') \n",
    "f_meta4 = os.path.join(SRC_DIR, dataset_name, 'sample_metadata.csv.gz') \n",
    "\n",
    "f_data_intron = os.path.join(SRC_DIR, dataset_name, 'intron.counts.csv.gz') \n",
    "f_data_exon = os.path.join(SRC_DIR, dataset_name, 'exon.counts.csv.gz') \n",
    "\n",
    "fout_meta = os.path.join(DST_DIR, dataset_name+'_metadata.tsv') \n",
    "fout_data = os.path.join(DST_DIR, dataset_name+'_raw.npz') \n",
    "fout_data_gene = os.path.join(DST_DIR, dataset_name+'_raw.gene') \n",
    "fout_data_cell = os.path.join(DST_DIR, dataset_name+'_raw.cell') \n",
    "\n",
    "print(f_meta)\n",
    "print(fout_meta)\n",
    "# print(f_data)\n",
    "print(fout_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6288, 129)\n",
      "(6288, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>QC</th>\n",
       "      <th>cluster_label</th>\n",
       "      <th>subclass_label</th>\n",
       "      <th>class_label</th>\n",
       "      <th>cluster_color</th>\n",
       "      <th>size</th>\n",
       "      <th>ar_id</th>\n",
       "      <th>exp_component_vendor_name</th>\n",
       "      <th>...</th>\n",
       "      <th>Injection_type</th>\n",
       "      <th>NewBatch</th>\n",
       "      <th>Vendor</th>\n",
       "      <th>MultiPlex</th>\n",
       "      <th>percent_reads_aligned_to_rrna_rmsk</th>\n",
       "      <th>percent_reads_aligned_to_trna_rmsk</th>\n",
       "      <th>percent_reads_aligned_to_ncrna</th>\n",
       "      <th>percent_reads_aligned_to_introns</th>\n",
       "      <th>percent_reads_aligned_to_intergenic</th>\n",
       "      <th>percent_reads_aligned_unique</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LS-15395_S41_E1-50</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>Sst Myh8_1</td>\n",
       "      <td>Sst</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#FF7F2C</td>\n",
       "      <td>136</td>\n",
       "      <td>543183476</td>\n",
       "      <td>LS-15395-41</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HistoricalData</td>\n",
       "      <td>Covance</td>\n",
       "      <td>index96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.162980</td>\n",
       "      <td>9.855384</td>\n",
       "      <td>3.264771</td>\n",
       "      <td>84.993663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LS-15395_S42_E1-50</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>Sst Myh8_1</td>\n",
       "      <td>Sst</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#FF7F2C</td>\n",
       "      <td>136</td>\n",
       "      <td>543183474</td>\n",
       "      <td>LS-15395-42</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HistoricalData</td>\n",
       "      <td>Covance</td>\n",
       "      <td>index96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16.281620</td>\n",
       "      <td>6.763949</td>\n",
       "      <td>2.856108</td>\n",
       "      <td>87.252036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LS-15501_S28_E1-50</td>\n",
       "      <td>19</td>\n",
       "      <td>94</td>\n",
       "      <td>Sst Myh8_1</td>\n",
       "      <td>Sst</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#FF7F2C</td>\n",
       "      <td>136</td>\n",
       "      <td>544930823</td>\n",
       "      <td>LS-15501-28</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HistoricalData</td>\n",
       "      <td>Covance</td>\n",
       "      <td>index96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17.066841</td>\n",
       "      <td>3.311346</td>\n",
       "      <td>2.437971</td>\n",
       "      <td>86.831969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LS-15501_S81_E1-50</td>\n",
       "      <td>19</td>\n",
       "      <td>120</td>\n",
       "      <td>Sst Myh8_1</td>\n",
       "      <td>Sst</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#FF7F2C</td>\n",
       "      <td>136</td>\n",
       "      <td>544930707</td>\n",
       "      <td>LS-15501-81</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HistoricalData</td>\n",
       "      <td>Covance</td>\n",
       "      <td>index96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.866006</td>\n",
       "      <td>11.769028</td>\n",
       "      <td>3.699682</td>\n",
       "      <td>85.252590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LS-15524_S21_E1-50</td>\n",
       "      <td>19</td>\n",
       "      <td>267</td>\n",
       "      <td>Sst Myh8_1</td>\n",
       "      <td>Sst</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#FF7F2C</td>\n",
       "      <td>136</td>\n",
       "      <td>549999237</td>\n",
       "      <td>LS-15524-21</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HistoricalData</td>\n",
       "      <td>Covance</td>\n",
       "      <td>index96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.912637</td>\n",
       "      <td>5.923318</td>\n",
       "      <td>2.385434</td>\n",
       "      <td>82.503705</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 129 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 cell  cluster_id   QC cluster_label subclass_label  \\\n",
       "0  LS-15395_S41_E1-50          19    1    Sst Myh8_1            Sst   \n",
       "1  LS-15395_S42_E1-50          19    2    Sst Myh8_1            Sst   \n",
       "2  LS-15501_S28_E1-50          19   94    Sst Myh8_1            Sst   \n",
       "3  LS-15501_S81_E1-50          19  120    Sst Myh8_1            Sst   \n",
       "4  LS-15524_S21_E1-50          19  267    Sst Myh8_1            Sst   \n",
       "\n",
       "  class_label cluster_color  size      ar_id exp_component_vendor_name  ...  \\\n",
       "0   GABAergic       #FF7F2C   136  543183476               LS-15395-41  ...   \n",
       "1   GABAergic       #FF7F2C   136  543183474               LS-15395-42  ...   \n",
       "2   GABAergic       #FF7F2C   136  544930823               LS-15501-28  ...   \n",
       "3   GABAergic       #FF7F2C   136  544930707               LS-15501-81  ...   \n",
       "4   GABAergic       #FF7F2C   136  549999237               LS-15524-21  ...   \n",
       "\n",
       "  Injection_type        NewBatch   Vendor MultiPlex  \\\n",
       "0            0.0  HistoricalData  Covance   index96   \n",
       "1            0.0  HistoricalData  Covance   index96   \n",
       "2            0.0  HistoricalData  Covance   index96   \n",
       "3            0.0  HistoricalData  Covance   index96   \n",
       "4            0.0  HistoricalData  Covance   index96   \n",
       "\n",
       "   percent_reads_aligned_to_rrna_rmsk  percent_reads_aligned_to_trna_rmsk  \\\n",
       "0                                 NaN                                 NaN   \n",
       "1                                 NaN                                 NaN   \n",
       "2                                 NaN                                 NaN   \n",
       "3                                 NaN                                 NaN   \n",
       "4                                 NaN                                 NaN   \n",
       "\n",
       "   percent_reads_aligned_to_ncrna  percent_reads_aligned_to_introns  \\\n",
       "0                       15.162980                          9.855384   \n",
       "1                       16.281620                          6.763949   \n",
       "2                       17.066841                          3.311346   \n",
       "3                       13.866006                         11.769028   \n",
       "4                       15.912637                          5.923318   \n",
       "\n",
       "  percent_reads_aligned_to_intergenic percent_reads_aligned_unique  \n",
       "0                            3.264771                    84.993663  \n",
       "1                            2.856108                    87.252036  \n",
       "2                            2.437971                    86.831969  \n",
       "3                            3.699682                    85.252590  \n",
       "4                            2.385434                    82.503705  \n",
       "\n",
       "[5 rows x 129 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv(f_meta, names=['cell', 'cluster_id'], header=0) \n",
    "meta2 = pd.read_csv(f_meta2)\n",
    "meta3 = pd.read_csv(f_meta3, names=['QC', 'cell'], header=0)\n",
    "meta4 = pd.read_csv(f_meta4, index_col=0)\n",
    "meta4.index.name = 'cell' \n",
    "meta4 = meta4.reset_index()\n",
    "\n",
    "meta = pd.merge(meta, meta3, on='cell')\n",
    "meta = pd.merge(meta, meta2, on='cluster_id')\n",
    "meta = pd.merge(meta, meta4, on='cell')\n",
    "\n",
    "print(meta.shape)\n",
    "print(meta3.shape)\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(32324,) (32324,)\n",
      "(45768, 6300)\n"
     ]
    }
   ],
   "source": [
    "data_intron = pd.read_csv(f_data_intron, index_col=0) \n",
    "data_exon = pd.read_csv(f_data_exon, index_col=0) \n",
    "\n",
    "data = data_intron + data_exon\n",
    "gene = data.index.values\n",
    "cell = data.columns.values\n",
    "\n",
    "gene_table = df_genes_v2.reindex(gene).dropna()\n",
    "gene_ids = gene_table['gene_id'].values\n",
    "gene_names = gene_table.index.values\n",
    "\n",
    "print(gene_ids.shape, gene_names.shape)\n",
    "print(data.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_sparse = sparse.coo_matrix(data.loc[gene_names])\n",
    "gene = gene_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "gc_mat = GC_matrix(gene, cell, data_sparse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta.to_csv(fout_meta, sep=\"\\t\", header=True, index=False)\n",
    "snmcseq_utils.save_gc_matrix(gc_mat, \n",
    "                             fout_data_gene, fout_data_cell, fout_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Smarter-nuclei"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze/smarter_nuclei/cluster.membership.csv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/smarter_nuclei_metadata.tsv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/smarter_nuclei_raw.npz\n"
     ]
    }
   ],
   "source": [
    "dataset_name = 'smarter_nuclei'\n",
    "\n",
    "f_meta = os.path.join(SRC_DIR, dataset_name, 'cluster.membership.csv') \n",
    "f_meta2 = os.path.join(SRC_DIR, dataset_name, 'cluster.annotation.csv') \n",
    "f_meta3 = os.path.join(SRC_DIR, dataset_name, 'QC.csv') \n",
    "f_meta4 = os.path.join(SRC_DIR, dataset_name, 'sample_metadata.csv.gz') \n",
    "\n",
    "f_data_intron = os.path.join(SRC_DIR, dataset_name, 'intron.counts.csv.gz') \n",
    "f_data_exon = os.path.join(SRC_DIR, dataset_name, 'exon.counts.csv.gz') \n",
    "\n",
    "fout_meta = os.path.join(DST_DIR, dataset_name+'_metadata.tsv') \n",
    "fout_data = os.path.join(DST_DIR, dataset_name+'_raw.npz') \n",
    "fout_data_gene = os.path.join(DST_DIR, dataset_name+'_raw.gene') \n",
    "fout_data_cell = os.path.join(DST_DIR, dataset_name+'_raw.cell') \n",
    "\n",
    "print(f_meta)\n",
    "print(fout_meta)\n",
    "print(fout_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6171, 146)\n",
      "(6171, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cndd/fangming/venvs/conda_dobro/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2728: DtypeWarning: Columns (61) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>QC</th>\n",
       "      <th>cluster_label</th>\n",
       "      <th>subclass_label</th>\n",
       "      <th>class_label</th>\n",
       "      <th>cluster_color</th>\n",
       "      <th>size</th>\n",
       "      <th>ar_id</th>\n",
       "      <th>exp_component_vendor_name</th>\n",
       "      <th>...</th>\n",
       "      <th>pred.cl.2</th>\n",
       "      <th>pred.score.2</th>\n",
       "      <th>ref.cl.2</th>\n",
       "      <th>ref.cl.1.1</th>\n",
       "      <th>pred.cl.3</th>\n",
       "      <th>pred.score.3</th>\n",
       "      <th>cluster_label.2</th>\n",
       "      <th>category_label</th>\n",
       "      <th>Region.1</th>\n",
       "      <th>gene.counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SM-GE653_S113_E1-50</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>Lamp5 Egln3_2</td>\n",
       "      <td>Lamp5</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#E69B73</td>\n",
       "      <td>87</td>\n",
       "      <td>647869537</td>\n",
       "      <td>SM-GE653-S113</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>0.917036</td>\n",
       "      <td>10</td>\n",
       "      <td>31</td>\n",
       "      <td>14</td>\n",
       "      <td>0.917036</td>\n",
       "      <td>Vip Htr1f</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>7792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SM-GE653_S116_E1-50</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>Lamp5 Egln3_2</td>\n",
       "      <td>Lamp5</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#E69B73</td>\n",
       "      <td>87</td>\n",
       "      <td>647869531</td>\n",
       "      <td>SM-GE653-S116</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>0.712671</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>0.712671</td>\n",
       "      <td>Lamp5 Pdlim5</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>8502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SM-GE653_S117_E1-50</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>Lamp5 Egln3_2</td>\n",
       "      <td>Lamp5</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#E69B73</td>\n",
       "      <td>87</td>\n",
       "      <td>647869529</td>\n",
       "      <td>SM-GE653-S117</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>0.899050</td>\n",
       "      <td>9</td>\n",
       "      <td>28</td>\n",
       "      <td>12</td>\n",
       "      <td>0.899050</td>\n",
       "      <td>Vip Chat_3</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>7021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SM-GE653_S120_E1-50</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>Lamp5 Egln3_2</td>\n",
       "      <td>Lamp5</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#E69B73</td>\n",
       "      <td>87</td>\n",
       "      <td>647869523</td>\n",
       "      <td>SM-GE653-S120</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>0.892925</td>\n",
       "      <td>9</td>\n",
       "      <td>28</td>\n",
       "      <td>11</td>\n",
       "      <td>0.892925</td>\n",
       "      <td>Vip Chat_2</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>7085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SM-GE653_S122_E1-50</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>Lamp5 Egln3_2</td>\n",
       "      <td>Lamp5</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#E69B73</td>\n",
       "      <td>87</td>\n",
       "      <td>647869519</td>\n",
       "      <td>SM-GE653-S122</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>0.822251</td>\n",
       "      <td>9</td>\n",
       "      <td>28</td>\n",
       "      <td>13</td>\n",
       "      <td>0.822251</td>\n",
       "      <td>Vip Lmo1</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>7516</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 146 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  cell  cluster_id  QC  cluster_label subclass_label  \\\n",
       "0  SM-GE653_S113_E1-50           2   1  Lamp5 Egln3_2          Lamp5   \n",
       "1  SM-GE653_S116_E1-50           2   4  Lamp5 Egln3_2          Lamp5   \n",
       "2  SM-GE653_S117_E1-50           2   5  Lamp5 Egln3_2          Lamp5   \n",
       "3  SM-GE653_S120_E1-50           2   8  Lamp5 Egln3_2          Lamp5   \n",
       "4  SM-GE653_S122_E1-50           2  10  Lamp5 Egln3_2          Lamp5   \n",
       "\n",
       "  class_label cluster_color  size      ar_id exp_component_vendor_name  ...  \\\n",
       "0   GABAergic       #E69B73    87  647869537             SM-GE653-S113  ...   \n",
       "1   GABAergic       #E69B73    87  647869531             SM-GE653-S116  ...   \n",
       "2   GABAergic       #E69B73    87  647869529             SM-GE653-S117  ...   \n",
       "3   GABAergic       #E69B73    87  647869523             SM-GE653-S120  ...   \n",
       "4   GABAergic       #E69B73    87  647869519             SM-GE653-S122  ...   \n",
       "\n",
       "  pred.cl.2 pred.score.2 ref.cl.2 ref.cl.1.1  pred.cl.3  pred.score.3  \\\n",
       "0        14     0.917036       10         31         14      0.917036   \n",
       "1         3     0.712671        2          7          3      0.712671   \n",
       "2        12     0.899050        9         28         12      0.899050   \n",
       "3        11     0.892925        9         28         11      0.892925   \n",
       "4        13     0.822251        9         28         13      0.822251   \n",
       "\n",
       "   cluster_label.2  category_label Region.1 gene.counts  \n",
       "0        Vip Htr1f       GABAergic   Cortex        7792  \n",
       "1     Lamp5 Pdlim5       GABAergic   Cortex        8502  \n",
       "2       Vip Chat_3       GABAergic   Cortex        7021  \n",
       "3       Vip Chat_2       GABAergic   Cortex        7085  \n",
       "4         Vip Lmo1       GABAergic   Cortex        7516  \n",
       "\n",
       "[5 rows x 146 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv(f_meta, names=['cell', 'cluster_id'], header=0) \n",
    "meta2 = pd.read_csv(f_meta2)\n",
    "meta3 = pd.read_csv(f_meta3, names=['QC', 'cell'], header=0)\n",
    "\n",
    "meta4 = pd.read_csv(f_meta4, index_col=0)\n",
    "meta4.index.name = 'cell' \n",
    "meta4 = meta4.reset_index()\n",
    "\n",
    "meta = pd.merge(meta, meta3, on='cell')\n",
    "meta = pd.merge(meta, meta2, on='cluster_id')\n",
    "meta = pd.merge(meta, meta4.drop(columns=['cluster_id', 'cluster_label', 'cluster_color', 'size']), \n",
    "                on='cell')\n",
    "\n",
    "print(meta.shape)\n",
    "print(meta3.shape)\n",
    "# meta.filter(regex='_x$', axis=1).head()\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(32324,) (32324,)\n",
      "(45768, 6278)\n"
     ]
    }
   ],
   "source": [
    "data_intron = pd.read_csv(f_data_intron, index_col=0) \n",
    "data_exon = pd.read_csv(f_data_exon, index_col=0) \n",
    "\n",
    "data = data_intron + data_exon\n",
    "cell = data.columns.values\n",
    "gene = data.index.values\n",
    "\n",
    "gene_table = df_genes_v2.reindex(gene).dropna()\n",
    "gene_ids = gene_table['gene_id'].values\n",
    "gene_names = gene_table.index.values\n",
    "\n",
    "print(gene_ids.shape, gene_names.shape)\n",
    "print(data.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_sparse = sparse.coo_matrix(data.loc[gene_names])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "gc_mat = GC_matrix(gene_ids, cell, data_sparse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta.to_csv(fout_meta, sep=\"\\t\", header=True, index=False)\n",
    "snmcseq_utils.save_gc_matrix(gc_mat, \n",
    "                             fout_data_gene, fout_data_cell, fout_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### mC\n",
    "\n",
    "- QC'ed cells\n",
    "- remove version on ensembl id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "import xarray as xr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_supptable/snmcseq_gene_metadata.tsv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_supptable/snmcseq_gene_mCH_counts.npz\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_supptable/snmcseq_gene_CH_counts.npz\n"
     ]
    }
   ],
   "source": [
    "dataset_name = 'snmcseq_gene'\n",
    "\n",
    "fout_meta = os.path.join(DST_DIR, dataset_name+'_metadata.tsv') \n",
    "fout_data_mc = os.path.join(DST_DIR, dataset_name+'_mCH_counts.npz') \n",
    "fout_data_c = os.path.join(DST_DIR, dataset_name+'_CH_counts.npz') \n",
    "fout_data_gene = os.path.join(DST_DIR, dataset_name+'_counts.gene') \n",
    "fout_data_cell = os.path.join(DST_DIR, dataset_name+'_counts.cell') \n",
    "\n",
    "print(fout_meta)\n",
    "print(fout_data_mc)\n",
    "print(fout_data_c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9876, 32)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>FullCellID</th>\n",
       "      <th>CCC_Rate</th>\n",
       "      <th>CG_Rate</th>\n",
       "      <th>CG_RateAdj</th>\n",
       "      <th>CH_Rate</th>\n",
       "      <th>CH_RateAdj</th>\n",
       "      <th>FinalReads</th>\n",
       "      <th>InputReads</th>\n",
       "      <th>MappedReads</th>\n",
       "      <th>...</th>\n",
       "      <th>Slice</th>\n",
       "      <th>PassFilter</th>\n",
       "      <th>pca_0</th>\n",
       "      <th>pca_1</th>\n",
       "      <th>umap_0</th>\n",
       "      <th>umap_1</th>\n",
       "      <th>tsne_0</th>\n",
       "      <th>tsne_1</th>\n",
       "      <th>MajorCluster</th>\n",
       "      <th>SubCluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2C_M_0</td>\n",
       "      <td>180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...</td>\n",
       "      <td>0.00511</td>\n",
       "      <td>0.77073</td>\n",
       "      <td>0.76955</td>\n",
       "      <td>0.02154</td>\n",
       "      <td>0.01651</td>\n",
       "      <td>1076259.0</td>\n",
       "      <td>2314862</td>\n",
       "      <td>1517338.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>-11.588627</td>\n",
       "      <td>-0.005045</td>\n",
       "      <td>10.051389</td>\n",
       "      <td>7.099836</td>\n",
       "      <td>-3.343437</td>\n",
       "      <td>18.330174</td>\n",
       "      <td>L4-IT-Rorb</td>\n",
       "      <td>L4-IT-Rorb_Rorb-Tenm2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2C_M_1</td>\n",
       "      <td>180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...</td>\n",
       "      <td>0.00531</td>\n",
       "      <td>0.75683</td>\n",
       "      <td>0.75553</td>\n",
       "      <td>0.02196</td>\n",
       "      <td>0.01674</td>\n",
       "      <td>1197092.0</td>\n",
       "      <td>2590626</td>\n",
       "      <td>1709334.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>-19.267288</td>\n",
       "      <td>-7.374098</td>\n",
       "      <td>13.717241</td>\n",
       "      <td>-1.786627</td>\n",
       "      <td>9.212860</td>\n",
       "      <td>-26.514785</td>\n",
       "      <td>L23-IT-Cux2</td>\n",
       "      <td>L23-IT-Cux2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2C_M_10</td>\n",
       "      <td>180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...</td>\n",
       "      <td>0.00581</td>\n",
       "      <td>0.75720</td>\n",
       "      <td>0.75578</td>\n",
       "      <td>0.02603</td>\n",
       "      <td>0.02034</td>\n",
       "      <td>1006630.0</td>\n",
       "      <td>2348484</td>\n",
       "      <td>1606410.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>-15.340144</td>\n",
       "      <td>-2.739380</td>\n",
       "      <td>4.829485</td>\n",
       "      <td>9.943986</td>\n",
       "      <td>-22.960995</td>\n",
       "      <td>13.509483</td>\n",
       "      <td>Outlier</td>\n",
       "      <td>Outlier</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2C_M_100</td>\n",
       "      <td>180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...</td>\n",
       "      <td>0.00634</td>\n",
       "      <td>0.74984</td>\n",
       "      <td>0.74824</td>\n",
       "      <td>0.02718</td>\n",
       "      <td>0.02097</td>\n",
       "      <td>2026078.0</td>\n",
       "      <td>4599922</td>\n",
       "      <td>3154619.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>-23.625140</td>\n",
       "      <td>-5.929033</td>\n",
       "      <td>13.739521</td>\n",
       "      <td>1.107174</td>\n",
       "      <td>12.798545</td>\n",
       "      <td>-17.126331</td>\n",
       "      <td>L23-IT-Cux2</td>\n",
       "      <td>L23-IT-Cux2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2C_M_1000</td>\n",
       "      <td>180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_3_CE...</td>\n",
       "      <td>0.00521</td>\n",
       "      <td>0.77073</td>\n",
       "      <td>0.76953</td>\n",
       "      <td>0.02117</td>\n",
       "      <td>0.01604</td>\n",
       "      <td>2368464.0</td>\n",
       "      <td>5320502</td>\n",
       "      <td>3648786.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>-21.815190</td>\n",
       "      <td>-7.130878</td>\n",
       "      <td>14.154541</td>\n",
       "      <td>-2.284118</td>\n",
       "      <td>8.443658</td>\n",
       "      <td>-27.943267</td>\n",
       "      <td>L23-IT-Cux2</td>\n",
       "      <td>L23-IT-Cux2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        cell                                         FullCellID  CCC_Rate  \\\n",
       "0     2C_M_0  180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...   0.00511   \n",
       "1     2C_M_1  180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...   0.00531   \n",
       "2    2C_M_10  180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...   0.00581   \n",
       "3   2C_M_100  180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...   0.00634   \n",
       "4  2C_M_1000  180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_3_CE...   0.00521   \n",
       "\n",
       "   CG_Rate  CG_RateAdj  CH_Rate  CH_RateAdj  FinalReads  InputReads  \\\n",
       "0  0.77073     0.76955  0.02154     0.01651   1076259.0     2314862   \n",
       "1  0.75683     0.75553  0.02196     0.01674   1197092.0     2590626   \n",
       "2  0.75720     0.75578  0.02603     0.02034   1006630.0     2348484   \n",
       "3  0.74984     0.74824  0.02718     0.02097   2026078.0     4599922   \n",
       "4  0.77073     0.76953  0.02117     0.01604   2368464.0     5320502   \n",
       "\n",
       "   MappedReads  ... Slice PassFilter      pca_0     pca_1     umap_0  \\\n",
       "0    1517338.0  ...     2       True -11.588627 -0.005045  10.051389   \n",
       "1    1709334.0  ...     2       True -19.267288 -7.374098  13.717241   \n",
       "2    1606410.0  ...     2       True -15.340144 -2.739380   4.829485   \n",
       "3    3154619.0  ...     2       True -23.625140 -5.929033  13.739521   \n",
       "4    3648786.0  ...     2       True -21.815190 -7.130878  14.154541   \n",
       "\n",
       "     umap_1     tsne_0     tsne_1  MajorCluster             SubCluster  \n",
       "0  7.099836  -3.343437  18.330174    L4-IT-Rorb  L4-IT-Rorb_Rorb-Tenm2  \n",
       "1 -1.786627   9.212860 -26.514785   L23-IT-Cux2            L23-IT-Cux2  \n",
       "2  9.943986 -22.960995  13.509483       Outlier                Outlier  \n",
       "3  1.107174  12.798545 -17.126331   L23-IT-Cux2            L23-IT-Cux2  \n",
       "4 -2.284118   8.443658 -27.943267   L23-IT-Cux2            L23-IT-Cux2  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# metadata after QC and clustering\n",
    "# meta = pd.read_msgpack(os.path.join(SRC_DIR, 'snmc_and_snatac/study/mc/Metadata.after_QC.msg'))\n",
    "# meta2 = pd.read_msgpack(os.path.join(SRC_DIR, 'snmc_and_snatac/study/mc/mop_mc_clustering/MOp.methylome_only.analysis_results.msg')).drop(columns=['cell'])\n",
    "# print(meta.shape)\n",
    "# print(meta2.shape)\n",
    "# print(meta.columns.values.shape)\n",
    "# print(np.intersect1d(meta.columns.values, meta2.columns.values).shape)\n",
    "# meta2 = meta2.drop(columns=meta2.filter(regex='^ENSMUSG', axis=1).columns).drop(columns='AllcPath')\n",
    "\n",
    "meta3 = pd.read_csv(os.path.join(SRC_DIR, 'snmc_and_snatac/study/mc/MOp.snmC-seq.AnalysisResult.csv.gz')).rename(columns={'index': 'cell'})\n",
    "print(meta3.shape)\n",
    "meta3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta3['index_name'].head()\n",
    "drop_cols = [\n",
    "    'index_name', 'uid', 'BamFilteringRate', 'MappingRate', 'Pos96',\n",
    "    'Plate', 'Col96', 'Row96', 'Col384', 'Row384', 'FACS_Date', 'Slice', \n",
    "    'pca_0', 'pca_1',\n",
    "    'Region',\n",
    "]\n",
    "meta3 = meta3.drop(drop_cols, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>FullCellID</th>\n",
       "      <th>CCC_Rate</th>\n",
       "      <th>CG_Rate</th>\n",
       "      <th>CG_RateAdj</th>\n",
       "      <th>CH_Rate</th>\n",
       "      <th>CH_RateAdj</th>\n",
       "      <th>FinalReads</th>\n",
       "      <th>InputReads</th>\n",
       "      <th>MappedReads</th>\n",
       "      <th>PassFilter</th>\n",
       "      <th>umap_0</th>\n",
       "      <th>umap_1</th>\n",
       "      <th>tsne_0</th>\n",
       "      <th>tsne_1</th>\n",
       "      <th>MajorCluster</th>\n",
       "      <th>SubCluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2C_M_0</td>\n",
       "      <td>180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...</td>\n",
       "      <td>0.00511</td>\n",
       "      <td>0.77073</td>\n",
       "      <td>0.76955</td>\n",
       "      <td>0.02154</td>\n",
       "      <td>0.01651</td>\n",
       "      <td>1076259.0</td>\n",
       "      <td>2314862</td>\n",
       "      <td>1517338.0</td>\n",
       "      <td>True</td>\n",
       "      <td>10.051389</td>\n",
       "      <td>7.099836</td>\n",
       "      <td>-3.343437</td>\n",
       "      <td>18.330174</td>\n",
       "      <td>L4-IT-Rorb</td>\n",
       "      <td>L4-IT-Rorb_Rorb-Tenm2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2C_M_1</td>\n",
       "      <td>180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...</td>\n",
       "      <td>0.00531</td>\n",
       "      <td>0.75683</td>\n",
       "      <td>0.75553</td>\n",
       "      <td>0.02196</td>\n",
       "      <td>0.01674</td>\n",
       "      <td>1197092.0</td>\n",
       "      <td>2590626</td>\n",
       "      <td>1709334.0</td>\n",
       "      <td>True</td>\n",
       "      <td>13.717241</td>\n",
       "      <td>-1.786627</td>\n",
       "      <td>9.212860</td>\n",
       "      <td>-26.514785</td>\n",
       "      <td>L23-IT-Cux2</td>\n",
       "      <td>L23-IT-Cux2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2C_M_10</td>\n",
       "      <td>180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...</td>\n",
       "      <td>0.00581</td>\n",
       "      <td>0.75720</td>\n",
       "      <td>0.75578</td>\n",
       "      <td>0.02603</td>\n",
       "      <td>0.02034</td>\n",
       "      <td>1006630.0</td>\n",
       "      <td>2348484</td>\n",
       "      <td>1606410.0</td>\n",
       "      <td>True</td>\n",
       "      <td>4.829485</td>\n",
       "      <td>9.943986</td>\n",
       "      <td>-22.960995</td>\n",
       "      <td>13.509483</td>\n",
       "      <td>Outlier</td>\n",
       "      <td>Outlier</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2C_M_100</td>\n",
       "      <td>180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...</td>\n",
       "      <td>0.00634</td>\n",
       "      <td>0.74984</td>\n",
       "      <td>0.74824</td>\n",
       "      <td>0.02718</td>\n",
       "      <td>0.02097</td>\n",
       "      <td>2026078.0</td>\n",
       "      <td>4599922</td>\n",
       "      <td>3154619.0</td>\n",
       "      <td>True</td>\n",
       "      <td>13.739521</td>\n",
       "      <td>1.107174</td>\n",
       "      <td>12.798545</td>\n",
       "      <td>-17.126331</td>\n",
       "      <td>L23-IT-Cux2</td>\n",
       "      <td>L23-IT-Cux2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2C_M_1000</td>\n",
       "      <td>180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_3_CE...</td>\n",
       "      <td>0.00521</td>\n",
       "      <td>0.77073</td>\n",
       "      <td>0.76953</td>\n",
       "      <td>0.02117</td>\n",
       "      <td>0.01604</td>\n",
       "      <td>2368464.0</td>\n",
       "      <td>5320502</td>\n",
       "      <td>3648786.0</td>\n",
       "      <td>True</td>\n",
       "      <td>14.154541</td>\n",
       "      <td>-2.284118</td>\n",
       "      <td>8.443658</td>\n",
       "      <td>-27.943267</td>\n",
       "      <td>L23-IT-Cux2</td>\n",
       "      <td>L23-IT-Cux2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        cell                                         FullCellID  CCC_Rate  \\\n",
       "0     2C_M_0  180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...   0.00511   \n",
       "1     2C_M_1  180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...   0.00531   \n",
       "2    2C_M_10  180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...   0.00581   \n",
       "3   2C_M_100  180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...   0.00634   \n",
       "4  2C_M_1000  180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_3_CE...   0.00521   \n",
       "\n",
       "   CG_Rate  CG_RateAdj  CH_Rate  CH_RateAdj  FinalReads  InputReads  \\\n",
       "0  0.77073     0.76955  0.02154     0.01651   1076259.0     2314862   \n",
       "1  0.75683     0.75553  0.02196     0.01674   1197092.0     2590626   \n",
       "2  0.75720     0.75578  0.02603     0.02034   1006630.0     2348484   \n",
       "3  0.74984     0.74824  0.02718     0.02097   2026078.0     4599922   \n",
       "4  0.77073     0.76953  0.02117     0.01604   2368464.0     5320502   \n",
       "\n",
       "   MappedReads  PassFilter     umap_0    umap_1     tsne_0     tsne_1  \\\n",
       "0    1517338.0        True  10.051389  7.099836  -3.343437  18.330174   \n",
       "1    1709334.0        True  13.717241 -1.786627   9.212860 -26.514785   \n",
       "2    1606410.0        True   4.829485  9.943986 -22.960995  13.509483   \n",
       "3    3154619.0        True  13.739521  1.107174  12.798545 -17.126331   \n",
       "4    3648786.0        True  14.154541 -2.284118   8.443658 -27.943267   \n",
       "\n",
       "  MajorCluster             SubCluster  \n",
       "0   L4-IT-Rorb  L4-IT-Rorb_Rorb-Tenm2  \n",
       "1  L23-IT-Cux2            L23-IT-Cux2  \n",
       "2      Outlier                Outlier  \n",
       "3  L23-IT-Cux2            L23-IT-Cux2  \n",
       "4  L23-IT-Cux2            L23-IT-Cux2  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['cell', 'FullCellID', 'CCC_Rate', 'CG_Rate', 'CG_RateAdj', 'CH_Rate',\n",
      "       'CH_RateAdj', 'FinalReads', 'InputReads', 'MappedReads', 'Region',\n",
      "       'index_name', 'uid', 'BamFilteringRate', 'MappingRate', 'Pos96',\n",
      "       'Plate', 'Col96', 'Row96', 'Col384', 'Row384', 'FACS_Date', 'Slice',\n",
      "       'PassFilter', 'pca_0', 'pca_1', 'umap_0', 'umap_1', 'tsne_0', 'tsne_1',\n",
      "       'MajorCluster', 'SubCluster'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(meta3.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta3.to_csv(fout_meta, sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "reps = ['2C-180409.mcds', \n",
    "        '2C-180410.mcds',\n",
    "        '3C-171206.mcds',\n",
    "        '3C-171207.mcds',\n",
    "        '4B-171212.mcds',\n",
    "        '4B-171213.mcds',\n",
    "        '4B-180104.mcds',\n",
    "        '5D-180605.mcds',\n",
    "        '5D-180612.mcds',\n",
    "       ]\n",
    "\n",
    "c_tables = []\n",
    "mc_tables = []\n",
    "for rep in reps:\n",
    "    mcds = xr.open_dataset(os.path.join(SRC_DIR, 'snmc_and_snatac/dataset/mc/{}'.format(rep)))\n",
    "    c_table = mcds['gene_da'].sel(mc_type='CHN', count_type='cov').squeeze().to_pandas()\n",
    "    mc_table = mcds['gene_da'].sel(mc_type='CHN', count_type='mc').squeeze().to_pandas()\n",
    "    c_tables.append(c_table)\n",
    "    mc_tables.append(mc_table)\n",
    "    \n",
    "c_tables = pd.concat(c_tables, axis=0)\n",
    "mc_tables = pd.concat(mc_tables, axis=0)\n",
    "assert np.all(c_tables.columns.values == mc_tables.columns.values)\n",
    "assert np.all(c_tables.index.values == mc_tables.columns.values)\n",
    "print(c_tables.shape, mc_tables.shape)\n",
    "    \n",
    "gene = c_tables.columns.values\n",
    "cell = c_tables.index.values \n",
    "c_table = c_tables.T\n",
    "c_table.index = gene\n",
    "\n",
    "mc_table = mc_tables.T\n",
    "mc_table.index = gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove version on ensembl id\n",
    "c_table = sparse.coo_matrix(c_table.values)\n",
    "mc_table = sparse.coo_matrix(mc_table.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['2C_M_10', '2C_M_0', '2C_M_1', ..., '5D_M_2370', '5D_M_2348',\n",
       "       '5D_M_2367'], dtype=object)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell = c_tables.index.values\n",
    "cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save mc and c tables\n",
    "\n",
    "data = {'mc': mc_table, \n",
    "        'c': c_table, \n",
    "       }\n",
    "gc_mat = GC_matrix(gene, cell, data) \n",
    "\n",
    "snmcseq_utils.save_gc_matrix_methylation(gc_mat, \n",
    "                                         fout_data_gene, fout_data_cell, \n",
    "                                         fout_data_mc, fout_data_c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ATAC \n",
    "- rename barcode and header\n",
    "- read snap files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_supptable/snatac_gene_metadata.tsv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_supptable/snatac_gene_raw.npz\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_supptable/snatac_gene_raw.gene\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_supptable/snatac_gene_raw.cell\n"
     ]
    }
   ],
   "source": [
    "dataset_name = 'snatac_gene'\n",
    "\n",
    "fout_meta = os.path.join(DST_DIR, dataset_name+'_metadata.tsv') \n",
    "fout_data = os.path.join(DST_DIR, dataset_name+'_raw.npz') \n",
    "fout_data_gene = os.path.join(DST_DIR, dataset_name+'_raw.gene') \n",
    "fout_data_cell = os.path.join(DST_DIR, dataset_name+'_raw.cell') \n",
    "\n",
    "print(fout_meta)\n",
    "print(fout_data)\n",
    "print(fout_data_gene)\n",
    "print(fout_data_cell)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "rename_atac = {\n",
    "    'TN': 'Num fragments',\n",
    "    'UM': 'Num uniquely mapped',\n",
    "    'PP': 'Num proper paired',\n",
    "    'UQ': \"Num unique fragments\",\n",
    "    'CM': \"Num chrM fragments\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cluster\n",
      "ASC                7215\n",
      "Chodl                86\n",
      "Endo               1245\n",
      "L23.a              6085\n",
      "L23.b              5758\n",
      "L23.c              3590\n",
      "L4                 7307\n",
      "L5.IT.a            3668\n",
      "L5.IT.b            4214\n",
      "L5.PT              1755\n",
      "L6.CT              8008\n",
      "L6.IT              4233\n",
      "Lamp5_Arhgdib       237\n",
      "Lamp5_Mettl21e      171\n",
      "Lamp5_Ndnf          204\n",
      "Lamp5_Smad3         940\n",
      "MGC                3899\n",
      "NP                 1291\n",
      "OGC                8557\n",
      "OPC                3279\n",
      "Other              1571\n",
      "Pv_Ntf3_Trim63     1105\n",
      "Pv_Tac1            1944\n",
      "Pv_Vsig2            431\n",
      "Smc                 586\n",
      "Sncg                246\n",
      "Sst_Chrna2_Myh8     538\n",
      "Sst_Man1a          1355\n",
      "Sst_Stk33           246\n",
      "Vip_Chat            500\n",
      "Vip_Gcnt4           292\n",
      "Vip_Hcls1           395\n",
      "Vip_Lipg            245\n",
      "dtype: int64\n",
      "(81196, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>sample</th>\n",
       "      <th>barcode</th>\n",
       "      <th>Num fragments</th>\n",
       "      <th>Num uniquely mapped</th>\n",
       "      <th>Num proper paired</th>\n",
       "      <th>Num unique fragments</th>\n",
       "      <th>Num chrM fragments</th>\n",
       "      <th>cluster</th>\n",
       "      <th>tsne1</th>\n",
       "      <th>tsne2</th>\n",
       "      <th>umap-1</th>\n",
       "      <th>umap-2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CEMBA171206_3C_AGCGATAGAACCAGGTAAGAGATGTATAGCCT</td>\n",
       "      <td>CEMBA171206_3C</td>\n",
       "      <td>AGCGATAGAACCAGGTAAGAGATGTATAGCCT</td>\n",
       "      <td>138330</td>\n",
       "      <td>128263</td>\n",
       "      <td>128039</td>\n",
       "      <td>64868</td>\n",
       "      <td>0</td>\n",
       "      <td>L6.CT</td>\n",
       "      <td>7.605117</td>\n",
       "      <td>0.697672</td>\n",
       "      <td>-2.125204</td>\n",
       "      <td>-4.396653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CEMBA171206_3C_AGCGATAGAACCAGGTAATGACGTCAGGACGT</td>\n",
       "      <td>CEMBA171206_3C</td>\n",
       "      <td>AGCGATAGAACCAGGTAATGACGTCAGGACGT</td>\n",
       "      <td>159133</td>\n",
       "      <td>146663</td>\n",
       "      <td>146348</td>\n",
       "      <td>69167</td>\n",
       "      <td>0</td>\n",
       "      <td>L23.b</td>\n",
       "      <td>-9.492925</td>\n",
       "      <td>-6.200938</td>\n",
       "      <td>8.959074</td>\n",
       "      <td>3.006819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CEMBA171206_3C_AGCGATAGAACCAGGTAGGATAACATAGAGGC</td>\n",
       "      <td>CEMBA171206_3C</td>\n",
       "      <td>AGCGATAGAACCAGGTAGGATAACATAGAGGC</td>\n",
       "      <td>19152</td>\n",
       "      <td>17586</td>\n",
       "      <td>17545</td>\n",
       "      <td>8348</td>\n",
       "      <td>0</td>\n",
       "      <td>ASC</td>\n",
       "      <td>14.227358</td>\n",
       "      <td>-11.023748</td>\n",
       "      <td>-6.880754</td>\n",
       "      <td>-2.903271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CEMBA171206_3C_AGCGATAGAACCAGGTATAGCCTTAGGCGAAG</td>\n",
       "      <td>CEMBA171206_3C</td>\n",
       "      <td>AGCGATAGAACCAGGTATAGCCTTAGGCGAAG</td>\n",
       "      <td>14876</td>\n",
       "      <td>13837</td>\n",
       "      <td>13801</td>\n",
       "      <td>6583</td>\n",
       "      <td>0</td>\n",
       "      <td>L6.CT</td>\n",
       "      <td>5.144768</td>\n",
       "      <td>4.790674</td>\n",
       "      <td>-0.955707</td>\n",
       "      <td>-4.635163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CEMBA171206_3C_AGCGATAGAACCAGGTATAGCCTTATAGAGGC</td>\n",
       "      <td>CEMBA171206_3C</td>\n",
       "      <td>AGCGATAGAACCAGGTATAGCCTTATAGAGGC</td>\n",
       "      <td>28532</td>\n",
       "      <td>26303</td>\n",
       "      <td>26202</td>\n",
       "      <td>12437</td>\n",
       "      <td>0</td>\n",
       "      <td>L5.IT.b</td>\n",
       "      <td>5.218895</td>\n",
       "      <td>-8.674734</td>\n",
       "      <td>0.796851</td>\n",
       "      <td>4.447933</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              cell          sample  \\\n",
       "0  CEMBA171206_3C_AGCGATAGAACCAGGTAAGAGATGTATAGCCT  CEMBA171206_3C   \n",
       "1  CEMBA171206_3C_AGCGATAGAACCAGGTAATGACGTCAGGACGT  CEMBA171206_3C   \n",
       "2  CEMBA171206_3C_AGCGATAGAACCAGGTAGGATAACATAGAGGC  CEMBA171206_3C   \n",
       "3  CEMBA171206_3C_AGCGATAGAACCAGGTATAGCCTTAGGCGAAG  CEMBA171206_3C   \n",
       "4  CEMBA171206_3C_AGCGATAGAACCAGGTATAGCCTTATAGAGGC  CEMBA171206_3C   \n",
       "\n",
       "                            barcode  Num fragments  Num uniquely mapped  \\\n",
       "0  AGCGATAGAACCAGGTAAGAGATGTATAGCCT         138330               128263   \n",
       "1  AGCGATAGAACCAGGTAATGACGTCAGGACGT         159133               146663   \n",
       "2  AGCGATAGAACCAGGTAGGATAACATAGAGGC          19152                17586   \n",
       "3  AGCGATAGAACCAGGTATAGCCTTAGGCGAAG          14876                13837   \n",
       "4  AGCGATAGAACCAGGTATAGCCTTATAGAGGC          28532                26303   \n",
       "\n",
       "   Num proper paired  Num unique fragments  Num chrM fragments  cluster  \\\n",
       "0             128039                 64868                   0    L6.CT   \n",
       "1             146348                 69167                   0    L23.b   \n",
       "2              17545                  8348                   0      ASC   \n",
       "3              13801                  6583                   0    L6.CT   \n",
       "4              26202                 12437                   0  L5.IT.b   \n",
       "\n",
       "       tsne1      tsne2    umap-1    umap-2  \n",
       "0   7.605117   0.697672 -2.125204 -4.396653  \n",
       "1  -9.492925  -6.200938  8.959074  3.006819  \n",
       "2  14.227358 -11.023748 -6.880754 -2.903271  \n",
       "3   5.144768   4.790674 -0.955707 -4.635163  \n",
       "4   5.218895  -8.674734  0.796851  4.447933  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = ('/cndd/Public_Datasets/CEMBA/BICCN_minibrain_data/data_freeze/snmc_and_snatac/study/atac/'\n",
    "     'yang_analysis_20190912/CEMBA_MOp.L2.cluster.meta.txt'\n",
    "    )\n",
    "meta_new = pd.read_csv(f, sep='\\t').rename(columns={'x.sp.joint@sample': 'sample', \n",
    "                                                    'x.sp.joint@cluster': 'cluster', \n",
    "                                                   })\n",
    "meta_cols = meta_new.columns.values\n",
    "meta_new['cell'] = meta_new['sample'] + '_' + meta_new['barcode']\n",
    "meta_new = meta_new[np.hstack([['cell'], meta_cols])]\n",
    "meta_new= meta_new.rename(columns=rename_atac)\n",
    "print(meta_new.groupby('cluster').size())\n",
    "print(meta_new.shape)\n",
    "meta_new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# meta = pd.read_msgpack(os.path.join(SRC_DIR, 'snmc_and_snatac/study/atac/ATAC.cell_tidy_data.msg'))\n",
    "# meta_cols = meta.columns.values\n",
    "# print(meta.shape)\n",
    "# meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# meta = meta.rename(columns={'barcode': 'sample', 'sample': 'barcode'})\n",
    "# meta['cell'] = meta['sample'] + '_' + meta['barcode']\n",
    "# meta = meta[np.hstack([['cell'], meta_cols])]\n",
    "# meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import h5py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CEMBA180409_2C\n",
      "CEMBA180410_2C\n",
      "CEMBA171206_3C\n",
      "CEMBA171207_3C\n",
      "CEMBA171212_4B\n",
      "CEMBA171213_4B\n",
      "CEMBA180104_4B\n",
      "CEMBA180612_5D\n",
      "CEMBA180618_5D\n"
     ]
    }
   ],
   "source": [
    "reps = ['CEMBA180409_2C', \n",
    "        'CEMBA180410_2C',\n",
    "        'CEMBA171206_3C',\n",
    "        'CEMBA171207_3C',\n",
    "        'CEMBA171212_4B',\n",
    "        'CEMBA171213_4B',\n",
    "        'CEMBA180104_4B',\n",
    "        'CEMBA180612_5D',\n",
    "        # 'CEMBA180605_5D',\n",
    "        'CEMBA180618_5D',\n",
    "       ]\n",
    "\n",
    "\n",
    "genes_all = []\n",
    "cells_all = []\n",
    "mat_all = []\n",
    "for rep in reps:\n",
    "    print(rep)\n",
    "    f = h5py.File(os.path.join(SRC_DIR, 'snmc_and_snatac/dataset/atac', rep+'.snap'), 'r')\n",
    "    # get genes cells mat for one replicate\n",
    "    genes = f['GM']['name'][:].astype(str)\n",
    "    cells = np.array([(rep+'_')+barcode for barcode in f['BD']['name'][:].astype(str)])\n",
    "    idx = f['GM']['idx'][:] - 1\n",
    "    idy = f['GM']['idy'][:] - 1\n",
    "    count = f['GM']['count']\n",
    "    mat = sparse.coo_matrix((count, (idx, idy)), shape=(len(cells), len(genes))).T\n",
    "    \n",
    "    # \n",
    "    genes_all.append(genes)\n",
    "    cells_all.append(cells)\n",
    "    mat_all.append(mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CEMBA180409_2C\n",
      "['AM', 'BD', 'FM', 'GM', 'HD', 'PM']\n"
     ]
    }
   ],
   "source": [
    "reps = ['CEMBA180409_2C', \n",
    "        'CEMBA180410_2C',\n",
    "        'CEMBA171206_3C',\n",
    "        'CEMBA171207_3C',\n",
    "        'CEMBA171212_4B',\n",
    "        'CEMBA171213_4B',\n",
    "        'CEMBA180104_4B',\n",
    "        'CEMBA180612_5D',\n",
    "        # 'CEMBA180605_5D',\n",
    "        'CEMBA180618_5D',\n",
    "       ]\n",
    "\n",
    "\n",
    "\n",
    "genes_all = []\n",
    "cells_all = []\n",
    "mat_all = []\n",
    "for rep in reps:\n",
    "    print(rep)\n",
    "    f = h5py.File(os.path.join(SRC_DIR, 'snmc_and_snatac/dataset/atac', rep+'.snap'), 'r')\n",
    "    print(list(f.keys()))\n",
    "#     # get genes cells mat for one replicate\n",
    "#     genes = f['GM']['name'][:].astype(str)\n",
    "#     cells = np.array([(rep+'_')+barcode for barcode in f['BD']['name'][:].astype(str)])\n",
    "#     idx = f['GM']['idx'][:] - 1\n",
    "#     idy = f['GM']['idy'][:] - 1\n",
    "#     count = f['GM']['count']\n",
    "#     mat = sparse.coo_matrix((count, (idx, idy)), shape=(len(cells), len(genes))).T\n",
    "    \n",
    "#     # \n",
    "#     genes_all.append(genes)\n",
    "#     cells_all.append(cells)\n",
    "#     mat_all.append(mat)\n",
    "    \n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(53278,) (135665,) (53278, 135665)\n"
     ]
    }
   ],
   "source": [
    "for i, genes in enumerate(genes_all):\n",
    "    if i == 0:\n",
    "        ref = genes\n",
    "    else:\n",
    "        assert np.all(genes == ref)\n",
    "\n",
    "genes_merge = genes\n",
    "cells_merge = np.hstack(cells_all)\n",
    "mat_merge = sparse.hstack(mat_all)\n",
    "\n",
    "print(genes_merge.shape, cells_merge.shape, mat_merge.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(53278,)\n",
      "(53278,)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['ENSMUSG00000037552', 'ENSMUSG00000052609', 'ENSMUSG00000040624',\n",
       "       ..., 'ENSMUSG00000110345', 'ENSMUSG00000110681',\n",
       "       'ENSMUSG00000110504'], dtype=object)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_table = df_genes_v2.reindex(genes_merge).dropna()\n",
    "assert len(gene_table) == len(genes_merge) #.shape\n",
    "genes_merge_id = gene_table['gene_id'].values\n",
    "\n",
    "print(genes.shape)\n",
    "print(genes_merge_id.shape)\n",
    "genes_merge_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "gc_mat = GC_matrix(genes_merge_id, cells_merge, mat_merge)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "snmcseq_utils.save_gc_matrix(gc_mat, \n",
    "                             fout_data_gene, fout_data_cell, fout_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_new.to_csv(fout_meta, sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tasic 2018 ALM "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>ENSMUSG00000109644</td>\n",
       "      <td>chr7</td>\n",
       "      <td>45567794</td>\n",
       "      <td>45575327</td>\n",
       "      <td>-</td>\n",
       "      <td>antisense_RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <td>ENSMUSG00000108652</td>\n",
       "      <td>chr7</td>\n",
       "      <td>74818818</td>\n",
       "      <td>74853813</td>\n",
       "      <td>-</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>ENSMUSG00000007777</td>\n",
       "      <td>chr11</td>\n",
       "      <td>51685386</td>\n",
       "      <td>51688874</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>ENSMUSG00000086714</td>\n",
       "      <td>chr2</td>\n",
       "      <td>26445696</td>\n",
       "      <td>26459390</td>\n",
       "      <td>+</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>ENSMUSG00000043644</td>\n",
       "      <td>chr11</td>\n",
       "      <td>120348678</td>\n",
       "      <td>120351190</td>\n",
       "      <td>+</td>\n",
       "      <td>bidirectional_promoter_lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          gene_id    chr      start        end strand  \\\n",
       "gene_name                                                               \n",
       "0610005C13Rik  ENSMUSG00000109644   chr7   45567794   45575327      -   \n",
       "0610006L08Rik  ENSMUSG00000108652   chr7   74818818   74853813      -   \n",
       "0610009B22Rik  ENSMUSG00000007777  chr11   51685386   51688874      -   \n",
       "0610009E02Rik  ENSMUSG00000086714   chr2   26445696   26459390      +   \n",
       "0610009L18Rik  ENSMUSG00000043644  chr11  120348678  120351190      +   \n",
       "\n",
       "                                   gene_type  \n",
       "gene_name                                     \n",
       "0610005C13Rik                  antisense_RNA  \n",
       "0610006L08Rik                        lincRNA  \n",
       "0610009B22Rik                 protein_coding  \n",
       "0610009E02Rik           processed_transcript  \n",
       "0610009L18Rik  bidirectional_promoter_lncRNA  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# genes_table = pd.read_csv(PATH_GENEBODY_ANNOTATION, sep='\\t')\n",
    "# print(genes_table.shape)\n",
    "# genes_table.head()\n",
    "df_genes_v2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(45768, 5)\n",
      "(10068, 50)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>organism</th>\n",
       "      <th>donor</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_days</th>\n",
       "      <th>eye_condition</th>\n",
       "      <th>genotype</th>\n",
       "      <th>driver_lines</th>\n",
       "      <th>...</th>\n",
       "      <th>genes_detected_cpm_criterion</th>\n",
       "      <th>genes_detected_fpkm_criterion</th>\n",
       "      <th>tdt_cpm</th>\n",
       "      <th>gfp_cpm</th>\n",
       "      <th>class</th>\n",
       "      <th>subclass</th>\n",
       "      <th>cluster</th>\n",
       "      <th>confusion_score</th>\n",
       "      <th>cluster_correlation</th>\n",
       "      <th>core_intermediate_call</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>F1S4_161216_001_A01</td>\n",
       "      <td>571878193</td>\n",
       "      <td>Cells</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>286575</td>\n",
       "      <td>M</td>\n",
       "      <td>57</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Ai14(RCL-tdT)/wt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>12965</td>\n",
       "      <td>12128</td>\n",
       "      <td>127.17</td>\n",
       "      <td>127.17</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>L5 PT</td>\n",
       "      <td>L5 PT ALM Slco2a1</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.845321</td>\n",
       "      <td>Core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>F1S4_180124_314_A01</td>\n",
       "      <td>665943217</td>\n",
       "      <td>Cells</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>364992</td>\n",
       "      <td>F</td>\n",
       "      <td>83</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Rasgrf2-T2A-dgFlpO/wt;Ai65F/wt</td>\n",
       "      <td>Rasgrf2-T2A-dgFlpO</td>\n",
       "      <td>...</td>\n",
       "      <td>10652</td>\n",
       "      <td>9966</td>\n",
       "      <td>173.35</td>\n",
       "      <td>173.35</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>L5 IT</td>\n",
       "      <td>L5 IT ALM Npw</td>\n",
       "      <td>0.6329</td>\n",
       "      <td>0.890471</td>\n",
       "      <td>Core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>F1S4_180124_315_A01</td>\n",
       "      <td>665943366</td>\n",
       "      <td>Cells</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>364992</td>\n",
       "      <td>F</td>\n",
       "      <td>83</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Rasgrf2-T2A-dgFlpO/wt;Ai65F/wt</td>\n",
       "      <td>Rasgrf2-T2A-dgFlpO</td>\n",
       "      <td>...</td>\n",
       "      <td>9626</td>\n",
       "      <td>8933</td>\n",
       "      <td>151.30</td>\n",
       "      <td>151.30</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Lamp5</td>\n",
       "      <td>Lamp5 Fam19a1 Pax6</td>\n",
       "      <td>0.4529</td>\n",
       "      <td>0.850266</td>\n",
       "      <td>Core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>F1S4_180124_315_B01</td>\n",
       "      <td>665943389</td>\n",
       "      <td>Cells</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>364992</td>\n",
       "      <td>F</td>\n",
       "      <td>83</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Rasgrf2-T2A-dgFlpO/wt;Ai65F/wt</td>\n",
       "      <td>Rasgrf2-T2A-dgFlpO</td>\n",
       "      <td>...</td>\n",
       "      <td>10463</td>\n",
       "      <td>9778</td>\n",
       "      <td>65.04</td>\n",
       "      <td>65.04</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Sncg</td>\n",
       "      <td>Sncg Slc17a8</td>\n",
       "      <td>0.0078</td>\n",
       "      <td>0.893137</td>\n",
       "      <td>Core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>F1S4_180124_315_C01</td>\n",
       "      <td>665943405</td>\n",
       "      <td>Cells</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>364992</td>\n",
       "      <td>F</td>\n",
       "      <td>83</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Rasgrf2-T2A-dgFlpO/wt;Ai65F/wt</td>\n",
       "      <td>Rasgrf2-T2A-dgFlpO</td>\n",
       "      <td>...</td>\n",
       "      <td>10848</td>\n",
       "      <td>10234</td>\n",
       "      <td>57.92</td>\n",
       "      <td>57.92</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Sncg</td>\n",
       "      <td>Sncg Slc17a8</td>\n",
       "      <td>0.0116</td>\n",
       "      <td>0.871497</td>\n",
       "      <td>Core</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           sample_name  sample_id sample_type      organism   donor sex  \\\n",
       "0  F1S4_161216_001_A01  571878193       Cells  Mus musculus  286575   M   \n",
       "1  F1S4_180124_314_A01  665943217       Cells  Mus musculus  364992   F   \n",
       "2  F1S4_180124_315_A01  665943366       Cells  Mus musculus  364992   F   \n",
       "3  F1S4_180124_315_B01  665943389       Cells  Mus musculus  364992   F   \n",
       "4  F1S4_180124_315_C01  665943405       Cells  Mus musculus  364992   F   \n",
       "\n",
       "   age_days eye_condition                        genotype        driver_lines  \\\n",
       "0        57        Normal                Ai14(RCL-tdT)/wt                 NaN   \n",
       "1        83        Normal  Rasgrf2-T2A-dgFlpO/wt;Ai65F/wt  Rasgrf2-T2A-dgFlpO   \n",
       "2        83        Normal  Rasgrf2-T2A-dgFlpO/wt;Ai65F/wt  Rasgrf2-T2A-dgFlpO   \n",
       "3        83        Normal  Rasgrf2-T2A-dgFlpO/wt;Ai65F/wt  Rasgrf2-T2A-dgFlpO   \n",
       "4        83        Normal  Rasgrf2-T2A-dgFlpO/wt;Ai65F/wt  Rasgrf2-T2A-dgFlpO   \n",
       "\n",
       "   ... genes_detected_cpm_criterion genes_detected_fpkm_criterion tdt_cpm  \\\n",
       "0  ...                        12965                         12128  127.17   \n",
       "1  ...                        10652                          9966  173.35   \n",
       "2  ...                         9626                          8933  151.30   \n",
       "3  ...                        10463                          9778   65.04   \n",
       "4  ...                        10848                         10234   57.92   \n",
       "\n",
       "  gfp_cpm          class subclass             cluster confusion_score  \\\n",
       "0  127.17  Glutamatergic    L5 PT   L5 PT ALM Slco2a1          0.0131   \n",
       "1  173.35  Glutamatergic    L5 IT       L5 IT ALM Npw          0.6329   \n",
       "2  151.30      GABAergic    Lamp5  Lamp5 Fam19a1 Pax6          0.4529   \n",
       "3   65.04      GABAergic     Sncg        Sncg Slc17a8          0.0078   \n",
       "4   57.92      GABAergic     Sncg        Sncg Slc17a8          0.0116   \n",
       "\n",
       "  cluster_correlation core_intermediate_call  \n",
       "0            0.845321                   Core  \n",
       "1            0.890471                   Core  \n",
       "2            0.850266                   Core  \n",
       "3            0.893137                   Core  \n",
       "4            0.871497                   Core  \n",
       "\n",
       "[5 rows x 50 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = '/cndd/Public_Datasets/Tasic_2018_nature/mouse_ALM_2018-06-14_genes-rows.csv'\n",
    "df_genes = pd.read_csv(f)\n",
    "print(df_genes.shape)\n",
    "df_genes.head()\n",
    "\n",
    "f = '/cndd/Public_Datasets/Tasic_2018_nature/mouse_ALM_2018-06-14_samples-columns.csv'\n",
    "df_cells = pd.read_csv(f, encoding=\"ISO-8859-1\")\n",
    "print(df_cells.shape)\n",
    "df_cells.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(45768, 10068)\n",
      "(45768, 10068)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>F1S4_161216_001_A01</th>\n",
       "      <th>F1S4_180124_314_A01</th>\n",
       "      <th>F1S4_180124_315_A01</th>\n",
       "      <th>F1S4_180124_315_B01</th>\n",
       "      <th>F1S4_180124_315_C01</th>\n",
       "      <th>F1S4_180124_315_D01</th>\n",
       "      <th>F1S4_180124_315_E01</th>\n",
       "      <th>F1S4_180124_315_F01</th>\n",
       "      <th>F1S4_180124_315_G01</th>\n",
       "      <th>F1S4_180124_315_H01</th>\n",
       "      <th>...</th>\n",
       "      <th>FJS4_170511_011_G01</th>\n",
       "      <th>FJS4_170511_011_H01</th>\n",
       "      <th>FJS4_170511_012_A01</th>\n",
       "      <th>FJS4_170511_012_B01</th>\n",
       "      <th>FJS4_170511_012_C01</th>\n",
       "      <th>FJS4_170511_012_D01</th>\n",
       "      <th>FJS4_170511_012_E01</th>\n",
       "      <th>FJS4_170511_012_F01</th>\n",
       "      <th>FJS4_170511_012_G01</th>\n",
       "      <th>FJS4_170511_012_H01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>71661</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76253</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58520</th>\n",
       "      <td>35</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>28</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66050</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100125929</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 10068 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           F1S4_161216_001_A01  F1S4_180124_314_A01  F1S4_180124_315_A01  \\\n",
       "71661                        0                    0                    0   \n",
       "76253                        0                    0                    0   \n",
       "58520                       35                    1                    0   \n",
       "66050                        4                    1                    7   \n",
       "100125929                    0                    0                    0   \n",
       "\n",
       "           F1S4_180124_315_B01  F1S4_180124_315_C01  F1S4_180124_315_D01  \\\n",
       "71661                        0                    0                    0   \n",
       "76253                        0                    0                    0   \n",
       "58520                        4                    1                   27   \n",
       "66050                        5                    3                    5   \n",
       "100125929                    3                    0                    0   \n",
       "\n",
       "           F1S4_180124_315_E01  F1S4_180124_315_F01  F1S4_180124_315_G01  \\\n",
       "71661                        0                    0                    0   \n",
       "76253                        0                    0                    0   \n",
       "58520                        0                    4                   25   \n",
       "66050                        2                    0                    0   \n",
       "100125929                    0                    0                    4   \n",
       "\n",
       "           F1S4_180124_315_H01  ...  FJS4_170511_011_G01  FJS4_170511_011_H01  \\\n",
       "71661                        0  ...                    0                    0   \n",
       "76253                        0  ...                    4                    0   \n",
       "58520                        0  ...                   28                   29   \n",
       "66050                        1  ...                    2                    3   \n",
       "100125929                    0  ...                    0                   22   \n",
       "\n",
       "           FJS4_170511_012_A01  FJS4_170511_012_B01  FJS4_170511_012_C01  \\\n",
       "71661                        0                    0                    0   \n",
       "76253                        0                    0                    0   \n",
       "58520                        0                    0                    0   \n",
       "66050                       12                    2                    6   \n",
       "100125929                    0                    0                    0   \n",
       "\n",
       "           FJS4_170511_012_D01  FJS4_170511_012_E01  FJS4_170511_012_F01  \\\n",
       "71661                        0                    0                    0   \n",
       "76253                        0                    0                    0   \n",
       "58520                        0                    0                    0   \n",
       "66050                        5                    4                   16   \n",
       "100125929                    0                   30                    0   \n",
       "\n",
       "           FJS4_170511_012_G01  FJS4_170511_012_H01  \n",
       "71661                        0                    0  \n",
       "76253                        0                    0  \n",
       "58520                        1                    0  \n",
       "66050                        0                    5  \n",
       "100125929                    0                    0  \n",
       "\n",
       "[5 rows x 10068 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = '/cndd/Public_Datasets/Tasic_2018_nature/mouse_ALM_2018-06-14_exon-matrix.csv'\n",
    "df1 = pd.read_csv(f, index_col=0)\n",
    "print(df1.shape)\n",
    "df1.head()\n",
    "\n",
    "f = '/cndd/Public_Datasets/Tasic_2018_nature/mouse_ALM_2018-06-14_intron-matrix.csv'\n",
    "df2 = pd.read_csv(f, index_col=0)\n",
    "print(df2.shape)\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(136, 45768)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <th>0610007P14Rik</th>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <th>0610009O20Rik</th>\n",
       "      <th>0610010B08Rik</th>\n",
       "      <th>0610010F05Rik</th>\n",
       "      <th>0610010K14Rik</th>\n",
       "      <th>...</th>\n",
       "      <th>n-R5s134</th>\n",
       "      <th>n-R5s136</th>\n",
       "      <th>n-R5s138</th>\n",
       "      <th>n-R5s139</th>\n",
       "      <th>n-R5s141</th>\n",
       "      <th>n-R5s142</th>\n",
       "      <th>n-R5s143</th>\n",
       "      <th>n-R5s144</th>\n",
       "      <th>n-R5s146</th>\n",
       "      <th>n-R5s149</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Astro Aqp4</th>\n",
       "      <td>2246</td>\n",
       "      <td>13</td>\n",
       "      <td>38561</td>\n",
       "      <td>32909</td>\n",
       "      <td>478</td>\n",
       "      <td>4643</td>\n",
       "      <td>8850</td>\n",
       "      <td>0</td>\n",
       "      <td>17353</td>\n",
       "      <td>14203</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Batch Grouping VISp L5 PT Ctxn3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>317</td>\n",
       "      <td>134</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>189</td>\n",
       "      <td>68</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CR Lhx5</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>272</td>\n",
       "      <td>1545</td>\n",
       "      <td>311</td>\n",
       "      <td>286</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1638</td>\n",
       "      <td>2323</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doublet Astro Aqp4 Ex</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2246</td>\n",
       "      <td>1662</td>\n",
       "      <td>0</td>\n",
       "      <td>90</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>591</td>\n",
       "      <td>508</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doublet Endo Peri SMC</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>193</td>\n",
       "      <td>341</td>\n",
       "      <td>0</td>\n",
       "      <td>94</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>281</td>\n",
       "      <td>237</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 45768 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 0610005C13Rik  0610006L08Rik  0610007P14Rik  \\\n",
       "cluster                                                                        \n",
       "Astro Aqp4                                2246             13          38561   \n",
       "Batch Grouping VISp L5 PT Ctxn3              0              0            317   \n",
       "CR Lhx5                                      0              0            272   \n",
       "Doublet Astro Aqp4 Ex                        0              0           2246   \n",
       "Doublet Endo Peri SMC                        0              0            193   \n",
       "\n",
       "                                 0610009B22Rik  0610009E02Rik  0610009L18Rik  \\\n",
       "cluster                                                                        \n",
       "Astro Aqp4                               32909            478           4643   \n",
       "Batch Grouping VISp L5 PT Ctxn3            134              0              3   \n",
       "CR Lhx5                                   1545            311            286   \n",
       "Doublet Astro Aqp4 Ex                     1662              0             90   \n",
       "Doublet Endo Peri SMC                      341              0             94   \n",
       "\n",
       "                                 0610009O20Rik  0610010B08Rik  0610010F05Rik  \\\n",
       "cluster                                                                        \n",
       "Astro Aqp4                                8850              0          17353   \n",
       "Batch Grouping VISp L5 PT Ctxn3              7              0            189   \n",
       "CR Lhx5                                      6              0           1638   \n",
       "Doublet Astro Aqp4 Ex                       32              0            591   \n",
       "Doublet Endo Peri SMC                       10              0            281   \n",
       "\n",
       "                                 0610010K14Rik  ...  n-R5s134  n-R5s136  \\\n",
       "cluster                                         ...                       \n",
       "Astro Aqp4                               14203  ...         0         6   \n",
       "Batch Grouping VISp L5 PT Ctxn3             68  ...         0         0   \n",
       "CR Lhx5                                   2323  ...         0         0   \n",
       "Doublet Astro Aqp4 Ex                      508  ...         0         0   \n",
       "Doublet Endo Peri SMC                      237  ...         0         0   \n",
       "\n",
       "                                 n-R5s138  n-R5s139  n-R5s141  n-R5s142  \\\n",
       "cluster                                                                   \n",
       "Astro Aqp4                              0         0         0         0   \n",
       "Batch Grouping VISp L5 PT Ctxn3         0         0         0         0   \n",
       "CR Lhx5                                 0         0         0         0   \n",
       "Doublet Astro Aqp4 Ex                   0         0         0         0   \n",
       "Doublet Endo Peri SMC                   0         0         0         0   \n",
       "\n",
       "                                 n-R5s143  n-R5s144  n-R5s146  n-R5s149  \n",
       "cluster                                                                  \n",
       "Astro Aqp4                              0         0         0         0  \n",
       "Batch Grouping VISp L5 PT Ctxn3         0         0         0         0  \n",
       "CR Lhx5                                 0         0         0         0  \n",
       "Doublet Astro Aqp4 Ex                   0         0         0         0  \n",
       "Doublet Endo Peri SMC                   0         0         0         0  \n",
       "\n",
       "[5 rows x 45768 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df1 + df2\n",
    "df.index = df_genes['gene_symbol'].values\n",
    "df = df.T\n",
    "df['cluster'] = df_cells.set_index('sample_name').loc[df.index, 'cluster']\n",
    "df = df.groupby('cluster').sum()\n",
    "\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>organism</th>\n",
       "      <th>donor</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_days</th>\n",
       "      <th>eye_condition</th>\n",
       "      <th>genotype</th>\n",
       "      <th>driver_lines</th>\n",
       "      <th>...</th>\n",
       "      <th>genes_detected_cpm_criterion</th>\n",
       "      <th>genes_detected_fpkm_criterion</th>\n",
       "      <th>tdt_cpm</th>\n",
       "      <th>gfp_cpm</th>\n",
       "      <th>class</th>\n",
       "      <th>subclass</th>\n",
       "      <th>cluster</th>\n",
       "      <th>confusion_score</th>\n",
       "      <th>cluster_correlation</th>\n",
       "      <th>core_intermediate_call</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>F1S4_161216_001_A01</td>\n",
       "      <td>571878193</td>\n",
       "      <td>Cells</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>286575</td>\n",
       "      <td>M</td>\n",
       "      <td>57</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Ai14(RCL-tdT)/wt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>12965</td>\n",
       "      <td>12128</td>\n",
       "      <td>127.17</td>\n",
       "      <td>127.17</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>L5 PT</td>\n",
       "      <td>L5 PT ALM Slco2a1</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>0.845321</td>\n",
       "      <td>Core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>F1S4_180124_314_A01</td>\n",
       "      <td>665943217</td>\n",
       "      <td>Cells</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>364992</td>\n",
       "      <td>F</td>\n",
       "      <td>83</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Rasgrf2-T2A-dgFlpO/wt;Ai65F/wt</td>\n",
       "      <td>Rasgrf2-T2A-dgFlpO</td>\n",
       "      <td>...</td>\n",
       "      <td>10652</td>\n",
       "      <td>9966</td>\n",
       "      <td>173.35</td>\n",
       "      <td>173.35</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>L5 IT</td>\n",
       "      <td>L5 IT ALM Npw</td>\n",
       "      <td>0.6329</td>\n",
       "      <td>0.890471</td>\n",
       "      <td>Core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>F1S4_180124_315_A01</td>\n",
       "      <td>665943366</td>\n",
       "      <td>Cells</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>364992</td>\n",
       "      <td>F</td>\n",
       "      <td>83</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Rasgrf2-T2A-dgFlpO/wt;Ai65F/wt</td>\n",
       "      <td>Rasgrf2-T2A-dgFlpO</td>\n",
       "      <td>...</td>\n",
       "      <td>9626</td>\n",
       "      <td>8933</td>\n",
       "      <td>151.30</td>\n",
       "      <td>151.30</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Lamp5</td>\n",
       "      <td>Lamp5 Fam19a1 Pax6</td>\n",
       "      <td>0.4529</td>\n",
       "      <td>0.850266</td>\n",
       "      <td>Core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>F1S4_180124_315_B01</td>\n",
       "      <td>665943389</td>\n",
       "      <td>Cells</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>364992</td>\n",
       "      <td>F</td>\n",
       "      <td>83</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Rasgrf2-T2A-dgFlpO/wt;Ai65F/wt</td>\n",
       "      <td>Rasgrf2-T2A-dgFlpO</td>\n",
       "      <td>...</td>\n",
       "      <td>10463</td>\n",
       "      <td>9778</td>\n",
       "      <td>65.04</td>\n",
       "      <td>65.04</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Sncg</td>\n",
       "      <td>Sncg Slc17a8</td>\n",
       "      <td>0.0078</td>\n",
       "      <td>0.893137</td>\n",
       "      <td>Core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>F1S4_180124_315_C01</td>\n",
       "      <td>665943405</td>\n",
       "      <td>Cells</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>364992</td>\n",
       "      <td>F</td>\n",
       "      <td>83</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Rasgrf2-T2A-dgFlpO/wt;Ai65F/wt</td>\n",
       "      <td>Rasgrf2-T2A-dgFlpO</td>\n",
       "      <td>...</td>\n",
       "      <td>10848</td>\n",
       "      <td>10234</td>\n",
       "      <td>57.92</td>\n",
       "      <td>57.92</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Sncg</td>\n",
       "      <td>Sncg Slc17a8</td>\n",
       "      <td>0.0116</td>\n",
       "      <td>0.871497</td>\n",
       "      <td>Core</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           sample_name  sample_id sample_type      organism   donor sex  \\\n",
       "0  F1S4_161216_001_A01  571878193       Cells  Mus musculus  286575   M   \n",
       "1  F1S4_180124_314_A01  665943217       Cells  Mus musculus  364992   F   \n",
       "2  F1S4_180124_315_A01  665943366       Cells  Mus musculus  364992   F   \n",
       "3  F1S4_180124_315_B01  665943389       Cells  Mus musculus  364992   F   \n",
       "4  F1S4_180124_315_C01  665943405       Cells  Mus musculus  364992   F   \n",
       "\n",
       "   age_days eye_condition                        genotype        driver_lines  \\\n",
       "0        57        Normal                Ai14(RCL-tdT)/wt                 NaN   \n",
       "1        83        Normal  Rasgrf2-T2A-dgFlpO/wt;Ai65F/wt  Rasgrf2-T2A-dgFlpO   \n",
       "2        83        Normal  Rasgrf2-T2A-dgFlpO/wt;Ai65F/wt  Rasgrf2-T2A-dgFlpO   \n",
       "3        83        Normal  Rasgrf2-T2A-dgFlpO/wt;Ai65F/wt  Rasgrf2-T2A-dgFlpO   \n",
       "4        83        Normal  Rasgrf2-T2A-dgFlpO/wt;Ai65F/wt  Rasgrf2-T2A-dgFlpO   \n",
       "\n",
       "   ... genes_detected_cpm_criterion genes_detected_fpkm_criterion tdt_cpm  \\\n",
       "0  ...                        12965                         12128  127.17   \n",
       "1  ...                        10652                          9966  173.35   \n",
       "2  ...                         9626                          8933  151.30   \n",
       "3  ...                        10463                          9778   65.04   \n",
       "4  ...                        10848                         10234   57.92   \n",
       "\n",
       "  gfp_cpm          class subclass             cluster confusion_score  \\\n",
       "0  127.17  Glutamatergic    L5 PT   L5 PT ALM Slco2a1          0.0131   \n",
       "1  173.35  Glutamatergic    L5 IT       L5 IT ALM Npw          0.6329   \n",
       "2  151.30      GABAergic    Lamp5  Lamp5 Fam19a1 Pax6          0.4529   \n",
       "3   65.04      GABAergic     Sncg        Sncg Slc17a8          0.0078   \n",
       "4   57.92      GABAergic     Sncg        Sncg Slc17a8          0.0116   \n",
       "\n",
       "  cluster_correlation core_intermediate_call  \n",
       "0            0.845321                   Core  \n",
       "1            0.890471                   Core  \n",
       "2            0.850266                   Core  \n",
       "3            0.893137                   Core  \n",
       "4            0.871497                   Core  \n",
       "\n",
       "[5 rows x 50 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_hqcells = df_cells[df_cells['class'].str.contains(r'^Glu|^GABA')]\n",
    "df_hqcells = df_hqcells[~df_hqcells['cluster'].str.contains(r'VISp')]\n",
    "df_visp = df_hqcells[df_hqcells['cluster'].str.contains(r'VISp')]\n",
    "df_hqcells.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2 ['Glutamatergic' 'GABAergic']\n",
      "15 ['L5 PT' 'L5 IT' 'Lamp5' 'Sncg' 'L2/3 IT' 'Sst' 'Vip' 'Pvalb' 'Serpinf1'\n",
      " 'L6 CT' 'L6 IT' 'NP' 'CR' 'Meis2' 'L6b']\n",
      "86 ['L5 PT ALM Slco2a1' 'L5 IT ALM Npw' 'Lamp5 Fam19a1 Pax6' 'Sncg Slc17a8'\n",
      " 'Lamp5 Krt73' 'Sncg Vip Itih5' 'L5 IT ALM Pld5' 'L2/3 IT ALM Macc1 Lrg1'\n",
      " 'Lamp5 Lsp1' 'L2/3 IT ALM Sla' 'Sst Myh8 Etv1 ' 'Vip Arhgap36 Hmcn1'\n",
      " 'Sst Tac1 Htr1d' 'Sst Hpse Sema3c' 'Sst Calb2 Pdlim5' 'Pvalb Reln Tac1'\n",
      " 'Sst Hpse Cbln4' 'Sst Tac2 Myh4' 'Sst Chrna2 Ptgdr' 'Sst Chrna2 Glra3'\n",
      " 'Sst Chodl' 'Sst Calb2 Necab1' 'Sst Crhr2 Efemp1' 'Sst Mme Fam114a1'\n",
      " 'Sst Nr2f2 Necab1' 'Sst Rxfp1 Prdm8' 'Pvalb Gabrg1' 'Sst Myh8 Fibin'\n",
      " 'Sst Tac2 Tacstd2' 'Sst Rxfp1 Eya1' 'Sst Crh 4930553C11Rik '\n",
      " 'Pvalb Th Sst' 'Sst Esm1' 'Sst Nts' 'L2/3 IT ALM Ptrf'\n",
      " 'Serpinf1 Aqp5 Vip' 'Vip Crispld2 Htr2c' 'Vip Ptprt Pkp2'\n",
      " 'Vip Crispld2 Kcne4' 'Vip Lmo1 Myl1' 'Vip Chat Htr1f' 'Vip Pygm C1ql1'\n",
      " 'Vip Lect1 Oxtr' 'Vip Rspo1 Itga4' 'Vip Igfbp4 Mab21l1'\n",
      " 'Vip Lmo1 Fam159b' 'Sncg Vip Nptx2' 'Vip Rspo4 Rxfp1 Chat'\n",
      " 'Vip Col15a1 Pde1a' 'Vip Gpc3 Slc18a3' 'Serpinf1 Clrn1'\n",
      " 'Vip Igfbp6 Car10' 'L6 CT ALM Nxph2 Sla' 'L6 CT ALM Cpa6' 'L5 IT ALM Tnc'\n",
      " 'L5 IT ALM Lypd1 Gpr88' 'L6 IT ALM Tgfb1' 'L5 IT ALM Cbln4 Fezf2'\n",
      " 'L5 IT ALM Gkn1 Pcdh19' 'L5 IT ALM Cpa6 Gpr88' 'L5 NP ALM Trhr Nefl'\n",
      " 'Lamp5 Plch2 Dock5' 'L5 IT ALM Tmem163 Dmrtb1' 'L6 IT ALM Oprk1'\n",
      " 'Pvalb Gpr149 Islr' 'Pvalb Sema3e Kank4' 'Pvalb Tpbg' 'Pvalb Vipr2'\n",
      " 'L6 NP ALM Trh' 'Vip Igfbp6 Pltp' 'Sncg Gpr50' 'Lamp5 Ntn1 Npy2r'\n",
      " 'CR Lhx5' 'Lamp5 Fam19a1 Tmem182' 'L5 IT ALM Tmem163 Arhgap25'\n",
      " 'Pvalb Akr1c18 Ntf3' 'Pvalb Calb1 Sst' 'Lamp5 Lhx6' 'Pvalb Reln Itm2a'\n",
      " 'Meis2 Adamts19' 'L5 PT ALM Npsr1' 'L6b Hsd17b2' 'L5 PT ALM Hpgd'\n",
      " 'L6b ALM Olfr111 Nxph1' 'L6b ALM Olfr111 Spon1' 'L6b P2ry12']\n"
     ]
    }
   ],
   "source": [
    "x = df_hqcells['class'].unique()\n",
    "y = df_hqcells['subclass'].unique()\n",
    "z = df_hqcells['cluster'].unique()\n",
    "\n",
    "print(len(x), x)\n",
    "print(len(y), y)\n",
    "print(len(z), z)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(86, 45768)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <th>0610007P14Rik</th>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <th>0610009O20Rik</th>\n",
       "      <th>0610010B08Rik</th>\n",
       "      <th>0610010F05Rik</th>\n",
       "      <th>0610010K14Rik</th>\n",
       "      <th>...</th>\n",
       "      <th>n-R5s134</th>\n",
       "      <th>n-R5s136</th>\n",
       "      <th>n-R5s138</th>\n",
       "      <th>n-R5s139</th>\n",
       "      <th>n-R5s141</th>\n",
       "      <th>n-R5s142</th>\n",
       "      <th>n-R5s143</th>\n",
       "      <th>n-R5s144</th>\n",
       "      <th>n-R5s146</th>\n",
       "      <th>n-R5s149</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>L5 PT ALM Slco2a1</th>\n",
       "      <td>24</td>\n",
       "      <td>40</td>\n",
       "      <td>39002</td>\n",
       "      <td>44210</td>\n",
       "      <td>630</td>\n",
       "      <td>4334</td>\n",
       "      <td>2240</td>\n",
       "      <td>2</td>\n",
       "      <td>47353</td>\n",
       "      <td>24068</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L5 IT ALM Npw</th>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>12899</td>\n",
       "      <td>9635</td>\n",
       "      <td>195</td>\n",
       "      <td>185</td>\n",
       "      <td>527</td>\n",
       "      <td>0</td>\n",
       "      <td>18519</td>\n",
       "      <td>4311</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lamp5 Fam19a1 Pax6</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8663</td>\n",
       "      <td>7407</td>\n",
       "      <td>87</td>\n",
       "      <td>363</td>\n",
       "      <td>654</td>\n",
       "      <td>0</td>\n",
       "      <td>6960</td>\n",
       "      <td>3188</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sncg Slc17a8</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5830</td>\n",
       "      <td>7590</td>\n",
       "      <td>121</td>\n",
       "      <td>732</td>\n",
       "      <td>856</td>\n",
       "      <td>0</td>\n",
       "      <td>7249</td>\n",
       "      <td>4122</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lamp5 Krt73</th>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>4262</td>\n",
       "      <td>5247</td>\n",
       "      <td>51</td>\n",
       "      <td>463</td>\n",
       "      <td>697</td>\n",
       "      <td>0</td>\n",
       "      <td>6836</td>\n",
       "      <td>3385</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 45768 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0610005C13Rik  0610006L08Rik  0610007P14Rik  \\\n",
       "cluster                                                           \n",
       "L5 PT ALM Slco2a1              24             40          39002   \n",
       "L5 IT ALM Npw                   5              3          12899   \n",
       "Lamp5 Fam19a1 Pax6              0              0           8663   \n",
       "Sncg Slc17a8                    0              0           5830   \n",
       "Lamp5 Krt73                     6              0           4262   \n",
       "\n",
       "                    0610009B22Rik  0610009E02Rik  0610009L18Rik  \\\n",
       "cluster                                                           \n",
       "L5 PT ALM Slco2a1           44210            630           4334   \n",
       "L5 IT ALM Npw                9635            195            185   \n",
       "Lamp5 Fam19a1 Pax6           7407             87            363   \n",
       "Sncg Slc17a8                 7590            121            732   \n",
       "Lamp5 Krt73                  5247             51            463   \n",
       "\n",
       "                    0610009O20Rik  0610010B08Rik  0610010F05Rik  \\\n",
       "cluster                                                           \n",
       "L5 PT ALM Slco2a1            2240              2          47353   \n",
       "L5 IT ALM Npw                 527              0          18519   \n",
       "Lamp5 Fam19a1 Pax6            654              0           6960   \n",
       "Sncg Slc17a8                  856              0           7249   \n",
       "Lamp5 Krt73                   697              0           6836   \n",
       "\n",
       "                    0610010K14Rik  ...  n-R5s134  n-R5s136  n-R5s138  \\\n",
       "cluster                            ...                                 \n",
       "L5 PT ALM Slco2a1           24068  ...         0         0         0   \n",
       "L5 IT ALM Npw                4311  ...         0         0         0   \n",
       "Lamp5 Fam19a1 Pax6           3188  ...         0         0         0   \n",
       "Sncg Slc17a8                 4122  ...         0         0         0   \n",
       "Lamp5 Krt73                  3385  ...         0         0         0   \n",
       "\n",
       "                    n-R5s139  n-R5s141  n-R5s142  n-R5s143  n-R5s144  \\\n",
       "cluster                                                                \n",
       "L5 PT ALM Slco2a1          0         0         0         0         0   \n",
       "L5 IT ALM Npw              0         0         0         0         0   \n",
       "Lamp5 Fam19a1 Pax6         0         0         0         0         0   \n",
       "Sncg Slc17a8               0         0         0         0         0   \n",
       "Lamp5 Krt73                0         0         0         0         0   \n",
       "\n",
       "                    n-R5s146  n-R5s149  \n",
       "cluster                                 \n",
       "L5 PT ALM Slco2a1          0         0  \n",
       "L5 IT ALM Npw              0         0  \n",
       "Lamp5 Fam19a1 Pax6         0         0  \n",
       "Sncg Slc17a8               0         0  \n",
       "Lamp5 Krt73                0         0  \n",
       "\n",
       "[5 rows x 45768 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_hq = df.loc[z]\n",
    "print(df_hq.shape)\n",
    "df_hq.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(32324, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>ENSMUSG00000109644</td>\n",
       "      <td>chr7</td>\n",
       "      <td>45567794.0</td>\n",
       "      <td>45575327.0</td>\n",
       "      <td>-</td>\n",
       "      <td>antisense_RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <td>ENSMUSG00000108652</td>\n",
       "      <td>chr7</td>\n",
       "      <td>74818818.0</td>\n",
       "      <td>74853813.0</td>\n",
       "      <td>-</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>ENSMUSG00000007777</td>\n",
       "      <td>chr11</td>\n",
       "      <td>51685386.0</td>\n",
       "      <td>51688874.0</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>ENSMUSG00000086714</td>\n",
       "      <td>chr2</td>\n",
       "      <td>26445696.0</td>\n",
       "      <td>26459390.0</td>\n",
       "      <td>+</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>ENSMUSG00000043644</td>\n",
       "      <td>chr11</td>\n",
       "      <td>120348678.0</td>\n",
       "      <td>120351190.0</td>\n",
       "      <td>+</td>\n",
       "      <td>bidirectional_promoter_lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          gene_id    chr        start          end strand  \\\n",
       "gene_name                                                                   \n",
       "0610005C13Rik  ENSMUSG00000109644   chr7   45567794.0   45575327.0      -   \n",
       "0610006L08Rik  ENSMUSG00000108652   chr7   74818818.0   74853813.0      -   \n",
       "0610009B22Rik  ENSMUSG00000007777  chr11   51685386.0   51688874.0      -   \n",
       "0610009E02Rik  ENSMUSG00000086714   chr2   26445696.0   26459390.0      +   \n",
       "0610009L18Rik  ENSMUSG00000043644  chr11  120348678.0  120351190.0      +   \n",
       "\n",
       "                                   gene_type  \n",
       "gene_name                                     \n",
       "0610005C13Rik                  antisense_RNA  \n",
       "0610006L08Rik                        lincRNA  \n",
       "0610009B22Rik                 protein_coding  \n",
       "0610009E02Rik           processed_transcript  \n",
       "0610009L18Rik  bidirectional_promoter_lncRNA  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genes_v3 = df_genes_v2.reindex(df_hq.columns.values).dropna()\n",
    "print(df_genes_v3.shape)\n",
    "df_genes_v3.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(86, 32324)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSMUSG00000109644</th>\n",
       "      <th>ENSMUSG00000108652</th>\n",
       "      <th>ENSMUSG00000007777</th>\n",
       "      <th>ENSMUSG00000086714</th>\n",
       "      <th>ENSMUSG00000043644</th>\n",
       "      <th>ENSMUSG00000024442</th>\n",
       "      <th>ENSMUSG00000042208</th>\n",
       "      <th>ENSMUSG00000020831</th>\n",
       "      <th>ENSMUSG00000107002</th>\n",
       "      <th>ENSMUSG00000046683</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSMUSG00000096015</th>\n",
       "      <th>ENSMUSG00000096037</th>\n",
       "      <th>ENSMUSG00000095183</th>\n",
       "      <th>ENSMUSG00000094760</th>\n",
       "      <th>ENSMUSG00000095749</th>\n",
       "      <th>ENSMUSG00000094657</th>\n",
       "      <th>ENSMUSG00000095645</th>\n",
       "      <th>ENSMUSG00000096765</th>\n",
       "      <th>ENSMUSG00000096170</th>\n",
       "      <th>ENSMUSG00000095453</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>L5 PT ALM Slco2a1</th>\n",
       "      <td>24</td>\n",
       "      <td>40</td>\n",
       "      <td>44210</td>\n",
       "      <td>630</td>\n",
       "      <td>4334</td>\n",
       "      <td>2240</td>\n",
       "      <td>47353</td>\n",
       "      <td>24068</td>\n",
       "      <td>93056</td>\n",
       "      <td>56</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L5 IT ALM Npw</th>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>9635</td>\n",
       "      <td>195</td>\n",
       "      <td>185</td>\n",
       "      <td>527</td>\n",
       "      <td>18519</td>\n",
       "      <td>4311</td>\n",
       "      <td>17587</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lamp5 Fam19a1 Pax6</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7407</td>\n",
       "      <td>87</td>\n",
       "      <td>363</td>\n",
       "      <td>654</td>\n",
       "      <td>6960</td>\n",
       "      <td>3188</td>\n",
       "      <td>8815</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sncg Slc17a8</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7590</td>\n",
       "      <td>121</td>\n",
       "      <td>732</td>\n",
       "      <td>856</td>\n",
       "      <td>7249</td>\n",
       "      <td>4122</td>\n",
       "      <td>14223</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lamp5 Krt73</th>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>5247</td>\n",
       "      <td>51</td>\n",
       "      <td>463</td>\n",
       "      <td>697</td>\n",
       "      <td>6836</td>\n",
       "      <td>3385</td>\n",
       "      <td>8224</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32324 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ENSMUSG00000109644  ENSMUSG00000108652  \\\n",
       "cluster                                                      \n",
       "L5 PT ALM Slco2a1                   24                  40   \n",
       "L5 IT ALM Npw                        5                   3   \n",
       "Lamp5 Fam19a1 Pax6                   0                   0   \n",
       "Sncg Slc17a8                         0                   0   \n",
       "Lamp5 Krt73                          6                   0   \n",
       "\n",
       "                    ENSMUSG00000007777  ENSMUSG00000086714  \\\n",
       "cluster                                                      \n",
       "L5 PT ALM Slco2a1                44210                 630   \n",
       "L5 IT ALM Npw                     9635                 195   \n",
       "Lamp5 Fam19a1 Pax6                7407                  87   \n",
       "Sncg Slc17a8                      7590                 121   \n",
       "Lamp5 Krt73                       5247                  51   \n",
       "\n",
       "                    ENSMUSG00000043644  ENSMUSG00000024442  \\\n",
       "cluster                                                      \n",
       "L5 PT ALM Slco2a1                 4334                2240   \n",
       "L5 IT ALM Npw                      185                 527   \n",
       "Lamp5 Fam19a1 Pax6                 363                 654   \n",
       "Sncg Slc17a8                       732                 856   \n",
       "Lamp5 Krt73                        463                 697   \n",
       "\n",
       "                    ENSMUSG00000042208  ENSMUSG00000020831  \\\n",
       "cluster                                                      \n",
       "L5 PT ALM Slco2a1                47353               24068   \n",
       "L5 IT ALM Npw                    18519                4311   \n",
       "Lamp5 Fam19a1 Pax6                6960                3188   \n",
       "Sncg Slc17a8                      7249                4122   \n",
       "Lamp5 Krt73                       6836                3385   \n",
       "\n",
       "                    ENSMUSG00000107002  ENSMUSG00000046683  ...  \\\n",
       "cluster                                                     ...   \n",
       "L5 PT ALM Slco2a1                93056                  56  ...   \n",
       "L5 IT ALM Npw                    17587                   3  ...   \n",
       "Lamp5 Fam19a1 Pax6                8815                   7  ...   \n",
       "Sncg Slc17a8                     14223                   4  ...   \n",
       "Lamp5 Krt73                       8224                   0  ...   \n",
       "\n",
       "                    ENSMUSG00000096015  ENSMUSG00000096037  \\\n",
       "cluster                                                      \n",
       "L5 PT ALM Slco2a1                    0                   0   \n",
       "L5 IT ALM Npw                        0                   0   \n",
       "Lamp5 Fam19a1 Pax6                   0                   0   \n",
       "Sncg Slc17a8                         0                   0   \n",
       "Lamp5 Krt73                          0                   0   \n",
       "\n",
       "                    ENSMUSG00000095183  ENSMUSG00000094760  \\\n",
       "cluster                                                      \n",
       "L5 PT ALM Slco2a1                    0                   0   \n",
       "L5 IT ALM Npw                        0                   0   \n",
       "Lamp5 Fam19a1 Pax6                   0                   0   \n",
       "Sncg Slc17a8                         0                   0   \n",
       "Lamp5 Krt73                          0                   0   \n",
       "\n",
       "                    ENSMUSG00000095749  ENSMUSG00000094657  \\\n",
       "cluster                                                      \n",
       "L5 PT ALM Slco2a1                    0                   0   \n",
       "L5 IT ALM Npw                        0                   0   \n",
       "Lamp5 Fam19a1 Pax6                   0                   0   \n",
       "Sncg Slc17a8                         0                   0   \n",
       "Lamp5 Krt73                          0                   0   \n",
       "\n",
       "                    ENSMUSG00000095645  ENSMUSG00000096765  \\\n",
       "cluster                                                      \n",
       "L5 PT ALM Slco2a1                    0                   0   \n",
       "L5 IT ALM Npw                        0                   0   \n",
       "Lamp5 Fam19a1 Pax6                   0                   0   \n",
       "Sncg Slc17a8                         0                   0   \n",
       "Lamp5 Krt73                          0                   0   \n",
       "\n",
       "                    ENSMUSG00000096170  ENSMUSG00000095453  \n",
       "cluster                                                     \n",
       "L5 PT ALM Slco2a1                    0                   0  \n",
       "L5 IT ALM Npw                        0                   0  \n",
       "Lamp5 Fam19a1 Pax6                   0                   0  \n",
       "Sncg Slc17a8                         0                   0  \n",
       "Lamp5 Krt73                          0                   0  \n",
       "\n",
       "[5 rows x 32324 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_hq = df_hq[df_genes_v3.index.values]\n",
    "df_hq.columns = df_genes_v3['gene_id'].values\n",
    "print(df_hq.shape)\n",
    "df_hq.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(32324, 86)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>cluster</th>\n",
       "      <th>L5 PT ALM Slco2a1</th>\n",
       "      <th>L5 IT ALM Npw</th>\n",
       "      <th>Lamp5 Fam19a1 Pax6</th>\n",
       "      <th>Sncg Slc17a8</th>\n",
       "      <th>Lamp5 Krt73</th>\n",
       "      <th>Sncg Vip Itih5</th>\n",
       "      <th>L5 IT ALM Pld5</th>\n",
       "      <th>L2/3 IT ALM Macc1 Lrg1</th>\n",
       "      <th>Lamp5 Lsp1</th>\n",
       "      <th>L2/3 IT ALM Sla</th>\n",
       "      <th>...</th>\n",
       "      <th>Pvalb Calb1 Sst</th>\n",
       "      <th>Lamp5 Lhx6</th>\n",
       "      <th>Pvalb Reln Itm2a</th>\n",
       "      <th>Meis2 Adamts19</th>\n",
       "      <th>L5 PT ALM Npsr1</th>\n",
       "      <th>L6b Hsd17b2</th>\n",
       "      <th>L5 PT ALM Hpgd</th>\n",
       "      <th>L6b ALM Olfr111 Nxph1</th>\n",
       "      <th>L6b ALM Olfr111 Spon1</th>\n",
       "      <th>L6b P2ry12</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000109644</th>\n",
       "      <td>24</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>23</td>\n",
       "      <td>51</td>\n",
       "      <td>16</td>\n",
       "      <td>173</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000108652</th>\n",
       "      <td>40</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>2</td>\n",
       "      <td>249</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000007777</th>\n",
       "      <td>44210</td>\n",
       "      <td>9635</td>\n",
       "      <td>7407</td>\n",
       "      <td>7590</td>\n",
       "      <td>5247</td>\n",
       "      <td>7352</td>\n",
       "      <td>72824</td>\n",
       "      <td>25487</td>\n",
       "      <td>86500</td>\n",
       "      <td>21912</td>\n",
       "      <td>...</td>\n",
       "      <td>9567</td>\n",
       "      <td>6357</td>\n",
       "      <td>985</td>\n",
       "      <td>1858</td>\n",
       "      <td>10203</td>\n",
       "      <td>2709</td>\n",
       "      <td>28964</td>\n",
       "      <td>12144</td>\n",
       "      <td>10188</td>\n",
       "      <td>1541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000086714</th>\n",
       "      <td>630</td>\n",
       "      <td>195</td>\n",
       "      <td>87</td>\n",
       "      <td>121</td>\n",
       "      <td>51</td>\n",
       "      <td>81</td>\n",
       "      <td>1154</td>\n",
       "      <td>669</td>\n",
       "      <td>913</td>\n",
       "      <td>369</td>\n",
       "      <td>...</td>\n",
       "      <td>68</td>\n",
       "      <td>347</td>\n",
       "      <td>30</td>\n",
       "      <td>3</td>\n",
       "      <td>95</td>\n",
       "      <td>20</td>\n",
       "      <td>442</td>\n",
       "      <td>191</td>\n",
       "      <td>204</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000043644</th>\n",
       "      <td>4334</td>\n",
       "      <td>185</td>\n",
       "      <td>363</td>\n",
       "      <td>732</td>\n",
       "      <td>463</td>\n",
       "      <td>413</td>\n",
       "      <td>2552</td>\n",
       "      <td>812</td>\n",
       "      <td>7158</td>\n",
       "      <td>1247</td>\n",
       "      <td>...</td>\n",
       "      <td>889</td>\n",
       "      <td>330</td>\n",
       "      <td>145</td>\n",
       "      <td>265</td>\n",
       "      <td>439</td>\n",
       "      <td>114</td>\n",
       "      <td>1710</td>\n",
       "      <td>560</td>\n",
       "      <td>731</td>\n",
       "      <td>197</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 86 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "cluster             L5 PT ALM Slco2a1  L5 IT ALM Npw  Lamp5 Fam19a1 Pax6  \\\n",
       "ENSMUSG00000109644                 24              5                   0   \n",
       "ENSMUSG00000108652                 40              3                   0   \n",
       "ENSMUSG00000007777              44210           9635                7407   \n",
       "ENSMUSG00000086714                630            195                  87   \n",
       "ENSMUSG00000043644               4334            185                 363   \n",
       "\n",
       "cluster             Sncg Slc17a8  Lamp5 Krt73  Sncg Vip Itih5  L5 IT ALM Pld5  \\\n",
       "ENSMUSG00000109644             0            6              23              51   \n",
       "ENSMUSG00000108652             0            0               0              27   \n",
       "ENSMUSG00000007777          7590         5247            7352           72824   \n",
       "ENSMUSG00000086714           121           51              81            1154   \n",
       "ENSMUSG00000043644           732          463             413            2552   \n",
       "\n",
       "cluster             L2/3 IT ALM Macc1 Lrg1  Lamp5 Lsp1  L2/3 IT ALM Sla  ...  \\\n",
       "ENSMUSG00000109644                      16         173                8  ...   \n",
       "ENSMUSG00000108652                       2         249                6  ...   \n",
       "ENSMUSG00000007777                   25487       86500            21912  ...   \n",
       "ENSMUSG00000086714                     669         913              369  ...   \n",
       "ENSMUSG00000043644                     812        7158             1247  ...   \n",
       "\n",
       "cluster             Pvalb Calb1 Sst  Lamp5 Lhx6  Pvalb Reln Itm2a  \\\n",
       "ENSMUSG00000109644                5          30                 0   \n",
       "ENSMUSG00000108652                1           0                 0   \n",
       "ENSMUSG00000007777             9567        6357               985   \n",
       "ENSMUSG00000086714               68         347                30   \n",
       "ENSMUSG00000043644              889         330               145   \n",
       "\n",
       "cluster             Meis2 Adamts19  L5 PT ALM Npsr1  L6b Hsd17b2  \\\n",
       "ENSMUSG00000109644               0                9            0   \n",
       "ENSMUSG00000108652               0               61            0   \n",
       "ENSMUSG00000007777            1858            10203         2709   \n",
       "ENSMUSG00000086714               3               95           20   \n",
       "ENSMUSG00000043644             265              439          114   \n",
       "\n",
       "cluster             L5 PT ALM Hpgd  L6b ALM Olfr111 Nxph1  \\\n",
       "ENSMUSG00000109644               8                      2   \n",
       "ENSMUSG00000108652              32                      1   \n",
       "ENSMUSG00000007777           28964                  12144   \n",
       "ENSMUSG00000086714             442                    191   \n",
       "ENSMUSG00000043644            1710                    560   \n",
       "\n",
       "cluster             L6b ALM Olfr111 Spon1  L6b P2ry12  \n",
       "ENSMUSG00000109644                      0           0  \n",
       "ENSMUSG00000108652                      0           0  \n",
       "ENSMUSG00000007777                  10188        1541  \n",
       "ENSMUSG00000086714                    204          24  \n",
       "ENSMUSG00000043644                    731         197  \n",
       "\n",
       "[5 rows x 86 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = '/cndd/fangming/CEMBA/data/MOp_all/Tasic_data/counts_ALM_gene_by_cluster.tsv'\n",
    "df_hq.T.to_csv(f, sep=\"\\t\", na_rep='NA', index=True, header=True)\n",
    "print(df_hq.T.shape)\n",
    "df_hq.T.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tasic 2018 processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(53379, 6)\n",
      "(53278, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>ENSMUSG00000109644</td>\n",
       "      <td>chr7</td>\n",
       "      <td>45567794</td>\n",
       "      <td>45575327</td>\n",
       "      <td>-</td>\n",
       "      <td>antisense_RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <td>ENSMUSG00000108652</td>\n",
       "      <td>chr7</td>\n",
       "      <td>74818818</td>\n",
       "      <td>74853813</td>\n",
       "      <td>-</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>ENSMUSG00000007777</td>\n",
       "      <td>chr11</td>\n",
       "      <td>51685386</td>\n",
       "      <td>51688874</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>ENSMUSG00000086714</td>\n",
       "      <td>chr2</td>\n",
       "      <td>26445696</td>\n",
       "      <td>26459390</td>\n",
       "      <td>+</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>ENSMUSG00000043644</td>\n",
       "      <td>chr11</td>\n",
       "      <td>120348678</td>\n",
       "      <td>120351190</td>\n",
       "      <td>+</td>\n",
       "      <td>bidirectional_promoter_lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          gene_id    chr      start        end strand  \\\n",
       "gene_name                                                               \n",
       "0610005C13Rik  ENSMUSG00000109644   chr7   45567794   45575327      -   \n",
       "0610006L08Rik  ENSMUSG00000108652   chr7   74818818   74853813      -   \n",
       "0610009B22Rik  ENSMUSG00000007777  chr11   51685386   51688874      -   \n",
       "0610009E02Rik  ENSMUSG00000086714   chr2   26445696   26459390      +   \n",
       "0610009L18Rik  ENSMUSG00000043644  chr11  120348678  120351190      +   \n",
       "\n",
       "                                   gene_type  \n",
       "gene_name                                     \n",
       "0610005C13Rik                  antisense_RNA  \n",
       "0610006L08Rik                        lincRNA  \n",
       "0610009B22Rik                 protein_coding  \n",
       "0610009E02Rik           processed_transcript  \n",
       "0610009L18Rik  bidirectional_promoter_lncRNA  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = PATH_GENEBODY_ANNOTATION \n",
    "df_genes = pd.read_csv(f, sep=\"\\t\")\n",
    "df_genes['gene_id'] = df_genes['gene_id'].apply(lambda x: x.split('.')[0])\n",
    "df_genes = df_genes.set_index('gene_id')\n",
    "df_genes_v2 = df_genes.reset_index().groupby('gene_name').first()\n",
    "print(df_genes.shape)\n",
    "df_genes.head()\n",
    "print(df_genes_v2.shape)\n",
    "df_genes_v2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(53379,)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "gene_id\n",
       "ENSMUSG00000102693      1069\n",
       "ENSMUSG00000064842       109\n",
       "ENSMUSG00000051951    465597\n",
       "ENSMUSG00000102851       479\n",
       "ENSMUSG00000103377      2818\n",
       "dtype: int64"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_lengths_base = df_genes['end'] - df_genes['start']\n",
    "print(gene_lengths_base.shape)\n",
    "gene_lengths_base.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(32324, 86)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>L5 PT ALM Slco2a1</th>\n",
       "      <th>L5 IT ALM Npw</th>\n",
       "      <th>Lamp5 Fam19a1 Pax6</th>\n",
       "      <th>Sncg Slc17a8</th>\n",
       "      <th>Lamp5 Krt73</th>\n",
       "      <th>Sncg Vip Itih5</th>\n",
       "      <th>L5 IT ALM Pld5</th>\n",
       "      <th>L2/3 IT ALM Macc1 Lrg1</th>\n",
       "      <th>Lamp5 Lsp1</th>\n",
       "      <th>L2/3 IT ALM Sla</th>\n",
       "      <th>...</th>\n",
       "      <th>Pvalb Calb1 Sst</th>\n",
       "      <th>Lamp5 Lhx6</th>\n",
       "      <th>Pvalb Reln Itm2a</th>\n",
       "      <th>Meis2 Adamts19</th>\n",
       "      <th>L5 PT ALM Npsr1</th>\n",
       "      <th>L6b Hsd17b2</th>\n",
       "      <th>L5 PT ALM Hpgd</th>\n",
       "      <th>L6b ALM Olfr111 Nxph1</th>\n",
       "      <th>L6b ALM Olfr111 Spon1</th>\n",
       "      <th>L6b P2ry12</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000109644</th>\n",
       "      <td>24</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>23</td>\n",
       "      <td>51</td>\n",
       "      <td>16</td>\n",
       "      <td>173</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000108652</th>\n",
       "      <td>40</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>2</td>\n",
       "      <td>249</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000007777</th>\n",
       "      <td>44210</td>\n",
       "      <td>9635</td>\n",
       "      <td>7407</td>\n",
       "      <td>7590</td>\n",
       "      <td>5247</td>\n",
       "      <td>7352</td>\n",
       "      <td>72824</td>\n",
       "      <td>25487</td>\n",
       "      <td>86500</td>\n",
       "      <td>21912</td>\n",
       "      <td>...</td>\n",
       "      <td>9567</td>\n",
       "      <td>6357</td>\n",
       "      <td>985</td>\n",
       "      <td>1858</td>\n",
       "      <td>10203</td>\n",
       "      <td>2709</td>\n",
       "      <td>28964</td>\n",
       "      <td>12144</td>\n",
       "      <td>10188</td>\n",
       "      <td>1541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000086714</th>\n",
       "      <td>630</td>\n",
       "      <td>195</td>\n",
       "      <td>87</td>\n",
       "      <td>121</td>\n",
       "      <td>51</td>\n",
       "      <td>81</td>\n",
       "      <td>1154</td>\n",
       "      <td>669</td>\n",
       "      <td>913</td>\n",
       "      <td>369</td>\n",
       "      <td>...</td>\n",
       "      <td>68</td>\n",
       "      <td>347</td>\n",
       "      <td>30</td>\n",
       "      <td>3</td>\n",
       "      <td>95</td>\n",
       "      <td>20</td>\n",
       "      <td>442</td>\n",
       "      <td>191</td>\n",
       "      <td>204</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000043644</th>\n",
       "      <td>4334</td>\n",
       "      <td>185</td>\n",
       "      <td>363</td>\n",
       "      <td>732</td>\n",
       "      <td>463</td>\n",
       "      <td>413</td>\n",
       "      <td>2552</td>\n",
       "      <td>812</td>\n",
       "      <td>7158</td>\n",
       "      <td>1247</td>\n",
       "      <td>...</td>\n",
       "      <td>889</td>\n",
       "      <td>330</td>\n",
       "      <td>145</td>\n",
       "      <td>265</td>\n",
       "      <td>439</td>\n",
       "      <td>114</td>\n",
       "      <td>1710</td>\n",
       "      <td>560</td>\n",
       "      <td>731</td>\n",
       "      <td>197</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 86 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    L5 PT ALM Slco2a1  L5 IT ALM Npw  Lamp5 Fam19a1 Pax6  \\\n",
       "ENSMUSG00000109644                 24              5                   0   \n",
       "ENSMUSG00000108652                 40              3                   0   \n",
       "ENSMUSG00000007777              44210           9635                7407   \n",
       "ENSMUSG00000086714                630            195                  87   \n",
       "ENSMUSG00000043644               4334            185                 363   \n",
       "\n",
       "                    Sncg Slc17a8  Lamp5 Krt73  Sncg Vip Itih5  L5 IT ALM Pld5  \\\n",
       "ENSMUSG00000109644             0            6              23              51   \n",
       "ENSMUSG00000108652             0            0               0              27   \n",
       "ENSMUSG00000007777          7590         5247            7352           72824   \n",
       "ENSMUSG00000086714           121           51              81            1154   \n",
       "ENSMUSG00000043644           732          463             413            2552   \n",
       "\n",
       "                    L2/3 IT ALM Macc1 Lrg1  Lamp5 Lsp1  L2/3 IT ALM Sla  ...  \\\n",
       "ENSMUSG00000109644                      16         173                8  ...   \n",
       "ENSMUSG00000108652                       2         249                6  ...   \n",
       "ENSMUSG00000007777                   25487       86500            21912  ...   \n",
       "ENSMUSG00000086714                     669         913              369  ...   \n",
       "ENSMUSG00000043644                     812        7158             1247  ...   \n",
       "\n",
       "                    Pvalb Calb1 Sst  Lamp5 Lhx6  Pvalb Reln Itm2a  \\\n",
       "ENSMUSG00000109644                5          30                 0   \n",
       "ENSMUSG00000108652                1           0                 0   \n",
       "ENSMUSG00000007777             9567        6357               985   \n",
       "ENSMUSG00000086714               68         347                30   \n",
       "ENSMUSG00000043644              889         330               145   \n",
       "\n",
       "                    Meis2 Adamts19  L5 PT ALM Npsr1  L6b Hsd17b2  \\\n",
       "ENSMUSG00000109644               0                9            0   \n",
       "ENSMUSG00000108652               0               61            0   \n",
       "ENSMUSG00000007777            1858            10203         2709   \n",
       "ENSMUSG00000086714               3               95           20   \n",
       "ENSMUSG00000043644             265              439          114   \n",
       "\n",
       "                    L5 PT ALM Hpgd  L6b ALM Olfr111 Nxph1  \\\n",
       "ENSMUSG00000109644               8                      2   \n",
       "ENSMUSG00000108652              32                      1   \n",
       "ENSMUSG00000007777           28964                  12144   \n",
       "ENSMUSG00000086714             442                    191   \n",
       "ENSMUSG00000043644            1710                    560   \n",
       "\n",
       "                    L6b ALM Olfr111 Spon1  L6b P2ry12  \n",
       "ENSMUSG00000109644                      0           0  \n",
       "ENSMUSG00000108652                      0           0  \n",
       "ENSMUSG00000007777                  10188        1541  \n",
       "ENSMUSG00000086714                    204          24  \n",
       "ENSMUSG00000043644                    731         197  \n",
       "\n",
       "[5 rows x 86 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = '/cndd/fangming/CEMBA/data/MOp_all/Tasic_data/counts_ALM_gene_by_cluster.tsv'\n",
    "counts = pd.read_csv(f, sep=\"\\t\", index_col=0)\n",
    "print(counts.shape)\n",
    "counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(32324, 86)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>L5 PT ALM Slco2a1</th>\n",
       "      <th>L5 IT ALM Npw</th>\n",
       "      <th>Lamp5 Fam19a1 Pax6</th>\n",
       "      <th>Sncg Slc17a8</th>\n",
       "      <th>Lamp5 Krt73</th>\n",
       "      <th>Sncg Vip Itih5</th>\n",
       "      <th>L5 IT ALM Pld5</th>\n",
       "      <th>L2/3 IT ALM Macc1 Lrg1</th>\n",
       "      <th>Lamp5 Lsp1</th>\n",
       "      <th>L2/3 IT ALM Sla</th>\n",
       "      <th>...</th>\n",
       "      <th>Pvalb Calb1 Sst</th>\n",
       "      <th>Lamp5 Lhx6</th>\n",
       "      <th>Pvalb Reln Itm2a</th>\n",
       "      <th>Meis2 Adamts19</th>\n",
       "      <th>L5 PT ALM Npsr1</th>\n",
       "      <th>L6b Hsd17b2</th>\n",
       "      <th>L5 PT ALM Hpgd</th>\n",
       "      <th>L6b ALM Olfr111 Nxph1</th>\n",
       "      <th>L6b ALM Olfr111 Spon1</th>\n",
       "      <th>L6b P2ry12</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000109644</th>\n",
       "      <td>0.034677</td>\n",
       "      <td>0.027257</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.066653</td>\n",
       "      <td>0.149167</td>\n",
       "      <td>0.037101</td>\n",
       "      <td>0.031673</td>\n",
       "      <td>0.107214</td>\n",
       "      <td>0.019169</td>\n",
       "      <td>...</td>\n",
       "      <td>0.034297</td>\n",
       "      <td>0.237029</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.056221</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018213</td>\n",
       "      <td>0.010616</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000108652</th>\n",
       "      <td>0.012762</td>\n",
       "      <td>0.003618</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004392</td>\n",
       "      <td>0.000883</td>\n",
       "      <td>0.036131</td>\n",
       "      <td>0.003153</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001534</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.079774</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015727</td>\n",
       "      <td>0.001155</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000007777</th>\n",
       "      <td>2.520725</td>\n",
       "      <td>2.432270</td>\n",
       "      <td>2.549302</td>\n",
       "      <td>2.443347</td>\n",
       "      <td>2.497324</td>\n",
       "      <td>2.453208</td>\n",
       "      <td>2.440950</td>\n",
       "      <td>2.417092</td>\n",
       "      <td>2.481972</td>\n",
       "      <td>2.428035</td>\n",
       "      <td>...</td>\n",
       "      <td>2.532211</td>\n",
       "      <td>2.522736</td>\n",
       "      <td>2.501284</td>\n",
       "      <td>2.546234</td>\n",
       "      <td>2.530684</td>\n",
       "      <td>2.605532</td>\n",
       "      <td>2.526183</td>\n",
       "      <td>2.512568</td>\n",
       "      <td>2.491395</td>\n",
       "      <td>2.421250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000086714</th>\n",
       "      <td>0.342476</td>\n",
       "      <td>0.378326</td>\n",
       "      <td>0.313194</td>\n",
       "      <td>0.326944</td>\n",
       "      <td>0.249348</td>\n",
       "      <td>0.253814</td>\n",
       "      <td>0.324298</td>\n",
       "      <td>0.437770</td>\n",
       "      <td>0.258375</td>\n",
       "      <td>0.331425</td>\n",
       "      <td>...</td>\n",
       "      <td>0.208111</td>\n",
       "      <td>0.749664</td>\n",
       "      <td>0.538158</td>\n",
       "      <td>0.058521</td>\n",
       "      <td>0.255875</td>\n",
       "      <td>0.244610</td>\n",
       "      <td>0.362042</td>\n",
       "      <td>0.361731</td>\n",
       "      <td>0.410960</td>\n",
       "      <td>0.310152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000043644</th>\n",
       "      <td>1.662881</td>\n",
       "      <td>0.913117</td>\n",
       "      <td>1.398597</td>\n",
       "      <td>1.580177</td>\n",
       "      <td>1.595336</td>\n",
       "      <td>1.363023</td>\n",
       "      <td>1.157834</td>\n",
       "      <td>1.097391</td>\n",
       "      <td>1.553195</td>\n",
       "      <td>1.344269</td>\n",
       "      <td>...</td>\n",
       "      <td>1.651418</td>\n",
       "      <td>1.397017</td>\n",
       "      <td>1.817070</td>\n",
       "      <td>1.847962</td>\n",
       "      <td>1.326664</td>\n",
       "      <td>1.389203</td>\n",
       "      <td>1.454121</td>\n",
       "      <td>1.338029</td>\n",
       "      <td>1.502259</td>\n",
       "      <td>1.678033</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 86 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    L5 PT ALM Slco2a1  L5 IT ALM Npw  Lamp5 Fam19a1 Pax6  \\\n",
       "ENSMUSG00000109644           0.034677       0.027257            0.000000   \n",
       "ENSMUSG00000108652           0.012762       0.003618            0.000000   \n",
       "ENSMUSG00000007777           2.520725       2.432270            2.549302   \n",
       "ENSMUSG00000086714           0.342476       0.378326            0.313194   \n",
       "ENSMUSG00000043644           1.662881       0.913117            1.398597   \n",
       "\n",
       "                    Sncg Slc17a8  Lamp5 Krt73  Sncg Vip Itih5  L5 IT ALM Pld5  \\\n",
       "ENSMUSG00000109644      0.000000     0.066653        0.149167        0.037101   \n",
       "ENSMUSG00000108652      0.000000     0.000000        0.000000        0.004392   \n",
       "ENSMUSG00000007777      2.443347     2.497324        2.453208        2.440950   \n",
       "ENSMUSG00000086714      0.326944     0.249348        0.253814        0.324298   \n",
       "ENSMUSG00000043644      1.580177     1.595336        1.363023        1.157834   \n",
       "\n",
       "                    L2/3 IT ALM Macc1 Lrg1  Lamp5 Lsp1  L2/3 IT ALM Sla  ...  \\\n",
       "ENSMUSG00000109644                0.031673    0.107214         0.019169  ...   \n",
       "ENSMUSG00000108652                0.000883    0.036131         0.003153  ...   \n",
       "ENSMUSG00000007777                2.417092    2.481972         2.428035  ...   \n",
       "ENSMUSG00000086714                0.437770    0.258375         0.331425  ...   \n",
       "ENSMUSG00000043644                1.097391    1.553195         1.344269  ...   \n",
       "\n",
       "                    Pvalb Calb1 Sst  Lamp5 Lhx6  Pvalb Reln Itm2a  \\\n",
       "ENSMUSG00000109644         0.034297    0.237029          0.000000   \n",
       "ENSMUSG00000108652         0.001534    0.000000          0.000000   \n",
       "ENSMUSG00000007777         2.532211    2.522736          2.501284   \n",
       "ENSMUSG00000086714         0.208111    0.749664          0.538158   \n",
       "ENSMUSG00000043644         1.651418    1.397017          1.817070   \n",
       "\n",
       "                    Meis2 Adamts19  L5 PT ALM Npsr1  L6b Hsd17b2  \\\n",
       "ENSMUSG00000109644        0.000000         0.056221     0.000000   \n",
       "ENSMUSG00000108652        0.000000         0.079774     0.000000   \n",
       "ENSMUSG00000007777        2.546234         2.530684     2.605532   \n",
       "ENSMUSG00000086714        0.058521         0.255875     0.244610   \n",
       "ENSMUSG00000043644        1.847962         1.326664     1.389203   \n",
       "\n",
       "                    L5 PT ALM Hpgd  L6b ALM Olfr111 Nxph1  \\\n",
       "ENSMUSG00000109644        0.018213               0.010616   \n",
       "ENSMUSG00000108652        0.015727               0.001155   \n",
       "ENSMUSG00000007777        2.526183               2.512568   \n",
       "ENSMUSG00000086714        0.362042               0.361731   \n",
       "ENSMUSG00000043644        1.454121               1.338029   \n",
       "\n",
       "                    L6b ALM Olfr111 Spon1  L6b P2ry12  \n",
       "ENSMUSG00000109644               0.000000    0.000000  \n",
       "ENSMUSG00000108652               0.000000    0.000000  \n",
       "ENSMUSG00000007777               2.491395    2.421250  \n",
       "ENSMUSG00000086714               0.410960    0.310152  \n",
       "ENSMUSG00000043644               1.502259    1.678033  \n",
       "\n",
       "[5 rows x 86 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_lengths = gene_lengths_base.reindex(counts.index)\n",
    "gene_lengths_mean = np.nanmean(gene_lengths)\n",
    "gene_lengths = gene_lengths.fillna(gene_lengths_mean)\n",
    "\n",
    "logtpm = snmcseq_utils.logtpm(counts, gene_lengths)\n",
    "print(logtpm.shape)\n",
    "logtpm.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
