{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Organize freezed data\n",
    "Aug 19th, 2019\n",
    "\n",
    "Sep 13th, 2019 (data freeze)\n",
    "\n",
    "Fangming Xie\n",
    "\n",
    "- reformat raw data (from the dropbox) into standardized python readable format\n",
    "- use ensembl id to refer genes\n",
    "- preserve content as much as possible, including naming conventions of table header and index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, '/cndd/fangming/CEMBA/snmcseq_dev')\n",
    "import importlib\n",
    "\n",
    "from __init__ import *\n",
    "from __init__jupyterlab import *\n",
    "import snmcseq_utils\n",
    "importlib.reload(snmcseq_utils)\n",
    "import CEMBA_clst_utils\n",
    "from scipy.io import mmread\n",
    "from scipy import sparse\n",
    "import time\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "SRC_DIR = '/cndd/fangming/CEMBA/data/MOp_all/data_freeze'\n",
    "DST_DIR = '/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(53379, 6)\n",
      "(53278, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>ENSMUSG00000109644</td>\n",
       "      <td>chr7</td>\n",
       "      <td>45567794</td>\n",
       "      <td>45575327</td>\n",
       "      <td>-</td>\n",
       "      <td>antisense_RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <td>ENSMUSG00000108652</td>\n",
       "      <td>chr7</td>\n",
       "      <td>74818818</td>\n",
       "      <td>74853813</td>\n",
       "      <td>-</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>ENSMUSG00000007777</td>\n",
       "      <td>chr11</td>\n",
       "      <td>51685386</td>\n",
       "      <td>51688874</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>ENSMUSG00000086714</td>\n",
       "      <td>chr2</td>\n",
       "      <td>26445696</td>\n",
       "      <td>26459390</td>\n",
       "      <td>+</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>ENSMUSG00000043644</td>\n",
       "      <td>chr11</td>\n",
       "      <td>120348678</td>\n",
       "      <td>120351190</td>\n",
       "      <td>+</td>\n",
       "      <td>bidirectional_promoter_lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          gene_id    chr      start        end strand  \\\n",
       "gene_name                                                               \n",
       "0610005C13Rik  ENSMUSG00000109644   chr7   45567794   45575327      -   \n",
       "0610006L08Rik  ENSMUSG00000108652   chr7   74818818   74853813      -   \n",
       "0610009B22Rik  ENSMUSG00000007777  chr11   51685386   51688874      -   \n",
       "0610009E02Rik  ENSMUSG00000086714   chr2   26445696   26459390      +   \n",
       "0610009L18Rik  ENSMUSG00000043644  chr11  120348678  120351190      +   \n",
       "\n",
       "                                   gene_type  \n",
       "gene_name                                     \n",
       "0610005C13Rik                  antisense_RNA  \n",
       "0610006L08Rik                        lincRNA  \n",
       "0610009B22Rik                 protein_coding  \n",
       "0610009E02Rik           processed_transcript  \n",
       "0610009L18Rik  bidirectional_promoter_lncRNA  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = PATH_GENEBODY_ANNOTATION \n",
    "df_genes = pd.read_csv(f, sep=\"\\t\")\n",
    "df_genes['gene_id'] = df_genes['gene_id'].apply(lambda x: x.split('.')[0])\n",
    "df_genes = df_genes.set_index('gene_id')\n",
    "df_genes_v2 = df_genes.reset_index().groupby('gene_name').first()\n",
    "print(df_genes.shape)\n",
    "df_genes.head()\n",
    "print(df_genes_v2.shape)\n",
    "df_genes_v2.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 10x_nuclei_v3_macosko \n",
    "- 2 metadata .csv\n",
    "- 1 data matrix in .rds format\n",
    "- concatenate male and female, get autosomo genes only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/10x_nuclei_v3_macosko_metadata.tsv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze/10x_nuclei_v3_macosko/./reformatted/raw_female.mtx\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/10x_nuclei_v3_macosko_counts.npz\n"
     ]
    }
   ],
   "source": [
    "dataset_name = '10x_nuclei_v3_macosko'\n",
    "\n",
    "\n",
    "# outdated\n",
    "f_meta = os.path.join(SRC_DIR, dataset_name, 'motor_cortex_v3_mop_data_metadata.csv') \n",
    "# f_meta4 = os.path.join(SRC_DIR, dataset_name, 'zizhen_processing', 'sample_metadata.csv') \n",
    "\n",
    "# zizhen's list \n",
    "f_meta1 = os.path.join(SRC_DIR, dataset_name, 'zizhen_processing', 'cluster.membership.csv') \n",
    "f_meta2 = os.path.join(SRC_DIR, dataset_name, 'zizhen_processing', 'cluster.annotation.csv') \n",
    "f_meta3 = os.path.join(SRC_DIR, dataset_name, 'zizhen_processing', 'QC.csv') \n",
    "\n",
    "f_data = os.path.join(SRC_DIR, dataset_name, './reformatted/raw_female.mtx') \n",
    "f_data_gene = os.path.join(SRC_DIR, dataset_name, './reformatted/raw_female_gene.csv') \n",
    "f_data_cell = os.path.join(SRC_DIR, dataset_name, './reformatted/raw_female_cell.csv') \n",
    "\n",
    "f_data2 = os.path.join(SRC_DIR, dataset_name, './reformatted/raw_male.mtx') \n",
    "f_data2_gene = os.path.join(SRC_DIR, dataset_name, './reformatted/raw_male_gene.csv') \n",
    "f_data2_cell = os.path.join(SRC_DIR, dataset_name, './reformatted/raw_male_cell.csv') \n",
    "\n",
    "fout_meta = os.path.join(DST_DIR, dataset_name+'_metadata.tsv') \n",
    "fout_data = os.path.join(DST_DIR, dataset_name+'_counts.npz') \n",
    "fout_data_gene = os.path.join(DST_DIR, dataset_name+'_counts.gene') \n",
    "fout_data_cell = os.path.join(DST_DIR, dataset_name+'_counts.cell') \n",
    "\n",
    "# print(f_meta)\n",
    "print(fout_meta)\n",
    "print(f_data)\n",
    "print(fout_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "672.5768630504608\n",
      "1225.8807742595673\n"
     ]
    }
   ],
   "source": [
    "ti = time.time()\n",
    "data = mmread(f_data)\n",
    "gene = pd.read_csv(f_data_gene)['x'].values\n",
    "cell = pd.read_csv(f_data_cell)['x'].values\n",
    "print(time.time()-ti)\n",
    "\n",
    "data2 = mmread(f_data2)\n",
    "gene2 = pd.read_csv(f_data2_gene)['x'].values\n",
    "cell2 = pd.read_csv(f_data2_cell)['x'].values\n",
    "print(time.time()-ti)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# assert np.all(gene==gene2)\n",
    "# gene.shape\n",
    "# gene2.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(27574,) (27495,) (24809,) (24809,)\n",
      "(24809,) (24809,)\n",
      "(24809, 215823)\n"
     ]
    }
   ],
   "source": [
    "# autosome genes\n",
    "chromosomes = ['chr{}'.format(chrom) for chrom in snmcseq_utils.get_mouse_chromosomes(include_x=False)]\n",
    "gene_table1 = df_genes_v2[df_genes_v2['chr'].isin(chromosomes)].reindex(gene).dropna()\n",
    "gene_table2 = df_genes_v2[df_genes_v2['chr'].isin(chromosomes)].reindex(gene2).dropna()\n",
    "\n",
    "genes_common = np.intersect1d(gene_table1.index.values, gene_table2.index.values)\n",
    "genes_common_ids = gene_table1.loc[genes_common, 'gene_id'].values \n",
    "print(gene.shape, \n",
    "      gene2.shape, \n",
    "      genes_common.shape, \n",
    "      genes_common_ids.shape, \n",
    "     )\n",
    "\n",
    "selected_genes_idx1 = snmcseq_utils.get_index_from_array(gene, genes_common)\n",
    "selected_genes_idx2 = snmcseq_utils.get_index_from_array(gene2, genes_common)\n",
    "print(selected_genes_idx1.shape, \n",
    "      selected_genes_idx2.shape, \n",
    "     )\n",
    "\n",
    "# \n",
    "data_comb = sparse.hstack([\n",
    "                data.tocsr()[selected_genes_idx1,:], data2.tocsr()[selected_genes_idx2,:]\n",
    "            ])\n",
    "print(data_comb.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "990.3496780395508\n"
     ]
    }
   ],
   "source": [
    "# output\n",
    "# save \n",
    "ti = time.time()\n",
    "gc_mat = GC_matrix(genes_common_ids, \n",
    "                   np.hstack([cell, cell2]), \n",
    "                   data_comb, \n",
    "                  )\n",
    "snmcseq_utils.save_gc_matrix(gc_mat, fout_data_gene, fout_data_cell, fout_data)\n",
    "print(time.time()-ti)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(215823, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nUMI</th>\n",
       "      <th>nGene</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pBICCNsMMrMOpRMiM003d190318_AAACCCACACAAGCTT</th>\n",
       "      <td>4859</td>\n",
       "      <td>2336</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pBICCNsMMrMOpRMiM003d190318_AAACCCACAGTCCCGA</th>\n",
       "      <td>24082</td>\n",
       "      <td>6164</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pBICCNsMMrMOpRMiM003d190318_AAACCCAGTGAATTAG</th>\n",
       "      <td>5194</td>\n",
       "      <td>2519</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pBICCNsMMrMOpRMiM003d190318_AAACCCAGTGGTAACG</th>\n",
       "      <td>1343</td>\n",
       "      <td>998</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pBICCNsMMrMOpRMiM003d190318_AAACCCAGTTCCAAAC</th>\n",
       "      <td>9022</td>\n",
       "      <td>3558</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               nUMI  nGene   sex\n",
       "pBICCNsMMrMOpRMiM003d190318_AAACCCACACAAGCTT   4859   2336  MALE\n",
       "pBICCNsMMrMOpRMiM003d190318_AAACCCACAGTCCCGA  24082   6164  MALE\n",
       "pBICCNsMMrMOpRMiM003d190318_AAACCCAGTGAATTAG   5194   2519  MALE\n",
       "pBICCNsMMrMOpRMiM003d190318_AAACCCAGTGGTAACG   1343    998  MALE\n",
       "pBICCNsMMrMOpRMiM003d190318_AAACCCAGTTCCAAAC   9022   3558  MALE"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # assemble metadata (keep order, add male and female)\n",
    "\n",
    "meta = pd.read_csv(f_meta, index_col=0).rename(columns={'dataset': 'sex'})[['nUMI', 'nGene', 'sex']]\n",
    "# meta.index.name = 'cell'\n",
    "# meta = meta.loc[np.hstack([cell, cell2])]\n",
    "print(meta.shape)\n",
    "meta.head()\n",
    "\n",
    "# # meta.to_csv(fout_meta, sep=\"\\t\", header=True, index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(159738, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>QC</th>\n",
       "      <th>cluster_label</th>\n",
       "      <th>cluster_color</th>\n",
       "      <th>class_label</th>\n",
       "      <th>subclass_label</th>\n",
       "      <th>size</th>\n",
       "      <th>gene.counts</th>\n",
       "      <th>umi.counts</th>\n",
       "      <th>Broad.QC.doublet</th>\n",
       "      <th>Broad.QC.Mito</th>\n",
       "      <th>Broad.passQC</th>\n",
       "      <th>MALE</th>\n",
       "      <th>Comb.QC</th>\n",
       "      <th>cl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pBICCNsMMrMOpRMiM003d190318_AAACCCACACAAGCTT</td>\n",
       "      <td>90</td>\n",
       "      <td>1</td>\n",
       "      <td>Oligo Opalin_4</td>\n",
       "      <td>#474662</td>\n",
       "      <td>Non-Neuronal</td>\n",
       "      <td>Oligo</td>\n",
       "      <td>16566</td>\n",
       "      <td>1740.348364</td>\n",
       "      <td>3419.937704</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.487323</td>\n",
       "      <td>NaN</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pBICCNsMMrMOpRMiM003d190318_AAAGGATCAGACCGCT</td>\n",
       "      <td>90</td>\n",
       "      <td>19</td>\n",
       "      <td>Oligo Opalin_4</td>\n",
       "      <td>#474662</td>\n",
       "      <td>Non-Neuronal</td>\n",
       "      <td>Oligo</td>\n",
       "      <td>16566</td>\n",
       "      <td>1740.348364</td>\n",
       "      <td>3419.937704</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.487323</td>\n",
       "      <td>NaN</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pBICCNsMMrMOpRMiM003d190318_AAAGGATCATGGAACG</td>\n",
       "      <td>90</td>\n",
       "      <td>20</td>\n",
       "      <td>Oligo Opalin_4</td>\n",
       "      <td>#474662</td>\n",
       "      <td>Non-Neuronal</td>\n",
       "      <td>Oligo</td>\n",
       "      <td>16566</td>\n",
       "      <td>1740.348364</td>\n",
       "      <td>3419.937704</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.487323</td>\n",
       "      <td>NaN</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pBICCNsMMrMOpRMiM003d190318_AAAGGTAGTAATACCC</td>\n",
       "      <td>90</td>\n",
       "      <td>37</td>\n",
       "      <td>Oligo Opalin_4</td>\n",
       "      <td>#474662</td>\n",
       "      <td>Non-Neuronal</td>\n",
       "      <td>Oligo</td>\n",
       "      <td>16566</td>\n",
       "      <td>1740.348364</td>\n",
       "      <td>3419.937704</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.487323</td>\n",
       "      <td>NaN</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pBICCNsMMrMOpRMiM003d190318_AAAGTCCCAAGGTCAG</td>\n",
       "      <td>90</td>\n",
       "      <td>39</td>\n",
       "      <td>Oligo Opalin_4</td>\n",
       "      <td>#474662</td>\n",
       "      <td>Non-Neuronal</td>\n",
       "      <td>Oligo</td>\n",
       "      <td>16566</td>\n",
       "      <td>1740.348364</td>\n",
       "      <td>3419.937704</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.487323</td>\n",
       "      <td>NaN</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           cell  cluster_id  QC  \\\n",
       "0  pBICCNsMMrMOpRMiM003d190318_AAACCCACACAAGCTT          90   1   \n",
       "1  pBICCNsMMrMOpRMiM003d190318_AAAGGATCAGACCGCT          90  19   \n",
       "2  pBICCNsMMrMOpRMiM003d190318_AAAGGATCATGGAACG          90  20   \n",
       "3  pBICCNsMMrMOpRMiM003d190318_AAAGGTAGTAATACCC          90  37   \n",
       "4  pBICCNsMMrMOpRMiM003d190318_AAAGTCCCAAGGTCAG          90  39   \n",
       "\n",
       "    cluster_label cluster_color   class_label subclass_label   size  \\\n",
       "0  Oligo Opalin_4       #474662  Non-Neuronal          Oligo  16566   \n",
       "1  Oligo Opalin_4       #474662  Non-Neuronal          Oligo  16566   \n",
       "2  Oligo Opalin_4       #474662  Non-Neuronal          Oligo  16566   \n",
       "3  Oligo Opalin_4       #474662  Non-Neuronal          Oligo  16566   \n",
       "4  Oligo Opalin_4       #474662  Non-Neuronal          Oligo  16566   \n",
       "\n",
       "   gene.counts   umi.counts  Broad.QC.doublet  Broad.QC.Mito  Broad.passQC  \\\n",
       "0  1740.348364  3419.937704              0.06            0.0          0.94   \n",
       "1  1740.348364  3419.937704              0.06            0.0          0.94   \n",
       "2  1740.348364  3419.937704              0.06            0.0          0.94   \n",
       "3  1740.348364  3419.937704              0.06            0.0          0.94   \n",
       "4  1740.348364  3419.937704              0.06            0.0          0.94   \n",
       "\n",
       "       MALE Comb.QC   cl  \n",
       "0  0.487323     NaN  150  \n",
       "1  0.487323     NaN  150  \n",
       "2  0.487323     NaN  150  \n",
       "3  0.487323     NaN  150  \n",
       "4  0.487323     NaN  150  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "meta1 = pd.read_csv(f_meta1, names=['cell', 'cluster_id'], header=0) # cluster assignment\n",
    "meta2 = pd.read_csv(f_meta2) # cluster annot [all metrics]\n",
    "meta3 = pd.read_csv(f_meta3, names=['QC', 'cell'], header=0) # QC'ed cells\n",
    "meta1 = pd.merge(meta1, meta3, on='cell') # cluster id on QC'ed cells\n",
    "meta1 = pd.merge(meta1, meta2, on='cluster_id')\n",
    "meta1 = pd.merge(meta1, meta, left_on='cell', right_index=True)\n",
    "\n",
    "print(meta1.shape)\n",
    "meta1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta1.to_csv(fout_meta, sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 10x_nuclei_v3\n",
    "\n",
    "- gene_ids\n",
    "- all cells passed QC (QC.csv) metadata\n",
    "\n",
    "- all cells data\n",
    "- output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze/10x_nuclei_v3/cluster.membership.csv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/10x_nuclei_v3_metadata.tsv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze/10x_nuclei_v3/matrix.mtx.gz\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/10x_nuclei_v3_counts.npz\n"
     ]
    }
   ],
   "source": [
    "dataset_name = '10x_nuclei_v3'\n",
    "\n",
    "f_meta = os.path.join(SRC_DIR, dataset_name, 'cluster.membership.csv') \n",
    "f_meta2 = os.path.join(SRC_DIR, dataset_name, 'cluster.annotation.csv') \n",
    "f_meta3 = os.path.join(SRC_DIR, dataset_name, 'QC.csv') \n",
    "\n",
    "f_data = os.path.join(SRC_DIR, dataset_name, 'matrix.mtx.gz') \n",
    "f_data_gene = os.path.join(SRC_DIR, dataset_name, 'features.tsv.gz') \n",
    "f_data_cell = os.path.join(SRC_DIR, dataset_name, 'barcodes.tsv.gz') \n",
    "\n",
    "fout_meta = os.path.join(DST_DIR, dataset_name+'_metadata.tsv') \n",
    "fout_data = os.path.join(DST_DIR, dataset_name+'_counts.npz') \n",
    "fout_data_gene = os.path.join(DST_DIR, dataset_name+'_counts.gene') \n",
    "fout_data_cell = os.path.join(DST_DIR, dataset_name+'_counts.cell') \n",
    "\n",
    "print(f_meta)\n",
    "print(fout_meta)\n",
    "print(f_data)\n",
    "print(fout_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(40166, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>QC</th>\n",
       "      <th>cluster_label</th>\n",
       "      <th>subclass_label</th>\n",
       "      <th>class_label</th>\n",
       "      <th>cluster_color</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAAGCTCTTCC-1L8TX_181211_01_A02</td>\n",
       "      <td>59</td>\n",
       "      <td>1</td>\n",
       "      <td>L6b Shisa6</td>\n",
       "      <td>L6b</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#2B9880</td>\n",
       "      <td>247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAGGGCTCGCGATCG-1L8TX_181211_01_A02</td>\n",
       "      <td>59</td>\n",
       "      <td>47</td>\n",
       "      <td>L6b Shisa6</td>\n",
       "      <td>L6b</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#2B9880</td>\n",
       "      <td>247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAGTGACATCGCTGG-1L8TX_181211_01_A02</td>\n",
       "      <td>59</td>\n",
       "      <td>80</td>\n",
       "      <td>L6b Shisa6</td>\n",
       "      <td>L6b</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#2B9880</td>\n",
       "      <td>247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AACAACCGTGCTCTCT-1L8TX_181211_01_A02</td>\n",
       "      <td>59</td>\n",
       "      <td>108</td>\n",
       "      <td>L6b Shisa6</td>\n",
       "      <td>L6b</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#2B9880</td>\n",
       "      <td>247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AACAAGAAGCCTGCCA-1L8TX_181211_01_A02</td>\n",
       "      <td>59</td>\n",
       "      <td>112</td>\n",
       "      <td>L6b Shisa6</td>\n",
       "      <td>L6b</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#2B9880</td>\n",
       "      <td>247</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   cell  cluster_id   QC cluster_label  \\\n",
       "0  AAACCCAAGCTCTTCC-1L8TX_181211_01_A02          59    1    L6b Shisa6   \n",
       "1  AAAGGGCTCGCGATCG-1L8TX_181211_01_A02          59   47    L6b Shisa6   \n",
       "2  AAAGTGACATCGCTGG-1L8TX_181211_01_A02          59   80    L6b Shisa6   \n",
       "3  AACAACCGTGCTCTCT-1L8TX_181211_01_A02          59  108    L6b Shisa6   \n",
       "4  AACAAGAAGCCTGCCA-1L8TX_181211_01_A02          59  112    L6b Shisa6   \n",
       "\n",
       "  subclass_label    class_label cluster_color  size  \n",
       "0            L6b  Glutamatergic       #2B9880   247  \n",
       "1            L6b  Glutamatergic       #2B9880   247  \n",
       "2            L6b  Glutamatergic       #2B9880   247  \n",
       "3            L6b  Glutamatergic       #2B9880   247  \n",
       "4            L6b  Glutamatergic       #2B9880   247  "
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv(f_meta, names=['cell', 'cluster_id'], header=0) \n",
    "meta2 = pd.read_csv(f_meta2)\n",
    "meta3 = pd.read_csv(f_meta3, names=['QC', 'cell'], header=0)\n",
    "meta = pd.merge(meta, meta3, on='cell')\n",
    "meta = pd.merge(meta, meta2, on='cluster_id')\n",
    "\n",
    "print(meta.shape)\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(90266,)\n",
      "(31053,)\n",
      "(31053, 90266)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['ENSMUSG00000051951', 'ENSMUSG00000089699', 'ENSMUSG00000102343',\n",
       "       ..., 'ENSMUSG00000063897', 'ENSMUSG00000096730',\n",
       "       'ENSMUSG00000095742'], dtype=object)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_cell = snmcseq_utils.import_single_textcol(f_data_cell)\n",
    "data_gene = snmcseq_utils.import_single_textcol(f_data_gene, col=0)\n",
    "data = mmread(f_data)\n",
    "print(data_cell.shape)\n",
    "print(data_gene.shape)\n",
    "print(data.shape)\n",
    "data_cell\n",
    "data_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "gc_mat = GC_matrix(data_gene, data_cell, data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta.to_csv(fout_meta, sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "snmcseq_utils.save_gc_matrix(gc_mat, \n",
    "                             fout_data_gene, fout_data_cell, fout_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 10x_cells_v3 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze/10x_cells_v3/cluster.membership.csv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/10x_cells_v3_metadata.tsv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze/10x_cells_v3/matrix.mtx.gz\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/10x_cells_v3_counts.npz\n"
     ]
    }
   ],
   "source": [
    "dataset_name = '10x_cells_v3'\n",
    "\n",
    "f_meta = os.path.join(SRC_DIR, dataset_name, 'cluster.membership.csv') \n",
    "f_meta2 = os.path.join(SRC_DIR, dataset_name, 'cluster.annotation.csv') \n",
    "f_meta3 = os.path.join(SRC_DIR, dataset_name, 'QC.csv') \n",
    "\n",
    "f_data = os.path.join(SRC_DIR, dataset_name, 'matrix.mtx.gz') \n",
    "f_data_gene = os.path.join(SRC_DIR, dataset_name, 'features.tsv.gz') \n",
    "f_data_cell = os.path.join(SRC_DIR, dataset_name, 'barcodes.tsv.gz') \n",
    "\n",
    "fout_meta = os.path.join(DST_DIR, dataset_name+'_metadata.tsv') \n",
    "fout_data = os.path.join(DST_DIR, dataset_name+'_counts.npz') \n",
    "fout_data_gene = os.path.join(DST_DIR, dataset_name+'_counts.gene') \n",
    "fout_data_cell = os.path.join(DST_DIR, dataset_name+'_counts.cell') \n",
    "\n",
    "print(f_meta)\n",
    "print(fout_meta)\n",
    "print(f_data)\n",
    "print(fout_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(71183, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>QC</th>\n",
       "      <th>cluster_label</th>\n",
       "      <th>subclass_label</th>\n",
       "      <th>class_label</th>\n",
       "      <th>cluster_color</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCCAAGCTTCATG-1L8TX_181211_01_G12</td>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>L5 IT Tcap_2</td>\n",
       "      <td>L5 IT</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#52CA74</td>\n",
       "      <td>17334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCCACACCAGCCA-1L8TX_181211_01_G12</td>\n",
       "      <td>42</td>\n",
       "      <td>3</td>\n",
       "      <td>L5 IT Tcap_2</td>\n",
       "      <td>L5 IT</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#52CA74</td>\n",
       "      <td>17334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACGAACAACGATTC-1L8TX_181211_01_G12</td>\n",
       "      <td>42</td>\n",
       "      <td>8</td>\n",
       "      <td>L5 IT Tcap_2</td>\n",
       "      <td>L5 IT</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#52CA74</td>\n",
       "      <td>17334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACGAATCTCGTGAA-1L8TX_181211_01_G12</td>\n",
       "      <td>42</td>\n",
       "      <td>11</td>\n",
       "      <td>L5 IT Tcap_2</td>\n",
       "      <td>L5 IT</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#52CA74</td>\n",
       "      <td>17334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACGCTGTAGTCACT-1L8TX_181211_01_G12</td>\n",
       "      <td>42</td>\n",
       "      <td>14</td>\n",
       "      <td>L5 IT Tcap_2</td>\n",
       "      <td>L5 IT</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#52CA74</td>\n",
       "      <td>17334</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   cell  cluster_id  QC cluster_label  \\\n",
       "0  AAACCCAAGCTTCATG-1L8TX_181211_01_G12          42   1  L5 IT Tcap_2   \n",
       "1  AAACCCACACCAGCCA-1L8TX_181211_01_G12          42   3  L5 IT Tcap_2   \n",
       "2  AAACGAACAACGATTC-1L8TX_181211_01_G12          42   8  L5 IT Tcap_2   \n",
       "3  AAACGAATCTCGTGAA-1L8TX_181211_01_G12          42  11  L5 IT Tcap_2   \n",
       "4  AAACGCTGTAGTCACT-1L8TX_181211_01_G12          42  14  L5 IT Tcap_2   \n",
       "\n",
       "  subclass_label    class_label cluster_color   size  \n",
       "0          L5 IT  Glutamatergic       #52CA74  17334  \n",
       "1          L5 IT  Glutamatergic       #52CA74  17334  \n",
       "2          L5 IT  Glutamatergic       #52CA74  17334  \n",
       "3          L5 IT  Glutamatergic       #52CA74  17334  \n",
       "4          L5 IT  Glutamatergic       #52CA74  17334  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv(f_meta, names=['cell', 'cluster_id'], header=0) \n",
    "meta2 = pd.read_csv(f_meta2)\n",
    "meta3 = pd.read_csv(f_meta3, names=['QC', 'cell'], header=0)\n",
    "meta = pd.merge(meta, meta3, on='cell')\n",
    "meta = pd.merge(meta, meta2, on='cluster_id')\n",
    "\n",
    "print(meta.shape)\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(176584,)\n",
      "(31053,)\n",
      "(31053, 176584)\n"
     ]
    }
   ],
   "source": [
    "data_cell = snmcseq_utils.import_single_textcol(f_data_cell)\n",
    "data_gene = snmcseq_utils.import_single_textcol(f_data_gene, col=0)\n",
    "data = mmread(f_data)\n",
    "print(data_cell.shape)\n",
    "print(data_gene.shape)\n",
    "print(data.shape)\n",
    "data_cell\n",
    "data_gene\n",
    "\n",
    "gc_mat = GC_matrix(data_gene, data_cell, data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta.to_csv(fout_meta, sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "snmcseq_utils.save_gc_matrix(gc_mat, \n",
    "                             fout_data_gene, fout_data_cell, fout_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 10x_cells_v2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze/10x_cells_v2/cluster.membership.csv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/10x_cells_v2_metadata.tsv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze/10x_cells_v2/matrix.mtx.gz\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/10x_cells_v2_counts.npz\n"
     ]
    }
   ],
   "source": [
    "dataset_name = '10x_cells_v2'\n",
    "\n",
    "f_meta = os.path.join(SRC_DIR, dataset_name, 'cluster.membership.csv') \n",
    "f_meta2 = os.path.join(SRC_DIR, dataset_name, 'cluster.annotation.csv') \n",
    "f_meta3 = os.path.join(SRC_DIR, dataset_name, 'QC.csv') \n",
    "\n",
    "f_data = os.path.join(SRC_DIR, dataset_name, 'matrix.mtx.gz') \n",
    "f_data_gene = os.path.join(SRC_DIR, dataset_name, 'features.tsv.gz') \n",
    "f_data_cell = os.path.join(SRC_DIR, dataset_name, 'barcodes.tsv.gz') \n",
    "\n",
    "fout_meta = os.path.join(DST_DIR, dataset_name+'_metadata.tsv') \n",
    "fout_data = os.path.join(DST_DIR, dataset_name+'_counts.npz') \n",
    "fout_data_gene = os.path.join(DST_DIR, dataset_name+'_counts.gene') \n",
    "fout_data_cell = os.path.join(DST_DIR, dataset_name+'_counts.cell') \n",
    "\n",
    "print(f_meta)\n",
    "print(fout_meta)\n",
    "print(f_data)\n",
    "print(fout_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(122641, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>QC</th>\n",
       "      <th>cluster_label</th>\n",
       "      <th>subclass_label</th>\n",
       "      <th>class_label</th>\n",
       "      <th>cluster_color</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGAGGAGTCTG-1L8TX_171026_01_F03</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Sncg Slc17a8</td>\n",
       "      <td>Sncg</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#9440F3</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACCTTTAGTACAGCAG-1L8TX_171026_01_F03</td>\n",
       "      <td>9</td>\n",
       "      <td>673</td>\n",
       "      <td>Sncg Slc17a8</td>\n",
       "      <td>Sncg</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#9440F3</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACGATACCACCCAGTG-1L8TX_171026_01_F03</td>\n",
       "      <td>9</td>\n",
       "      <td>724</td>\n",
       "      <td>Sncg Slc17a8</td>\n",
       "      <td>Sncg</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#9440F3</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACGGGTCAGTGGGCTA-1L8TX_171026_01_F03</td>\n",
       "      <td>9</td>\n",
       "      <td>885</td>\n",
       "      <td>Sncg Slc17a8</td>\n",
       "      <td>Sncg</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#9440F3</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGCAGCCCAGTTCATG-1L8TX_171026_01_F03</td>\n",
       "      <td>9</td>\n",
       "      <td>1205</td>\n",
       "      <td>Sncg Slc17a8</td>\n",
       "      <td>Sncg</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#9440F3</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   cell  cluster_id    QC cluster_label  \\\n",
       "0  AAACCTGAGGAGTCTG-1L8TX_171026_01_F03           9     1  Sncg Slc17a8   \n",
       "1  ACCTTTAGTACAGCAG-1L8TX_171026_01_F03           9   673  Sncg Slc17a8   \n",
       "2  ACGATACCACCCAGTG-1L8TX_171026_01_F03           9   724  Sncg Slc17a8   \n",
       "3  ACGGGTCAGTGGGCTA-1L8TX_171026_01_F03           9   885  Sncg Slc17a8   \n",
       "4  AGCAGCCCAGTTCATG-1L8TX_171026_01_F03           9  1205  Sncg Slc17a8   \n",
       "\n",
       "  subclass_label class_label cluster_color  size  \n",
       "0           Sncg   GABAergic       #9440F3   293  \n",
       "1           Sncg   GABAergic       #9440F3   293  \n",
       "2           Sncg   GABAergic       #9440F3   293  \n",
       "3           Sncg   GABAergic       #9440F3   293  \n",
       "4           Sncg   GABAergic       #9440F3   293  "
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv(f_meta, names=['cell', 'cluster_id'], header=0) \n",
    "meta2 = pd.read_csv(f_meta2)\n",
    "meta3 = pd.read_csv(f_meta3, names=['QC', 'cell'], header=0)\n",
    "meta = pd.merge(meta, meta3, on='cell')\n",
    "meta = pd.merge(meta, meta2, on='cluster_id')\n",
    "\n",
    "print(meta.shape)\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(145748,)\n",
      "(31053,)\n",
      "(31053, 145748)\n"
     ]
    }
   ],
   "source": [
    "data_cell = snmcseq_utils.import_single_textcol(f_data_cell)\n",
    "data_gene = snmcseq_utils.import_single_textcol(f_data_gene, col=0)\n",
    "data = mmread(f_data)\n",
    "print(data_cell.shape)\n",
    "print(data_gene.shape)\n",
    "print(data.shape)\n",
    "data_cell\n",
    "data_gene\n",
    "\n",
    "gc_mat = GC_matrix(data_gene, data_cell, data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta.to_csv(fout_meta, sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "snmcseq_utils.save_gc_matrix(gc_mat, \n",
    "                             fout_data_gene, fout_data_cell, fout_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Smart-cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze/smarter_cells/cluster.membership.csv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/smarter_cells_metadata.tsv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/smarter_cells_raw.npz\n"
     ]
    }
   ],
   "source": [
    "dataset_name = 'smarter_cells'\n",
    "\n",
    "f_meta = os.path.join(SRC_DIR, dataset_name, 'cluster.membership.csv') \n",
    "f_meta2 = os.path.join(SRC_DIR, dataset_name, 'cluster.annotation.csv') \n",
    "f_meta3 = os.path.join(SRC_DIR, dataset_name, 'QC.csv') \n",
    "f_meta4 = os.path.join(SRC_DIR, dataset_name, 'sample_metadata.csv.gz') \n",
    "\n",
    "f_data_intron = os.path.join(SRC_DIR, dataset_name, 'intron.counts.csv.gz') \n",
    "f_data_exon = os.path.join(SRC_DIR, dataset_name, 'exon.counts.csv.gz') \n",
    "\n",
    "fout_meta = os.path.join(DST_DIR, dataset_name+'_metadata.tsv') \n",
    "fout_data = os.path.join(DST_DIR, dataset_name+'_raw.npz') \n",
    "fout_data_gene = os.path.join(DST_DIR, dataset_name+'_raw.gene') \n",
    "fout_data_cell = os.path.join(DST_DIR, dataset_name+'_raw.cell') \n",
    "\n",
    "print(f_meta)\n",
    "print(fout_meta)\n",
    "# print(f_data)\n",
    "print(fout_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6288, 129)\n",
      "(6288, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>QC</th>\n",
       "      <th>cluster_label</th>\n",
       "      <th>subclass_label</th>\n",
       "      <th>class_label</th>\n",
       "      <th>cluster_color</th>\n",
       "      <th>size</th>\n",
       "      <th>ar_id</th>\n",
       "      <th>exp_component_vendor_name</th>\n",
       "      <th>...</th>\n",
       "      <th>Injection_type</th>\n",
       "      <th>NewBatch</th>\n",
       "      <th>Vendor</th>\n",
       "      <th>MultiPlex</th>\n",
       "      <th>percent_reads_aligned_to_rrna_rmsk</th>\n",
       "      <th>percent_reads_aligned_to_trna_rmsk</th>\n",
       "      <th>percent_reads_aligned_to_ncrna</th>\n",
       "      <th>percent_reads_aligned_to_introns</th>\n",
       "      <th>percent_reads_aligned_to_intergenic</th>\n",
       "      <th>percent_reads_aligned_unique</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LS-15395_S41_E1-50</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>Sst Myh8_1</td>\n",
       "      <td>Sst</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#FF7F2C</td>\n",
       "      <td>136</td>\n",
       "      <td>543183476</td>\n",
       "      <td>LS-15395-41</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HistoricalData</td>\n",
       "      <td>Covance</td>\n",
       "      <td>index96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.162980</td>\n",
       "      <td>9.855384</td>\n",
       "      <td>3.264771</td>\n",
       "      <td>84.993663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LS-15395_S42_E1-50</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>Sst Myh8_1</td>\n",
       "      <td>Sst</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#FF7F2C</td>\n",
       "      <td>136</td>\n",
       "      <td>543183474</td>\n",
       "      <td>LS-15395-42</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HistoricalData</td>\n",
       "      <td>Covance</td>\n",
       "      <td>index96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16.281620</td>\n",
       "      <td>6.763949</td>\n",
       "      <td>2.856108</td>\n",
       "      <td>87.252036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LS-15501_S28_E1-50</td>\n",
       "      <td>19</td>\n",
       "      <td>94</td>\n",
       "      <td>Sst Myh8_1</td>\n",
       "      <td>Sst</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#FF7F2C</td>\n",
       "      <td>136</td>\n",
       "      <td>544930823</td>\n",
       "      <td>LS-15501-28</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HistoricalData</td>\n",
       "      <td>Covance</td>\n",
       "      <td>index96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17.066841</td>\n",
       "      <td>3.311346</td>\n",
       "      <td>2.437971</td>\n",
       "      <td>86.831969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LS-15501_S81_E1-50</td>\n",
       "      <td>19</td>\n",
       "      <td>120</td>\n",
       "      <td>Sst Myh8_1</td>\n",
       "      <td>Sst</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#FF7F2C</td>\n",
       "      <td>136</td>\n",
       "      <td>544930707</td>\n",
       "      <td>LS-15501-81</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HistoricalData</td>\n",
       "      <td>Covance</td>\n",
       "      <td>index96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.866006</td>\n",
       "      <td>11.769028</td>\n",
       "      <td>3.699682</td>\n",
       "      <td>85.252590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LS-15524_S21_E1-50</td>\n",
       "      <td>19</td>\n",
       "      <td>267</td>\n",
       "      <td>Sst Myh8_1</td>\n",
       "      <td>Sst</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#FF7F2C</td>\n",
       "      <td>136</td>\n",
       "      <td>549999237</td>\n",
       "      <td>LS-15524-21</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HistoricalData</td>\n",
       "      <td>Covance</td>\n",
       "      <td>index96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.912637</td>\n",
       "      <td>5.923318</td>\n",
       "      <td>2.385434</td>\n",
       "      <td>82.503705</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  129 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 cell  cluster_id   QC cluster_label subclass_label  \\\n",
       "0  LS-15395_S41_E1-50          19    1    Sst Myh8_1            Sst   \n",
       "1  LS-15395_S42_E1-50          19    2    Sst Myh8_1            Sst   \n",
       "2  LS-15501_S28_E1-50          19   94    Sst Myh8_1            Sst   \n",
       "3  LS-15501_S81_E1-50          19  120    Sst Myh8_1            Sst   \n",
       "4  LS-15524_S21_E1-50          19  267    Sst Myh8_1            Sst   \n",
       "\n",
       "  class_label cluster_color  size      ar_id exp_component_vendor_name  ...  \\\n",
       "0   GABAergic       #FF7F2C   136  543183476               LS-15395-41  ...   \n",
       "1   GABAergic       #FF7F2C   136  543183474               LS-15395-42  ...   \n",
       "2   GABAergic       #FF7F2C   136  544930823               LS-15501-28  ...   \n",
       "3   GABAergic       #FF7F2C   136  544930707               LS-15501-81  ...   \n",
       "4   GABAergic       #FF7F2C   136  549999237               LS-15524-21  ...   \n",
       "\n",
       "  Injection_type        NewBatch   Vendor MultiPlex  \\\n",
       "0            0.0  HistoricalData  Covance   index96   \n",
       "1            0.0  HistoricalData  Covance   index96   \n",
       "2            0.0  HistoricalData  Covance   index96   \n",
       "3            0.0  HistoricalData  Covance   index96   \n",
       "4            0.0  HistoricalData  Covance   index96   \n",
       "\n",
       "   percent_reads_aligned_to_rrna_rmsk  percent_reads_aligned_to_trna_rmsk  \\\n",
       "0                                 NaN                                 NaN   \n",
       "1                                 NaN                                 NaN   \n",
       "2                                 NaN                                 NaN   \n",
       "3                                 NaN                                 NaN   \n",
       "4                                 NaN                                 NaN   \n",
       "\n",
       "   percent_reads_aligned_to_ncrna  percent_reads_aligned_to_introns  \\\n",
       "0                       15.162980                          9.855384   \n",
       "1                       16.281620                          6.763949   \n",
       "2                       17.066841                          3.311346   \n",
       "3                       13.866006                         11.769028   \n",
       "4                       15.912637                          5.923318   \n",
       "\n",
       "  percent_reads_aligned_to_intergenic percent_reads_aligned_unique  \n",
       "0                            3.264771                    84.993663  \n",
       "1                            2.856108                    87.252036  \n",
       "2                            2.437971                    86.831969  \n",
       "3                            3.699682                    85.252590  \n",
       "4                            2.385434                    82.503705  \n",
       "\n",
       "[5 rows x 129 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv(f_meta, names=['cell', 'cluster_id'], header=0) \n",
    "meta2 = pd.read_csv(f_meta2)\n",
    "meta3 = pd.read_csv(f_meta3, names=['QC', 'cell'], header=0)\n",
    "meta4 = pd.read_csv(f_meta4, index_col=0)\n",
    "meta4.index.name = 'cell' \n",
    "meta4 = meta4.reset_index()\n",
    "\n",
    "meta = pd.merge(meta, meta3, on='cell')\n",
    "meta = pd.merge(meta, meta2, on='cluster_id')\n",
    "meta = pd.merge(meta, meta4, on='cell')\n",
    "\n",
    "print(meta.shape)\n",
    "print(meta3.shape)\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(32324,) (32324,)\n",
      "(45768, 6300)\n"
     ]
    }
   ],
   "source": [
    "data_intron = pd.read_csv(f_data_intron, index_col=0) \n",
    "data_exon = pd.read_csv(f_data_exon, index_col=0) \n",
    "\n",
    "data = data_intron + data_exon\n",
    "gene = data.index.values\n",
    "cell = data.columns.values\n",
    "\n",
    "gene_table = df_genes_v2.reindex(gene).dropna()\n",
    "gene_ids = gene_table['gene_id'].values\n",
    "gene_names = gene_table.index.values\n",
    "\n",
    "print(gene_ids.shape, gene_names.shape)\n",
    "print(data.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_sparse = sparse.coo_matrix(data.loc[gene_names])\n",
    "gene = gene_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "gc_mat = GC_matrix(gene, cell, data_sparse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta.to_csv(fout_meta, sep=\"\\t\", header=True, index=False)\n",
    "snmcseq_utils.save_gc_matrix(gc_mat, \n",
    "                             fout_data_gene, fout_data_cell, fout_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Smarter-nuclei"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze/smarter_nuclei/cluster.membership.csv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/smarter_nuclei_metadata.tsv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/smarter_nuclei_raw.npz\n"
     ]
    }
   ],
   "source": [
    "dataset_name = 'smarter_nuclei'\n",
    "\n",
    "f_meta = os.path.join(SRC_DIR, dataset_name, 'cluster.membership.csv') \n",
    "f_meta2 = os.path.join(SRC_DIR, dataset_name, 'cluster.annotation.csv') \n",
    "f_meta3 = os.path.join(SRC_DIR, dataset_name, 'QC.csv') \n",
    "f_meta4 = os.path.join(SRC_DIR, dataset_name, 'sample_metadata.csv.gz') \n",
    "\n",
    "f_data_intron = os.path.join(SRC_DIR, dataset_name, 'intron.counts.csv.gz') \n",
    "f_data_exon = os.path.join(SRC_DIR, dataset_name, 'exon.counts.csv.gz') \n",
    "\n",
    "fout_meta = os.path.join(DST_DIR, dataset_name+'_metadata.tsv') \n",
    "fout_data = os.path.join(DST_DIR, dataset_name+'_raw.npz') \n",
    "fout_data_gene = os.path.join(DST_DIR, dataset_name+'_raw.gene') \n",
    "fout_data_cell = os.path.join(DST_DIR, dataset_name+'_raw.cell') \n",
    "\n",
    "print(f_meta)\n",
    "print(fout_meta)\n",
    "print(fout_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6171, 146)\n",
      "(6171, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cndd/fangming/venvs/conda_dobro/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2728: DtypeWarning: Columns (61) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>QC</th>\n",
       "      <th>cluster_label</th>\n",
       "      <th>subclass_label</th>\n",
       "      <th>class_label</th>\n",
       "      <th>cluster_color</th>\n",
       "      <th>size</th>\n",
       "      <th>ar_id</th>\n",
       "      <th>exp_component_vendor_name</th>\n",
       "      <th>...</th>\n",
       "      <th>pred.cl.2</th>\n",
       "      <th>pred.score.2</th>\n",
       "      <th>ref.cl.2</th>\n",
       "      <th>ref.cl.1.1</th>\n",
       "      <th>pred.cl.3</th>\n",
       "      <th>pred.score.3</th>\n",
       "      <th>cluster_label.2</th>\n",
       "      <th>category_label</th>\n",
       "      <th>Region.1</th>\n",
       "      <th>gene.counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SM-GE653_S113_E1-50</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>Lamp5 Egln3_2</td>\n",
       "      <td>Lamp5</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#E69B73</td>\n",
       "      <td>87</td>\n",
       "      <td>647869537</td>\n",
       "      <td>SM-GE653-S113</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>0.917036</td>\n",
       "      <td>10</td>\n",
       "      <td>31</td>\n",
       "      <td>14</td>\n",
       "      <td>0.917036</td>\n",
       "      <td>Vip Htr1f</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>7792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SM-GE653_S116_E1-50</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>Lamp5 Egln3_2</td>\n",
       "      <td>Lamp5</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#E69B73</td>\n",
       "      <td>87</td>\n",
       "      <td>647869531</td>\n",
       "      <td>SM-GE653-S116</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>0.712671</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>0.712671</td>\n",
       "      <td>Lamp5 Pdlim5</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>8502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SM-GE653_S117_E1-50</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>Lamp5 Egln3_2</td>\n",
       "      <td>Lamp5</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#E69B73</td>\n",
       "      <td>87</td>\n",
       "      <td>647869529</td>\n",
       "      <td>SM-GE653-S117</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>0.899050</td>\n",
       "      <td>9</td>\n",
       "      <td>28</td>\n",
       "      <td>12</td>\n",
       "      <td>0.899050</td>\n",
       "      <td>Vip Chat_3</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>7021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SM-GE653_S120_E1-50</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>Lamp5 Egln3_2</td>\n",
       "      <td>Lamp5</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#E69B73</td>\n",
       "      <td>87</td>\n",
       "      <td>647869523</td>\n",
       "      <td>SM-GE653-S120</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>0.892925</td>\n",
       "      <td>9</td>\n",
       "      <td>28</td>\n",
       "      <td>11</td>\n",
       "      <td>0.892925</td>\n",
       "      <td>Vip Chat_2</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>7085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SM-GE653_S122_E1-50</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>Lamp5 Egln3_2</td>\n",
       "      <td>Lamp5</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#E69B73</td>\n",
       "      <td>87</td>\n",
       "      <td>647869519</td>\n",
       "      <td>SM-GE653-S122</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>0.822251</td>\n",
       "      <td>9</td>\n",
       "      <td>28</td>\n",
       "      <td>13</td>\n",
       "      <td>0.822251</td>\n",
       "      <td>Vip Lmo1</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>7516</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  146 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  cell  cluster_id  QC  cluster_label subclass_label  \\\n",
       "0  SM-GE653_S113_E1-50           2   1  Lamp5 Egln3_2          Lamp5   \n",
       "1  SM-GE653_S116_E1-50           2   4  Lamp5 Egln3_2          Lamp5   \n",
       "2  SM-GE653_S117_E1-50           2   5  Lamp5 Egln3_2          Lamp5   \n",
       "3  SM-GE653_S120_E1-50           2   8  Lamp5 Egln3_2          Lamp5   \n",
       "4  SM-GE653_S122_E1-50           2  10  Lamp5 Egln3_2          Lamp5   \n",
       "\n",
       "  class_label cluster_color  size      ar_id exp_component_vendor_name  ...  \\\n",
       "0   GABAergic       #E69B73    87  647869537             SM-GE653-S113  ...   \n",
       "1   GABAergic       #E69B73    87  647869531             SM-GE653-S116  ...   \n",
       "2   GABAergic       #E69B73    87  647869529             SM-GE653-S117  ...   \n",
       "3   GABAergic       #E69B73    87  647869523             SM-GE653-S120  ...   \n",
       "4   GABAergic       #E69B73    87  647869519             SM-GE653-S122  ...   \n",
       "\n",
       "  pred.cl.2 pred.score.2 ref.cl.2 ref.cl.1.1  pred.cl.3  pred.score.3  \\\n",
       "0        14     0.917036       10         31         14      0.917036   \n",
       "1         3     0.712671        2          7          3      0.712671   \n",
       "2        12     0.899050        9         28         12      0.899050   \n",
       "3        11     0.892925        9         28         11      0.892925   \n",
       "4        13     0.822251        9         28         13      0.822251   \n",
       "\n",
       "   cluster_label.2  category_label Region.1 gene.counts  \n",
       "0        Vip Htr1f       GABAergic   Cortex        7792  \n",
       "1     Lamp5 Pdlim5       GABAergic   Cortex        8502  \n",
       "2       Vip Chat_3       GABAergic   Cortex        7021  \n",
       "3       Vip Chat_2       GABAergic   Cortex        7085  \n",
       "4         Vip Lmo1       GABAergic   Cortex        7516  \n",
       "\n",
       "[5 rows x 146 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv(f_meta, names=['cell', 'cluster_id'], header=0) \n",
    "meta2 = pd.read_csv(f_meta2)\n",
    "meta3 = pd.read_csv(f_meta3, names=['QC', 'cell'], header=0)\n",
    "\n",
    "meta4 = pd.read_csv(f_meta4, index_col=0)\n",
    "meta4.index.name = 'cell' \n",
    "meta4 = meta4.reset_index()\n",
    "\n",
    "meta = pd.merge(meta, meta3, on='cell')\n",
    "meta = pd.merge(meta, meta2, on='cluster_id')\n",
    "meta = pd.merge(meta, meta4.drop(columns=['cluster_id', 'cluster_label', 'cluster_color', 'size']), \n",
    "                on='cell')\n",
    "\n",
    "print(meta.shape)\n",
    "print(meta3.shape)\n",
    "# meta.filter(regex='_x$', axis=1).head()\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(32324,) (32324,)\n",
      "(45768, 6278)\n"
     ]
    }
   ],
   "source": [
    "data_intron = pd.read_csv(f_data_intron, index_col=0) \n",
    "data_exon = pd.read_csv(f_data_exon, index_col=0) \n",
    "\n",
    "data = data_intron + data_exon\n",
    "cell = data.columns.values\n",
    "gene = data.index.values\n",
    "\n",
    "gene_table = df_genes_v2.reindex(gene).dropna()\n",
    "gene_ids = gene_table['gene_id'].values\n",
    "gene_names = gene_table.index.values\n",
    "\n",
    "print(gene_ids.shape, gene_names.shape)\n",
    "print(data.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_sparse = sparse.coo_matrix(data.loc[gene_names])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "gc_mat = GC_matrix(gene_ids, cell, data_sparse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta.to_csv(fout_meta, sep=\"\\t\", header=True, index=False)\n",
    "snmcseq_utils.save_gc_matrix(gc_mat, \n",
    "                             fout_data_gene, fout_data_cell, fout_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### mC\n",
    "\n",
    "- QC'ed cells\n",
    "- remove version on ensembl id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import xarray as xr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/snmcseq_gene_metadata.tsv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/snmcseq_gene_mCH_counts.npz\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/snmcseq_gene_CH_counts.npz\n"
     ]
    }
   ],
   "source": [
    "dataset_name = 'snmcseq_gene'\n",
    "\n",
    "fout_meta = os.path.join(DST_DIR, dataset_name+'_metadata.tsv') \n",
    "fout_data_mc = os.path.join(DST_DIR, dataset_name+'_mCH_counts.npz') \n",
    "fout_data_c = os.path.join(DST_DIR, dataset_name+'_CH_counts.npz') \n",
    "fout_data_gene = os.path.join(DST_DIR, dataset_name+'_counts.gene') \n",
    "fout_data_cell = os.path.join(DST_DIR, dataset_name+'_counts.cell') \n",
    "\n",
    "print(fout_meta)\n",
    "print(fout_data_mc)\n",
    "print(fout_data_c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9876, 32)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>FullCellID</th>\n",
       "      <th>CCC_Rate</th>\n",
       "      <th>CG_Rate</th>\n",
       "      <th>CG_RateAdj</th>\n",
       "      <th>CH_Rate</th>\n",
       "      <th>CH_RateAdj</th>\n",
       "      <th>FinalReads</th>\n",
       "      <th>InputReads</th>\n",
       "      <th>MappedReads</th>\n",
       "      <th>...</th>\n",
       "      <th>Slice</th>\n",
       "      <th>PassFilter</th>\n",
       "      <th>pca_0</th>\n",
       "      <th>pca_1</th>\n",
       "      <th>umap_0</th>\n",
       "      <th>umap_1</th>\n",
       "      <th>tsne_0</th>\n",
       "      <th>tsne_1</th>\n",
       "      <th>MajorCluster</th>\n",
       "      <th>SubCluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2C_M_0</td>\n",
       "      <td>180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...</td>\n",
       "      <td>0.00511</td>\n",
       "      <td>0.77073</td>\n",
       "      <td>0.76955</td>\n",
       "      <td>0.02154</td>\n",
       "      <td>0.01651</td>\n",
       "      <td>1076259.0</td>\n",
       "      <td>2314862</td>\n",
       "      <td>1517338.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>-11.588627</td>\n",
       "      <td>-0.005045</td>\n",
       "      <td>10.051389</td>\n",
       "      <td>7.099836</td>\n",
       "      <td>-3.343437</td>\n",
       "      <td>18.330174</td>\n",
       "      <td>L4-IT-Rorb</td>\n",
       "      <td>L4-IT-Rorb_Rorb-Tenm2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2C_M_1</td>\n",
       "      <td>180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...</td>\n",
       "      <td>0.00531</td>\n",
       "      <td>0.75683</td>\n",
       "      <td>0.75553</td>\n",
       "      <td>0.02196</td>\n",
       "      <td>0.01674</td>\n",
       "      <td>1197092.0</td>\n",
       "      <td>2590626</td>\n",
       "      <td>1709334.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>-19.267288</td>\n",
       "      <td>-7.374098</td>\n",
       "      <td>13.717241</td>\n",
       "      <td>-1.786627</td>\n",
       "      <td>9.212860</td>\n",
       "      <td>-26.514785</td>\n",
       "      <td>L23-IT-Cux2</td>\n",
       "      <td>L23-IT-Cux2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2C_M_10</td>\n",
       "      <td>180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...</td>\n",
       "      <td>0.00581</td>\n",
       "      <td>0.75720</td>\n",
       "      <td>0.75578</td>\n",
       "      <td>0.02603</td>\n",
       "      <td>0.02034</td>\n",
       "      <td>1006630.0</td>\n",
       "      <td>2348484</td>\n",
       "      <td>1606410.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>-15.340144</td>\n",
       "      <td>-2.739380</td>\n",
       "      <td>4.829485</td>\n",
       "      <td>9.943986</td>\n",
       "      <td>-22.960995</td>\n",
       "      <td>13.509483</td>\n",
       "      <td>Outlier</td>\n",
       "      <td>Outlier</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2C_M_100</td>\n",
       "      <td>180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...</td>\n",
       "      <td>0.00634</td>\n",
       "      <td>0.74984</td>\n",
       "      <td>0.74824</td>\n",
       "      <td>0.02718</td>\n",
       "      <td>0.02097</td>\n",
       "      <td>2026078.0</td>\n",
       "      <td>4599922</td>\n",
       "      <td>3154619.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>-23.625140</td>\n",
       "      <td>-5.929033</td>\n",
       "      <td>13.739521</td>\n",
       "      <td>1.107174</td>\n",
       "      <td>12.798545</td>\n",
       "      <td>-17.126331</td>\n",
       "      <td>L23-IT-Cux2</td>\n",
       "      <td>L23-IT-Cux2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2C_M_1000</td>\n",
       "      <td>180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_3_CE...</td>\n",
       "      <td>0.00521</td>\n",
       "      <td>0.77073</td>\n",
       "      <td>0.76953</td>\n",
       "      <td>0.02117</td>\n",
       "      <td>0.01604</td>\n",
       "      <td>2368464.0</td>\n",
       "      <td>5320502</td>\n",
       "      <td>3648786.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>-21.815190</td>\n",
       "      <td>-7.130878</td>\n",
       "      <td>14.154541</td>\n",
       "      <td>-2.284118</td>\n",
       "      <td>8.443658</td>\n",
       "      <td>-27.943267</td>\n",
       "      <td>L23-IT-Cux2</td>\n",
       "      <td>L23-IT-Cux2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        cell                                         FullCellID  CCC_Rate  \\\n",
       "0     2C_M_0  180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...   0.00511   \n",
       "1     2C_M_1  180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...   0.00531   \n",
       "2    2C_M_10  180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...   0.00581   \n",
       "3   2C_M_100  180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_1_CE...   0.00634   \n",
       "4  2C_M_1000  180508_CEMBA_mm_P56_P63_2C_CEMBA180409_2C_3_CE...   0.00521   \n",
       "\n",
       "   CG_Rate  CG_RateAdj  CH_Rate  CH_RateAdj  FinalReads  InputReads  \\\n",
       "0  0.77073     0.76955  0.02154     0.01651   1076259.0     2314862   \n",
       "1  0.75683     0.75553  0.02196     0.01674   1197092.0     2590626   \n",
       "2  0.75720     0.75578  0.02603     0.02034   1006630.0     2348484   \n",
       "3  0.74984     0.74824  0.02718     0.02097   2026078.0     4599922   \n",
       "4  0.77073     0.76953  0.02117     0.01604   2368464.0     5320502   \n",
       "\n",
       "   MappedReads  ... Slice PassFilter      pca_0     pca_1     umap_0  \\\n",
       "0    1517338.0  ...     2       True -11.588627 -0.005045  10.051389   \n",
       "1    1709334.0  ...     2       True -19.267288 -7.374098  13.717241   \n",
       "2    1606410.0  ...     2       True -15.340144 -2.739380   4.829485   \n",
       "3    3154619.0  ...     2       True -23.625140 -5.929033  13.739521   \n",
       "4    3648786.0  ...     2       True -21.815190 -7.130878  14.154541   \n",
       "\n",
       "     umap_1     tsne_0     tsne_1  MajorCluster             SubCluster  \n",
       "0  7.099836  -3.343437  18.330174    L4-IT-Rorb  L4-IT-Rorb_Rorb-Tenm2  \n",
       "1 -1.786627   9.212860 -26.514785   L23-IT-Cux2            L23-IT-Cux2  \n",
       "2  9.943986 -22.960995  13.509483       Outlier                Outlier  \n",
       "3  1.107174  12.798545 -17.126331   L23-IT-Cux2            L23-IT-Cux2  \n",
       "4 -2.284118   8.443658 -27.943267   L23-IT-Cux2            L23-IT-Cux2  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# metadata after QC and clustering\n",
    "# meta = pd.read_msgpack(os.path.join(SRC_DIR, 'snmc_and_snatac/study/mc/Metadata.after_QC.msg'))\n",
    "# meta2 = pd.read_msgpack(os.path.join(SRC_DIR, 'snmc_and_snatac/study/mc/mop_mc_clustering/MOp.methylome_only.analysis_results.msg')).drop(columns=['cell'])\n",
    "# print(meta.shape)\n",
    "# print(meta2.shape)\n",
    "# print(meta.columns.values.shape)\n",
    "# print(np.intersect1d(meta.columns.values, meta2.columns.values).shape)\n",
    "# meta2 = meta2.drop(columns=meta2.filter(regex='^ENSMUSG', axis=1).columns).drop(columns='AllcPath')\n",
    "\n",
    "meta3 = pd.read_csv(os.path.join(SRC_DIR, 'snmc_and_snatac/study/mc/MOp.snmC-seq.AnalysisResult.csv.gz')).rename(columns={'index': 'cell'})\n",
    "print(meta3.shape)\n",
    "meta3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta3.to_csv(fout_meta, sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "reps = ['2C-180409.mcds', \n",
    "        '2C-180410.mcds',\n",
    "        '3C-171206.mcds',\n",
    "        '3C-171207.mcds',\n",
    "        '4B-171212.mcds',\n",
    "        '4B-171213.mcds',\n",
    "        '4B-180104.mcds',\n",
    "        '5D-180605.mcds',\n",
    "        '5D-180612.mcds',\n",
    "       ]\n",
    "\n",
    "c_tables = []\n",
    "mc_tables = []\n",
    "for rep in reps:\n",
    "    mcds = xr.open_dataset(os.path.join(SRC_DIR, 'snmc_and_snatac/dataset/mc/{}'.format(rep)))\n",
    "    c_table = mcds['gene_da'].sel(mc_type='CHN', count_type='cov').squeeze().to_pandas()\n",
    "    mc_table = mcds['gene_da'].sel(mc_type='CHN', count_type='mc').squeeze().to_pandas()\n",
    "    c_tables.append(c_table)\n",
    "    mc_tables.append(mc_table)\n",
    "    \n",
    "c_tables = pd.concat(c_tables, axis=0)\n",
    "mc_tables = pd.concat(mc_tables, axis=0)\n",
    "assert np.all(c_tables.columns.values == mc_tables.columns.values)\n",
    "assert np.all(c_tables.index.values == mc_tables.columns.values)\n",
    "print(c_tables.shape, mc_tables.shape)\n",
    "    \n",
    "gene = c_tables.columns.values\n",
    "cell = c_tables.index.values \n",
    "c_table = c_tables.T\n",
    "c_table.index = gene\n",
    "\n",
    "mc_table = mc_tables.T\n",
    "mc_table.index = gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove version on ensembl id\n",
    "c_table = sparse.coo_matrix(c_table.values)\n",
    "mc_table = sparse.coo_matrix(mc_table.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['2C_M_10', '2C_M_0', '2C_M_1', ..., '5D_M_2370', '5D_M_2348',\n",
       "       '5D_M_2367'], dtype=object)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell = c_tables.index.values\n",
    "cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save mc and c tables\n",
    "\n",
    "data = {'mc': mc_table, \n",
    "        'c': c_table, \n",
    "       }\n",
    "gc_mat = GC_matrix(gene, cell, data) \n",
    "\n",
    "snmcseq_utils.save_gc_matrix_methylation(gc_mat, \n",
    "                                         fout_data_gene, fout_data_cell, \n",
    "                                         fout_data_mc, fout_data_c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ATAC \n",
    "- rename barcode and header\n",
    "- read snap files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/snatac_gene_metadata.tsv\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/snatac_gene_raw.npz\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/snatac_gene_raw.gene\n",
      "/cndd/fangming/CEMBA/data/MOp_all/data_freeze_formatted/snatac_gene_raw.cell\n"
     ]
    }
   ],
   "source": [
    "dataset_name = 'snatac_gene'\n",
    "\n",
    "fout_meta = os.path.join(DST_DIR, dataset_name+'_metadata.tsv') \n",
    "fout_data = os.path.join(DST_DIR, dataset_name+'_raw.npz') \n",
    "fout_data_gene = os.path.join(DST_DIR, dataset_name+'_raw.gene') \n",
    "fout_data_cell = os.path.join(DST_DIR, dataset_name+'_raw.cell') \n",
    "\n",
    "print(fout_meta)\n",
    "print(fout_data)\n",
    "print(fout_data_gene)\n",
    "print(fout_data_cell)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cluster\n",
      "ASC                7215\n",
      "Chodl                86\n",
      "Endo               1245\n",
      "L23.a              6085\n",
      "L23.b              5758\n",
      "L23.c              3590\n",
      "L4                 7307\n",
      "L5.IT.a            3668\n",
      "L5.IT.b            4214\n",
      "L5.PT              1755\n",
      "L6.CT              8008\n",
      "L6.IT              4233\n",
      "Lamp5_Arhgdib       237\n",
      "Lamp5_Mettl21e      171\n",
      "Lamp5_Ndnf          204\n",
      "Lamp5_Smad3         940\n",
      "MGC                3899\n",
      "NP                 1291\n",
      "OGC                8557\n",
      "OPC                3279\n",
      "Other              1571\n",
      "Pv_Ntf3_Trim63     1105\n",
      "Pv_Tac1            1944\n",
      "Pv_Vsig2            431\n",
      "Smc                 586\n",
      "Sncg                246\n",
      "Sst_Chrna2_Myh8     538\n",
      "Sst_Man1a          1355\n",
      "Sst_Stk33           246\n",
      "Vip_Chat            500\n",
      "Vip_Gcnt4           292\n",
      "Vip_Hcls1           395\n",
      "Vip_Lipg            245\n",
      "dtype: int64\n",
      "(81196, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>sample</th>\n",
       "      <th>barcode</th>\n",
       "      <th>TN</th>\n",
       "      <th>UM</th>\n",
       "      <th>PP</th>\n",
       "      <th>UQ</th>\n",
       "      <th>CM</th>\n",
       "      <th>cluster</th>\n",
       "      <th>tsne1</th>\n",
       "      <th>tsne2</th>\n",
       "      <th>umap-1</th>\n",
       "      <th>umap-2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CEMBA171206_3C_AGCGATAGAACCAGGTAAGAGATGTATAGCCT</td>\n",
       "      <td>CEMBA171206_3C</td>\n",
       "      <td>AGCGATAGAACCAGGTAAGAGATGTATAGCCT</td>\n",
       "      <td>138330</td>\n",
       "      <td>128263</td>\n",
       "      <td>128039</td>\n",
       "      <td>64868</td>\n",
       "      <td>0</td>\n",
       "      <td>L6.CT</td>\n",
       "      <td>7.605117</td>\n",
       "      <td>0.697672</td>\n",
       "      <td>-2.125204</td>\n",
       "      <td>-4.396653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CEMBA171206_3C_AGCGATAGAACCAGGTAATGACGTCAGGACGT</td>\n",
       "      <td>CEMBA171206_3C</td>\n",
       "      <td>AGCGATAGAACCAGGTAATGACGTCAGGACGT</td>\n",
       "      <td>159133</td>\n",
       "      <td>146663</td>\n",
       "      <td>146348</td>\n",
       "      <td>69167</td>\n",
       "      <td>0</td>\n",
       "      <td>L23.b</td>\n",
       "      <td>-9.492925</td>\n",
       "      <td>-6.200938</td>\n",
       "      <td>8.959074</td>\n",
       "      <td>3.006819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CEMBA171206_3C_AGCGATAGAACCAGGTAGGATAACATAGAGGC</td>\n",
       "      <td>CEMBA171206_3C</td>\n",
       "      <td>AGCGATAGAACCAGGTAGGATAACATAGAGGC</td>\n",
       "      <td>19152</td>\n",
       "      <td>17586</td>\n",
       "      <td>17545</td>\n",
       "      <td>8348</td>\n",
       "      <td>0</td>\n",
       "      <td>ASC</td>\n",
       "      <td>14.227358</td>\n",
       "      <td>-11.023748</td>\n",
       "      <td>-6.880754</td>\n",
       "      <td>-2.903271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CEMBA171206_3C_AGCGATAGAACCAGGTATAGCCTTAGGCGAAG</td>\n",
       "      <td>CEMBA171206_3C</td>\n",
       "      <td>AGCGATAGAACCAGGTATAGCCTTAGGCGAAG</td>\n",
       "      <td>14876</td>\n",
       "      <td>13837</td>\n",
       "      <td>13801</td>\n",
       "      <td>6583</td>\n",
       "      <td>0</td>\n",
       "      <td>L6.CT</td>\n",
       "      <td>5.144768</td>\n",
       "      <td>4.790674</td>\n",
       "      <td>-0.955707</td>\n",
       "      <td>-4.635163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CEMBA171206_3C_AGCGATAGAACCAGGTATAGCCTTATAGAGGC</td>\n",
       "      <td>CEMBA171206_3C</td>\n",
       "      <td>AGCGATAGAACCAGGTATAGCCTTATAGAGGC</td>\n",
       "      <td>28532</td>\n",
       "      <td>26303</td>\n",
       "      <td>26202</td>\n",
       "      <td>12437</td>\n",
       "      <td>0</td>\n",
       "      <td>L5.IT.b</td>\n",
       "      <td>5.218895</td>\n",
       "      <td>-8.674734</td>\n",
       "      <td>0.796851</td>\n",
       "      <td>4.447933</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              cell          sample  \\\n",
       "0  CEMBA171206_3C_AGCGATAGAACCAGGTAAGAGATGTATAGCCT  CEMBA171206_3C   \n",
       "1  CEMBA171206_3C_AGCGATAGAACCAGGTAATGACGTCAGGACGT  CEMBA171206_3C   \n",
       "2  CEMBA171206_3C_AGCGATAGAACCAGGTAGGATAACATAGAGGC  CEMBA171206_3C   \n",
       "3  CEMBA171206_3C_AGCGATAGAACCAGGTATAGCCTTAGGCGAAG  CEMBA171206_3C   \n",
       "4  CEMBA171206_3C_AGCGATAGAACCAGGTATAGCCTTATAGAGGC  CEMBA171206_3C   \n",
       "\n",
       "                            barcode      TN      UM      PP     UQ  CM  \\\n",
       "0  AGCGATAGAACCAGGTAAGAGATGTATAGCCT  138330  128263  128039  64868   0   \n",
       "1  AGCGATAGAACCAGGTAATGACGTCAGGACGT  159133  146663  146348  69167   0   \n",
       "2  AGCGATAGAACCAGGTAGGATAACATAGAGGC   19152   17586   17545   8348   0   \n",
       "3  AGCGATAGAACCAGGTATAGCCTTAGGCGAAG   14876   13837   13801   6583   0   \n",
       "4  AGCGATAGAACCAGGTATAGCCTTATAGAGGC   28532   26303   26202  12437   0   \n",
       "\n",
       "   cluster      tsne1      tsne2    umap-1    umap-2  \n",
       "0    L6.CT   7.605117   0.697672 -2.125204 -4.396653  \n",
       "1    L23.b  -9.492925  -6.200938  8.959074  3.006819  \n",
       "2      ASC  14.227358 -11.023748 -6.880754 -2.903271  \n",
       "3    L6.CT   5.144768   4.790674 -0.955707 -4.635163  \n",
       "4  L5.IT.b   5.218895  -8.674734  0.796851  4.447933  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = ('/cndd/Public_Datasets/CEMBA/BICCN_minibrain_data/data_freeze/snmc_and_snatac/study/atac/'\n",
    "     'yang_analysis_20190912/CEMBA_MOp.L2.cluster.meta.txt'\n",
    "    )\n",
    "meta_new = pd.read_csv(f, sep='\\t').rename(columns={'x.sp.joint@sample': 'sample', \n",
    "                                                    'x.sp.joint@cluster': 'cluster', \n",
    "                                                   })\n",
    "meta_cols = meta_new.columns.values\n",
    "meta_new['cell'] = meta_new['sample'] + '_' + meta_new['barcode']\n",
    "meta_new = meta_new[np.hstack([['cell'], meta_cols])]\n",
    "print(meta_new.groupby('cluster').size())\n",
    "print(meta_new.shape)\n",
    "meta_new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# meta = pd.read_msgpack(os.path.join(SRC_DIR, 'snmc_and_snatac/study/atac/ATAC.cell_tidy_data.msg'))\n",
    "# meta_cols = meta.columns.values\n",
    "# print(meta.shape)\n",
    "# meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# meta = meta.rename(columns={'barcode': 'sample', 'sample': 'barcode'})\n",
    "# meta['cell'] = meta['sample'] + '_' + meta['barcode']\n",
    "# meta = meta[np.hstack([['cell'], meta_cols])]\n",
    "# meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import h5py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CEMBA180409_2C\n",
      "CEMBA180410_2C\n",
      "CEMBA171206_3C\n",
      "CEMBA171207_3C\n",
      "CEMBA171212_4B\n",
      "CEMBA171213_4B\n",
      "CEMBA180104_4B\n",
      "CEMBA180612_5D\n",
      "CEMBA180618_5D\n"
     ]
    }
   ],
   "source": [
    "reps = ['CEMBA180409_2C', \n",
    "        'CEMBA180410_2C',\n",
    "        'CEMBA171206_3C',\n",
    "        'CEMBA171207_3C',\n",
    "        'CEMBA171212_4B',\n",
    "        'CEMBA171213_4B',\n",
    "        'CEMBA180104_4B',\n",
    "        'CEMBA180612_5D',\n",
    "        # 'CEMBA180605_5D',\n",
    "        'CEMBA180618_5D',\n",
    "       ]\n",
    "\n",
    "\n",
    "genes_all = []\n",
    "cells_all = []\n",
    "mat_all = []\n",
    "for rep in reps:\n",
    "    print(rep)\n",
    "    f = h5py.File(os.path.join(SRC_DIR, 'snmc_and_snatac/dataset/atac', rep+'.snap'), 'r')\n",
    "    # get genes cells mat for one replicate\n",
    "    genes = f['GM']['name'][:].astype(str)\n",
    "    cells = np.array([(rep+'_')+barcode for barcode in f['BD']['name'][:].astype(str)])\n",
    "    idx = f['GM']['idx'][:] - 1\n",
    "    idy = f['GM']['idy'][:] - 1\n",
    "    count = f['GM']['count']\n",
    "    mat = sparse.coo_matrix((count, (idx, idy)), shape=(len(cells), len(genes))).T\n",
    "    \n",
    "    # \n",
    "    genes_all.append(genes)\n",
    "    cells_all.append(cells)\n",
    "    mat_all.append(mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(53278,) (135665,) (53278, 135665)\n"
     ]
    }
   ],
   "source": [
    "for i, genes in enumerate(genes_all):\n",
    "    if i == 0:\n",
    "        ref = genes\n",
    "    else:\n",
    "        assert np.all(genes == ref)\n",
    "\n",
    "genes_merge = genes\n",
    "cells_merge = np.hstack(cells_all)\n",
    "mat_merge = sparse.hstack(mat_all)\n",
    "\n",
    "print(genes_merge.shape, cells_merge.shape, mat_merge.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(53278,)\n",
      "(53278,)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['ENSMUSG00000037552', 'ENSMUSG00000052609', 'ENSMUSG00000040624',\n",
       "       ..., 'ENSMUSG00000110345', 'ENSMUSG00000110681',\n",
       "       'ENSMUSG00000110504'], dtype=object)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_table = df_genes_v2.reindex(genes_merge).dropna()\n",
    "assert len(gene_table) == len(genes_merge) #.shape\n",
    "genes_merge_id = gene_table['gene_id'].values\n",
    "\n",
    "print(genes.shape)\n",
    "print(genes_merge_id.shape)\n",
    "genes_merge_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "gc_mat = GC_matrix(genes_merge_id, cells_merge, mat_merge)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "snmcseq_utils.save_gc_matrix(gc_mat, \n",
    "                             fout_data_gene, fout_data_cell, fout_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_new.to_csv(fout_meta, sep=\"\\t\", header=True, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
